var store = [{
        "title": "Choosing a platform for a personal website",
        "excerpt":"Hello Internet! I finally have a website. I’ve been thinking about having a website for several years now but never actually got around to making one. It was mostly due to lack of focussed time rather than lack of ideas. I probably won’t have and don’t plan on having regular popular content. So my idea was to have a personal website showcasing my work, a bit about myself, contact,  various social links and an occasional blog post. I spent quite a bit of time researching on potential set-up for the website from various hosting services to themes and templates. I thought I would go through some of the web hosting options out there especially for a personal website.   Dynamic Websites   Unhosted websites                           Unhosted websites.                If you do not want to bother with web hosting and management, the best choice is to go for a system that is ready and managed. These come as restricted free services or full-fledged paid services. Drag and drop builders are provided. No coding or server management is required. One of the most popular and beautiful looking options is Squarespace. Others include Wix, Virb, Weebly, Strikingly and lots more. Another good example is wordpress.com for example, which is heavily restricted in terms of domain name, theme modification etc. Some other options in similar category with limited customisation include Blogger and Google sites. Google sites is probably from the 1990s and still feels and looks the same. I won’t include things like Facebook or Tumblr since I don’t see them as personal websites, but rather social media.   Free hosted websites                           Free hosted websites.                There are several options for free hosted websites where you just have to create a user account and put in some details and you are set to go. These sort of web services have a range of features from custom domain names to site building tools. The best feature is of course that they are free. They might also provide free domain names like yoursite.provider.com. Most of them offer support for php, mysql databases etc with a graphical user interface without diving into code. The drawbacks with such services are that they are almost always shared which affects performance. The storage space and traffic bandwidth is limited. But, for a small personal website, this really shouldn’t be a problem. Some of such providers include 000webshost, awardspace, freewebhostingarea etc. Here is a small list on prchecker.                               Shared hosted websites   Most basic paid web hosting are so called ‘shared hosting’. The previously mentioned providers and almost all free providers also provide this sort of paid hosting. Free domain names are provided and the possibility of a custom domain is also possible. Storage and traffic bandwidth is increased. A wide range of pricing to fit your needs makes this an attractive option. Nevertheless, your website will still be shared on a physical location which means the cores and memory are shared between all other users/websites on the same computer. This might cause a drop in performance especially if another demanding website runs on the same computer. You also do not have root access to system administration. But, this option should suffice for most general users, private bloggers, small scale websites etc. PCMag has a nice list of such providers for 2015.   Virtual Private Server   These are the upgraded version of shared hosting. You have your own private area that you have full control over as if you own it. You can configure anything any way you like. Handles traffic better. You need to load your own LAMP stack (Linux, Apache, MySQL, PHP or Perl) and set up your own server management. This means more technical work and demanding system administration and self regulated security. If done properly, has more security than shared hosting. You can install your own resource-hungry softwares. There are lots of resources on the web to compare differences between shared and VPS hosting. All major web hosting providers such as bluehost, hostgator, dreamhost etc. have the VPS option.   Dedicated server   A dedicated server means that you are not getting a slice of the server, instead you get the whole server to your self. You can choose the cores and RAM and storage etc. System management and administration is demanding and usually command line. Some services may provide control panels and auto-installer scripts for commonly used applications. You can install any server OS, any web-server from Apache to Nginx. While there is effectively no sharing on computing resources, there might be bandwidth sharing. PCMag has yet another list for dedicated servers.   Cloud computing                           Cloud computing.                Cloud computing is the newest trend. Cloud computing provides users with something called instances. These are virtual machines with any configuration as you would like. Any number of cores, whatever RAM, storage etc. Instances can be created and installed with any OS on the fly in a couple of minutes. You can install, configure, management anything anyway you like. Billing is usually time based. Hourly billing is usually used for intensive high-computing instances while a monthly billing is used for lower configurations. Websites implemented through cloud computing have their content on multiple servers at multiple locations at the same time which means that they less prone to downtime if a single physical server crashes somewhere. This may not always be the case depending on the set-up. There are countless articles debating the pros and cons of cloud computing. As a bottom line, cloud computing might be overkill for hosting a small website. The most prominent players are Amazon EC2 and Digital Ocean. Others include Vultr, Linode, Google Cloud and Microsoft Azure. Amazon AWS has a free tier option for new customers for the first one year.   Static Websites                           Static websites.                Static websites contain fixed static content. They do not use server-side scripting like PHP or databases like SQL. They are simple and easy to develop and maintain. They are used for small websites where content do not change very often. Static websites use only HTML, javascript and CSS. They do not have any fancy features or functionalities. They are the cheapest to host as they do not need any sort of live server. They can be hosted off Github or even Dropbox. Beautiful looking web pages can be created and updated simply using Markdown. Some of the popular platforms for generating static sites include Jekyll, Hugo, GitBook etc. Here is Dean Attali’s website which is a good example of a static website.   Conclusion   Ultimately, after all that brain racking, I decided (for now) to go with the cloud computing option of Vultr. I am running a wordpress site. I decided to go for the cloud computing option since I also wanted a live server to run an R shiny server for R shiny applications. Or else, I would’ve opted the static website option on Github pages.  ","categories": ["Web"],
        "tags": ["Blogging","Personal Website","Web Hosting"],
        "url": "/choosing-personal-website/",
        "teaser": null
      },{
        "title": "Exploring South Sweden",
        "excerpt":"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               As you drive from Denmark into Sweden, the first thing you notice is the lower traffic density, which makes driving a lot more pleasant. The route from Helsingborg to Nybro was mostly uneventful with long roads, beautiful fields and the occasional road breaks.   Glasrikets Älgpark, Nybro (Moose Park)                           Moose park.                This was very conveniently on our route and we decided to go have a look. Its a relatively small enclosure with around 5-6 moose of various ages and sexes. The best park is that you can get really close to them, feed them and even touch them. The tour guide was friendly and explained about moose behaviour and random facts. Moose parks are all over the place in Sweden and its not hard to find one. The details for this one is here.   James Bond Museum, Nybro   Passing through Nybro, looking out, we randomly saw a sign that read James Bond Museum. We decided to check it out. It turned out to be some sort of personal museum managed and run by a James Bond fan named Gunnar Shäfer who calls himself Gunnar Shäfer Bond. The man is a middle aged eccentric lunatic who goes around driving his BMW Z3 through the streets of Nybro and brandishing his toy pistol at unsuspecting strangers. The museum might be appreciated by a hardcore James Bond fan for the random collectables and memorabilia, but for an average curious tourist, it’s just a load of  junk. The quality of the museum is evident from its website. As we were leaving, we learned that Mr Gunnar Shäfer has officially added James Bond to his name as seen in his driving license. He proudly offered to autograph one of the brochures for us which we complied reluctantly.   Norra Kvills National Park, Vimmerby                           Norra Kvills national park.                The town of Vimmerby is worth visiting. It’s a very cute little town and the home of children’s stories author Astrid Lindgren including the theme park Astrid Lindgren’s world. Vimmerby is also close to several sites of interest such as Virum Moose park, Kvill Oak, Norra Kvills national park etc. The Norra Kvills national park is a well-preserved pristine piece of primeval forest. The best feature of Norra Kvills is the large moss covered rocks.   Air-Force Museum, Linköping (Flygvapenmuseum)   The Air-Force museum is one of the biggest attractions in and around Linköping. It is worth a visit for any random tourist and an excellent location for aircraft enthusiasts such as myself. The museum is the home to numerous full-size military airplanes from various eras of Swedish aviation history. You can get up close of the famous aircrafts such as the Douglas DC-3, Mustang P-51, Gripen etc. The musuem also has several other collections of military interests such as weapons, ammunitions and historic relics. This museum can be thought of as the smaller brother of the big badass Imperial War Museum in London.   Tåkern Bird Park                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Tåkern bird park          The Naturum Tåkern is a protected bird park and an ideal location to observe birds. The visitor centre in itself is an intriguing piece of architectural work. The visitor centre has everything you need to know about birds. There are lots of educational material such as bird identification, sounds and behaviour. There are also models of all the 200 or so birds spotted at Tåkern lake over the years.   Ekopark Omberg   Omberg is located on the east coast of Sweden’s largest lake Vätten between Jönköping and Motala. Ekopark in Omberg is a forest covered hill managed by the Swedish forest authority Sveaskog. There are several view point across the cliffs providing a spectacular view of the coastline. This is a great location to hike, pick berries, enjoy nature and to relax. I am able to find an excellent cliff-side to camp.                           Cliff side camping.                Omberg was the last major stop, we drove past some beautiful lakes, farms and meadows all the way back to Uppsala.                               ","categories": ["Photography","Travel"],
        "tags": ["Camping","Forest","National Park","Nature","Road Trip","Sweden","Wildlife"],
        "url": "/exploring-south-sweden/",
        "teaser": null
      },{
        "title": "Low-Polygon Portrait",
        "excerpt":"I ran into this video and was intrigued by the idea of creating a low-poly portrait for myself. So I thought of documenting the steps involved.        The first step is to find a good portrait with a clear face. Search on the web for _low-poly face_ to get an idea of what sort of portraits work. It doesn't have to be a portrait, but a low-complexity image works best. Since I didn’t have the patience to take a portrait for this purpose alone, I went through my phone and found a picture that I thought might work.      Import the image into _Adobe Illustrator_ or any vector handling application, add a new layer above the image and start creating polygons (2). This is the most time-consuming part. Here are some suggestions for creating polygons:      Create polygons for areas with similar colours/tones.    Stick to 3 sided polygons (triangles). It works best.    Larger and fewer triangles for low complexity areas and several smaller triangles for higher complexity areas.    Direction and general flow of the mesh must compliment the contours in the image.    Too many or too few triangles can ruin the final result.    Take your time to do the polygons as this makes all the difference.           Complete all the polgons and connect the edges. Do not leave any gaps.    Once the polygons are done, select each polygon and with the *eyedropper* tool, fill it with the colour at the centre of the polygon in the image.    Once the polygons are all coloured, you might want to change the background or leave it plain. I decided to go for a defocussed foliage with sunlight as my background. Export the image as a .png with transparent background and import into *Adobe Photoshop* or any other image editing software. Position the poly layer over your desired background as you see fit. Adjust brightness/contrast, hue/saturation of both layers as required.    Flatten the layers and use _Lighting effects_ filter to add a directional light/spot light. Do not overdo this. Then add a _Photo filter_ if required to further blend the portrait and background. Go to _Curves_ and lower highlights in blues slightly to create softer whites. Add a light vignette and bam! You are done.                                  Here is the final result.                               A detailed tutorial is on this blog by the original author.  Don’t forget to check out the comments section of the blog for user generated low-poly portraits.  ","categories": ["Design"],
        "tags": ["Adobe Illustrator","Adobe Photoshop","Portrait Design"],
        "url": "/low-poly-portrait/",
        "teaser": null
      },{
        "title": "Markdown basics",
        "excerpt":"Text formatting   Headings can be defined as shown below.   # Level 1 heading ## Level 2 heading   ### Level 3 heading   #### Level 4 heading   ##### Level 5 heading   ###### Level 6 heading   Level 2 heading  Level 3 heading  Level 4 heading  Level 5 heading  Level 6 heading   Six custom classes are defined for text scaling. This can be defined inside a paragraph (&lt;p&gt;) or &lt;span&gt;.   &lt;span class=\"largest\"&gt;Largest text.&lt;/span&gt;    &lt;span class=\"larger\"&gt;Larger text.&lt;/span&gt;   &lt;span class=\"large\"&gt;Large text.&lt;/span&gt;   Normal text.   &lt;span class=\"small\"&gt;Small text.&lt;/span&gt;   &lt;span class=\"smaller\"&gt;Smaller text.&lt;/span&gt;   &lt;span class=\"smallest\"&gt;Smallest text.&lt;/span&gt;     Largest text.   Larger text.  Large text.  Normal text.  Small text.  Smaller text.  Smallest text.   A horizontal line can be created using three or more * or -.   ***     This is __Bold text__ This is Bold text  This is _Italic text_ This is Italic text  ~~Strikethrough~~ text Strikethrough text   This is Subscript H~2~O displayed as H~2~O  This is Superscript 2^10^ displayed as 2^10^  This is a [link](r-project.org) This is a link  An example of footnote reference 1   &gt; This is a block quote. This &gt; paragraph has two lines. &gt; &gt; 1. This is a list inside a block quote. &gt; 2. Second item.      This is a block quote. This paragraph has two lines.          This is a list inside a block quote.     Second item.      Icons from FontAwesome can be displayed using the HTML &lt;i&gt; tag. Note that not all icons may work.   Here is a &lt;i class='fa fa-calendar'&gt;&lt;/i&gt; calendar and a &lt;i class='fa fa-couch'&gt;&lt;/i&gt; couch.   Here is a  calendar and a  couch.   Code formatting   Code can be defined inline where `this` looks like this. Code can also be defined inside code blocks.    ``` This is code ```   This is code   Code formatting can also be achieved by four spaces        This is code   This is code   Lists  Bulleted List   Unordered lists are created using asterisks.      Bullet 1   Bullet 2            Sub-bullet 2.1       Sub-bullet 2.2           Bullet 3   Ordered lists are created using numbers.      Point 1   Point 2   Point 3   Images   Images can be placed as such.   ![](/assets/images/featured.jpg)      The dimensions are based on image and/or fill up the entire available space. You can control the dimension as shown below.   [&lt;img src=\"/assets/images/featured.jpg\" width=\"30%\"/&gt;](/assets/images/featured.jpg)       This image above is now 30% of it’s original width.   Tables   Simple tables can be created using markdown. Below is an example of a table.    |Sepal.Length|Sepal.Width|Petal.Length|Petal.Width|Species| |---|---|---|---|---| |5.1|3.5|1.4|0.2|setosa| |4.9|3.0|1.4|0.2|setosa| |4.7|3.2|1.3|0.2|setosa|   which renders into an HTML table                  Sepal.Length       Sepal.Width       Petal.Length       Petal.Width       Species                       5.1       3.5       1.4       0.2       setosa                 4.9       3.0       1.4       0.2       setosa                 4.7       3.2       1.3       0.2       setosa                         That reference refers to this footnote. &#8617;           ","categories": ["Code"],
        "tags": ["markdown"],
        "url": "/markdown-basics/",
        "teaser": null
      },{
        "title": "Hamra National Park",
        "excerpt":"Hamra National Park is 14 sq km of wide open marshlands, woods, forests, streams and rugged trails. Founded in 1909, Hamra national park is one of the oldest preserved forests in Sweden.   Hamra national park (NP) is one of sweden’s nine oldest national parks, located at the western edge of the Gävleborg municipality (Kommune). We took a 2-day trip in the last week of August. The drive from Uppsala to Hamra takes about 4.5 hours. We stayed at the STF accomodation in the nearest village called Fågelsjö. The house is nice and comfy with several bedrooms and a pretty lawn. This was the off-season which meant that we had the whole house to ourselves. Hamra NP is a short drive away from Fågelsjö.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  Hamra national park has several well marked trails of which the Hamra trail (red route) is the longest covering 10km. We took the Hamra trail on day 1 and the short virgin forest trail (yellow route) on day 2. The Hamra trail takes you through a diverse landscape consisting of woodland forests, marshlands and swamps. Most of the national park consist of wetlands which makes it demanding to walk through.                           Map and guide to Hamra National Park from the Swedish National Parks authority.                A lot of the Hamra trail is is marshes and bogs with ankle high water to wade through. A solid pair of hiking boots or waterproof gumboots would be the suitable choice. In fact, anything would be better than my choice of Nike trainers which were totally wet 10 mins into the trail. The wettest and the most unstable regions of the trails are installed with log bridges which makes it easy and convenient to walk through.   Hamra National Park is one of 29 national parks in Sweden. It was one of the first nine national parks which was established in Sweden in 1909. During a visit to Hamra National Park, you can experience bearded forests, expansive marshes and wild rapids. - Sverigesnationalparker.  Larger plant life consists of vast areas of firs, pines, and larches. Most old and dying wood are covered by reindeer lichens, moss and various fungi. Other smaller shrubs and bushes include crowberries, heather, marsh orchids, creeping lady’s tresses and heath spotted orchids.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Some of the common plant life.          Hamra NP is also home to a range of animal life from insects to birds and mammals. The NP is claimed to host 450 different species of beetles. Bird life includes swans, dippers, black grouse (capercaillies), woodpeckers, owls as well as green shank, snipes and sandpipers. Large mammals include the moose (Alces alces, also called elk in Europe and Sweden), beavers, wolves and lynx. Hamra NP is also home to one of the largest densities of brown bears. Wild Sweden has a nice article about wildlife in Sweden. This is all text book information. In reality, the only animals we came across were mosquitoes, butterflies and dragonflies. We also saw a couple of capercaillies, a frog and a moose run across the road in the night. Surprisingly, there was very little bird life, or any animal life for that matter. We did come across occasional fecal matter which we assume to be some sort of larger berry-eating animal. The whole place had a sterile feel to it. It is possible that the animals are familiar with the trail route and prefer to stay away.   The national park has excellent facilities for rest, view points, observation towers and barbeque sites. We had picked Hamra as it was closet NP one could go from Uppsala that would have minimal light pollution. Hamra lies in relatively dark neighbourhood in terms of light pollution. We had hoped on finding auroras or starry skies. Unfortunately, we were greeted by the full moon which ruined our night skies.                           Long hair-like lichens grow on the old wood.                On day 2, we took the yellow route which trailed through the oldest part of Hamra NP; the old growth forest. The most interesting aspect is the dead trees which provide food and shelter to many specialised species of plants and animals. Old and dead trees have rough barks which are covered with hanging lichens and mosses. This gives the appearance of a tuft of hair. This is why these forests are sometimes called bearded forests. Siberian jays are known to hide food in the lichens. The yellow trail is also more dense with vegetation. Carnivorous sundew plants grow along the side of the trail in some places.                           Spiral grooves on the stem.                There are many curious trees with spiral barks. It turns out they are not barks after all. Those are the xylem vessels visible after the barks have fallen off. But xylem channels mostly run straight and these spiral columns are usually rare to find. In this case, the xylem vessels curve in a certain direction as they grow and the spiral becomes more obvious with age. Some explanations for spiral growth include greater fexibility for the tree or better distribution of water around the tree. Either way spiral wood is undesirable for woodwork as they are weaker and brittle.   The NP has outdoor activities for every season. In winter, the snow is often very deep and its perfect for cross-country skiing. Husky sledge rides are also popular. Perhaps I should think about coming back in winter.                               ","categories": ["Photography","Travel"],
        "tags": ["Forest","Hiking","National Park","Nature","Road Trip","Sweden","Wildlife"],
        "url": "/hamra-national-park/",
        "teaser": null
      },{
        "title": "Elegant scientific graphs: Learning from examples",
        "excerpt":"Scientific graphics are key to understanding complex data. In addition to graphing data, it is important that the graphics are clean, elegant and visually stunning. This post explores examples of attractive graphs from popular magazines and news websites.   This is the flow of understanding data.   Data collection &gt; Data analysis &gt; Data visualisation   The most important part of a plot is of course its content. Once you have good content/data, you need to think about how to represent this data. Which sort of plot to use. How to best convey this information. See this article for most common types of data visualisation. Some of the popular programming environments for plotting include R and Python with Julia being the latest addition. Other options include Excel, Tableau, Plotly and more. R is a great tool for graphics as evident from the numerous images, blogs and publications over the past years. There are several resources that can help you find code to create all sorts of graphs in R.     But, just managing to create a graph is not sufficient in my opinion. The graphic has to be beautiful, elegant, user-friendly and attractive. Getting from data to a plot is one thing, but creating a high-quality, publishable and professional looking figure is a different story. A Basic plot is the initial basic output figure from any plotting software. This uses the default setting and default looks. Most people stop at this point because they have gone through considerable effort to get the data, analyse it and finally figure out the code to plot it. But a basic plot is usually not going to look professional or elegant. It will need some level of customisation to make it attractive. The ggplot2 plotting package in R, for example, produces a pretty decent default output, but they are overused and the graphics are not unique or catchy. I haven’t come across many sources that go into the fine details of the making a professional looking graph.  We will go into plot customisation in R in a future post, but, in this post, we explore some examples of customised plots.   Professional graphs   The first step to creating elegant plots is to understand what makes a good plot. Since its hard to say exactly what is a good-looking graph and a bad-looking one, I will use examples. I usually take figures, plots or graphics in popular magazines as a quality benchmark. My favourite choices are scientific journals like Nature and Science that produce customised graphics, scientific magazines such as National Geographic, Scientific American, New Scientist, American Scientist etc. and popular news websites such as BBC, The Guardian, The Huffington Post, The Telegraph, The New York Times etc. They usually have high quality data visualisation, illustrations, figures and graphics. This post will focus more on graphs and plots. We will avoid highly customised complex illustrations, infographics and interactive graphics for now. Those will be perhaps dealt with in a future post. The examples are roughly organised from simple to complex graphics.   BBC   The BBC example used is from a recent article on the European migrant crisis.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  These are typical BBC graphics: clean and minimal. Text is clear and legible. Colours are few, distinct and catchy. Text colour is around 80% grey. The first figure combines a stacked horizontal barplot with an annotated map. Colours in the barplot match the arrows on the map. Second and third figures show standard barplots. Notice the use of simple gridlines to enable easier comparison of values. Axes lines are removed where not necessary. Second and third figures show standard barplots. In the fourth figure, notice the use of white background fill colour for city names to increase legibility. The fifth figure uses circles on maps to denote number of claims per country. Notice the use of white border for text on dark background. In the last figure, notice the use of sequential colour palette to represent binned data. And neighbouring countries and lines have been removed to reduce clutter.   National Geographic   We will focus on two graphs from an article on Dolphin intelligence.                           Comparing brain evolution between humans and dolphins.                Here we have a timeline plot. Small pictograms (to scale) allow the viewer to quickly understand the animals in comparison and get a sense of their sizes. Subtle colours (dark teal blue &amp; grey) are used but they are easy to differentiate. Timeline annotation and gridlines are about 20% grey and muted such that they can be used as guide but not strong enough to overpower the subject. Notice the font size and the font weight. The three important events on the timeline are emphasized with bold font.                           Hearing range in animals.                Even without a caption, the boldest font ‘Hearing range’ captures your attention. The figure compares the hearing frequency range of five animals. This is a slight variation of a regular barplot. The bars don’t always go to zero. The dog for example can only hear down to about 44 hertz. Note the number of colours used and their hues. Note the font sizes and the use of full caps. Also, note that all figures in the article use a similar colour scheme and consistent look.   Here are a few more examples below showcasing customised lineplots, barplots, maps and slopecharts.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 The figures are from the article: Weather Gone Wild (Sep, 2012), Beauty Markets (Dec, 2012), Cheetas on the edge (Nov, 2012) and Frank Rankings (Sep, 2012).          Have a look at the two plots below. They both show the same information. On left is what they call a slopechart and right is a scatterplot. I think they are both good and it’s a matter of personal taste as to which is better. Here is a discussion about the use of these chart types.                           Comparing the same data as a slopegraph and a scatterplot.                The Huffington Post                           The Huffington Post article on lead poisoning in the US.                Here is an example from The Huffington Post on the risk of lead poisoning. The figure shows lineplots, scatterplots and barplots. Most of the text is in black although guide text is light grey. The colour scheme is shared between the three plots as they belong as a group. Selected data points are highlighted on the scatterplot. Notice level 1, level 2 and level 3 title fonts.   Thomson Reuters                           Unemployment rates in Europe.                Here is a Reuters graphic showing unemployment rates in Europe using a barplot combined with map. The colour scheme is a bit strange in using different colours for sequential continuous data. The title and number font may seem a bit too bold and black (85% gray). But, the graphic is still quite appealing in general. The eurozone countries are marked by crossline shading. Neighbouring countries are grayed out to be less prominent but still puts the subject in context.   The New York Times   The examples from The New York Times include the value of honey bees in pollination (barplot), chemical weapons found in Iraq, deaths in Afghanistan (barplots and maps) and the ebola outbreak.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    Relying on bees is an interesting use of a stacked barplot along with a table. The chemical weapons graphic is a standard map with subtle colours and location names. The deaths in Afghanistan page has a collection of barplots and scatterplots on maps. The greyish colour scheme is clear and simple and frequently used in The New York Times graphics. Events of interest are marked on the timeline. The ebola outbreak page has a few examples of simple barplots, lineplots and maps. The last figure is a customised dotplot showing new ebola treatment centers. Another great example is below.                           Fukushima exclusion zone.                This example combines a map plot with table and text. The map shows the region around the nuclear plant, estimated population in the surrounding areas (shown in a subtle yellow colour) and the distance from the plant as concentric circles. The distance markers are combine with an inline barplot and table to reveal more data. This is more in the realm of infographics.   In R, barplots can be plotted using the base function barplot as well as using lattice and ggplot2 packages. Creating maps in R can be achieved using a variety of functions from the base function maps to packages ggplot2, ggmap and RGoogleMaps. Creating maps in base R is like a jungle as one needs several functions spread across numerous packages such as maptools, mapproj, raster, rgeos, rgdal, sp, spatstat etc. depending on the task at hand. Interactive mapping options are available from the rCharts and ggvis package. Choropleth maps are very popular to visualise all sorts of variables over spatial data.   The Wall Street Journal   The Wall Street Journal is quite well known for their stellar graphics. They even have a guide to information graphics.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        The first figure shows the external debt of some Asian countries. These graphs are referred to as dotplots or stripcharts. And this is an excellent example of how and when to use it. When there are few data points as such, it better to show the actual points rather than a mean or boxplot. Dotplots are not ideal when dealing with too many points. The second figure shows the value of advertising during the superbowl from this article. The figure shows the proportion of advertising categories and the absolute value over time. It’s similar to a stacked barplot except that heights are not fixed. Each bar is 100% of the expenses but adds up to a different amount each year. Each bar is also sorted top to bottom by highest to lowest spending category. It’s easy to visualise the rise in automobile advertising over the years. The third figure is a good example of heatmap from this article on diseases and vaccination. The last figure shows a heatmap of unemployment rate in the US over time. Heatmaps are great for easily visualising trend in tabular data. Notice the complex colour palette for the disease heatmap and a simple diverging red-green heatmap for the unemployment graph. In fact, the palette is not just a simple red-green, it is also colour-blind compatible. Test this figure at Coblis.   Here is a complex example from The Wall Street Journal combining dotplots, scatterplots and what-not. The figure is quite pretty although I am not sure exactly what it is about.                           A complicated, but pretty figure.                In R, dotplots can be created using the base functions dotchart or stripchart, or dotplot from lattice package or using the ggplot2 package. Heatmap options in R include the base function heatmap, heatmap.2 from gplots package and using the ggplot2 package.   The Guardian                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                The Guardian has some beautifully customised graphics. Some of them use colours that are bold and attention seeking. The Guardian uses their own font Guardian Egyptian. A few examples are shown here are mapping rainfall in the UK, world CO2 emissions in 2008-2009, a decade of Iraq war visualised and the 2015 refugee crisis around the Hungary-Serbia border. Differentially sized circles has been used in several infographics from The Guardian. Sometimes, there are better ways to represent the data than circles. Here is a discussion on it. The last two examples have muted colours. Note the use of an inset world map to direct the viewer to the region on the main map.   The Guardian is a good source of pie charts and donut charts.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            They are usually bright and colourful. They come in various levels of complexity. These examples are from the following articles on human rights, household food waste and UK government spending. No talk about pie charts is complete without a visit here.   In R, pie charts can be created using the base function pie as well as using ggplot2 package. There is also the option of function gvisPieChart from the googleVis package for interactive pie and donut plots.   Scientific American   Scientific American usually has heavily customised graphics with illustrations which is more in the realm of infographics. This is not possible without major investment in time as well as some serious digital arts skills. Nevertheless, it can still be used as inspirational material to get creative about how to best represent your data. The colours and fonts should also be noted. Check out Jens Christiansen’s work for SA.                           Comparison of Ebola outbreaks in the recent history.                Here is an article on Ebola outbreaks which shows an interesting way to represent multivariate data, something called radarplots. There is a lot of debate on the use of radarplots with regard to their readability and comparison on independent axes. In this example, I think this is a good choice of data visualisation. The four variables denote mortality rate, infected individuals, transmission clusters and number of days of outbreak. Essentially, smaller values are bad, but larger values are worse. Therefore, just looking at the area of the polygon gives you a sense of how bad the outbreak was and makes it easy to compare to other outbreaks. Radarplots are not advisable for all types of data and fewer the variables, the better. Radarplots can be created in R using the base function stars, radial.plot from plotrix package or package ggplot2.                           A phylogenetic tree showing the divergence of dog from wolf.                This is a phylogenetic tree commonly used to represent the divergence of species over time from the article: How Wolves became Dogs. Notice the vertical timeline with length of lines representing time, width of lines representing population sizes, variations of similar colours to denote clades and pictograms to identify species. The tight narrow region with few individuals at 15,000 years is a bottleneck event which then led to a radiation of lineages.   Phylogenetic trees in R needs the use of several packages such as ape, ade4, adegenet, adephylo, pegas, phangorn etc. This book has pretty much everything you need to know about working with phylogenetics in R.   New Scientist                           Network graph of journal citations by authors working in the field of stem cell research.                This article called Paper Trail: Inside Stem Cell Wars from June, 2010 uses a network graph to representation article citations between authors working in the field of stem cell research. Here is a discussion on this figure. Network graphs are useful to represent relationships between objects. Network graphs are widely used to present social media connections. Network graphs can be created in R using packages igraphs or ggplot2 and network.   Another approach to presenting relationships is called Sankey plots. This has become more popular lately.                           Sankey plot showing US petroleum consumption.                Here is a Sankey plot showing US petroleum consumption from this webpage. Static Sankey plots can be created in R using the riverplot package and interactive versions can be created using the rCharts or googleVis packages.      Here is an example of a treemap. They were invented for visualising distribution and usage of storage space on hard drives which is still their best application. TreeMaps can be created in R using packages portfolio or treemap.   Conclusion   In summary, we have explored examples of elegant graphics from some of the leading magazines and newspapers, looked at various types of plots, discussed properties of some of the plots and noted the functions used to create these plots in R. There are of course lot more functions and packages in R to do the same task. It is beyond the scope of this post to go into the details of every function. The purpose of this post to take a good look at high quality graphs, so that we are aware of what to aim for, while creating our own figures. In future posts, I will go into creating customised plots in R.   Further reading   Interested in more scientific figures and data visualisation? Here are some links. Some of them fall into the category of infographics.      http://weather-radials.com/   http://flowingdata.com/   http://www.informationisbeautiful.net/   http://betterfigures.org/   http://chartsnthings.tumblr.com/   http://coffeespoons.me/   http://spatial.ly/   http://www.r-bloggers.com/   http://www.dailyinfographic.com/   http://www.coolinfographics.com/   http://visual.ly/get-inspired  ","categories": ["Design","Research"],
        "tags": ["Data Visualisation","R","Scientific Graphs"],
        "url": "/elegant-scientific-graphs-learning-from-examples/",
        "teaser": null
      },{
        "title": "A guide to elegant tiled heatmaps in R",
        "excerpt":"This post/code is now outdated. Please see this new post for updated code.   This was inspired by the disease incidence rate in the US featured on the Wall Street Journal which I mentioned in one of the previous posts. The disease incidence dataset was originally used in this article in the New England Journal of Medicine. Here, I use the measles level 1 incidence (cases per 100,000 people) dataset obtained as a .csv file from Project Tycho. Download the .csv file here or head over to Project Tycho for other datasets.   In this post, we will look into creating a neat, clean and elegant heatmap in R. No clustering, no dendrograms, no trace  lines, no bullshit. We will go through some basic data cleanup, reformatting and finally plotting. We go through this step by step. For the whole code with minimal explanations, scroll to the bottom of the page.   I am running R version 3.2.0 64-bit on Windows 8.1 64-bit. Packages in use are ggplot2 (1.0.1), reshape2 (1.4.1), plyr (1.8.3) and Cairo (1.5-9). Install the necessary packages if not already installed and load them.   #install packages install.packages(pkgs = c(\"ggplot2\",\"reshape2\",\"plyr\",\"Cairo\"), dependencies = T)  #load packages library(ggplot2) #ggplot() for plotting library(reshape2) #melt(), dcast() for data reformatting library(plyr) #ddply() for data reformatting library(Cairo) #better aliasing of output images   Data preparation   Read in the .csv file and inspect the data. The first two rows with non-table data is skipped.   #read csv file m &lt;- read.csv(\"measles_lev1.csv\",header = T,stringsAsFactors = F,skip = 2)  #inspect data head(m) str(m) table(m$YEAR) table(m$WEEK)   The head() function shows us the header names and the first 6 rows of the data. The str() function shows that YEAR and WEEK columns are stored as integers and the incidences as characters. The incidences although numeric have been read in as character because missing values are coded as “-“.  The table() function is used to check for any missing years or weeks. The data is currently stored in so called ‘wide’ format which we will convert to ‘long’ format. The ggplot2 plotting package prefers long format. The YEARand WEEK variables are kept as is and all incidence values are collapsed into a variable and value column.   #convert data to long format m2 &lt;- melt(m,id.vars = c(\"YEAR\", \"WEEK\")) #rename column names colnames(m2) &lt;- c(\"Year\", \"Week\", \"State\", \"Value\")  #inspect data head(m2) str(m2)                           (Left) shows data in wide format while (right) shows data in long format.                The State variable is currently in all caps and multi-word states have dot separators. I prefer to have them in Camel case and separated by space to be shown on the plot later.   #custom function to convert to camel case camelCase &lt;- function(string=NULL,separator=\"\\.\") { if(is.null(string)) stop(\"No input string.n\") s &lt;- strsplit(string, separator) s &lt;- tolower(s[[1]]) paste(toupper(substring(s, 1, 1)), substring(s, 2),sep = \"\", collapse = \" \") }  #change variable to character, convert to camel case, #remove dot and change variable back to factor m2$State &lt;- factor(as.character(sapply(as.character(m2$State),camelCase)))   A custom function is used to change the states into camel case. Multi-word states are split at the dot separator and each word is changed into camel case. Now we change other columns Year and Weekinto factors. We also need to convert the Value field from character to numeric and change the “-“ to NA. We could use sub() function to replace the “-“ with NA. But, an easier way to to just use as.numeric() which automatically converts non-numeric content into NAs.   #change variable types m2$Year &lt;- factor(m2$Year) m2$Week &lt;- factor(m2$Week) #also converts '-' to NA m2$Value &lt;- as.numeric(m2$Value)   Now, I would like to plot the heatmap with the Year on the x-axis and State on the y axis, which means we have to deal with the Week variable in some way. We will sum all the incidences from all weeks for each year and discard the Week variable. The ddply() function from package plyr is suitable for this purpose. We need to use the sum() function inside ddply() to sum incidences over weeks. The way sum() handles NAs is a bit strange. By default, it returns NA if one or more elements in the input vector is NA. If we set argument na.rm=TRUE, then NAs are removed and the remaining numbers are summed. But if all the elements are NA, the sum is returned as zero. This is weird and undesirable in this situation. Therefore, I have a custom sum function to remove NAs and return the remaining sum or return NA if all elements are NA. We then use this custom function inside ddply() function to summarise the data by year and state while getting rid of week.   #custom sum function returns NA when all values in set are NA, #in a set mixed with NAs, NAs are removed and remaining summed. naSum &lt;- function(x) { if(all(is.na(x))) val &lt;- sum(x,na.rm=F) if(!all(is.na(x))) val &lt;- sum(x,na.rm=T) return(val) }  #sums incidences for all weeks into one year m3 &lt;- ddply(m2,c(\"Year\",\"State\"),Incidence=round(naSum(Value),0),summarise)  #inspect data head(m3) str(m3)                           Long format data ready for plotting in ggplot2                At this point, the data preparation is essentially over. The data is in ‘long’ format, the x, y and z variables for the plot are available and are of the correct type: factor, factor and numeric. If your data is already in this format, it is easy to jump straight into visualisation. But depending on what sort of data you start off with, the data preparation and reformatting can be complicated and tedious.   Plotting the data   I prefer to use the ggplot2 plotting package to plot graphs in R due to its consistent code structure. I will focus mostly on ggplot2 code here. But, just for the sake of completeness, I will also include some basic heatmap code using base graphics.   Plotting using ggplot2   Once the data is in the right format, plotting the data is rather simple code in ggplot2. The ggplot2 index page has the code syntax and arguments.   #basic ggplot p &lt;- ggplot(m3,aes(x=Year,y=State,fill=Incidence))+ geom_tile()  #save plot to working directory ggsave(filename=\"measles-basic.png\",plot = p)                           The basic plot produced from ggplot2.                The default vanilla output from ggplot2 is decent and quite frankly impressive. A hell of a lot better than the monstrosity produced from the base graphics function heatmap(). But of course, there are several aspects of the basic plot that can be modified and tweaked. To begin with, the years on the x-axis can be rotated 90° or 45° to make it more readable. The ggplot code is modified below.   #modified ggplot p &lt;- ggplot(m3,aes(x=Year,y=State,fill=Incidence))+ #add border white colour of line thickness 0.25 geom_tile(colour=\"white\",size=0.25)+ #remove x and y axis labels labs(x=\"\",y=\"\")+ #remove extra space scale_y_discrete(expand=c(0,0))+ #define new breaks on x-axis scale_x_discrete(expand=c(0,0), breaks=c(\"1930\",\"1940\",\"1950\",\"1960\",\"1970\",\"1980\",\"1990\",\"2000\"))+ #one unit on x-axis is equal to one unit on y-axis. #maintains aspect ratio. coord_fixed()+ #set a base size for all fonts theme_grey(base_size=8)+ #theme options theme( #bold font for both axis text axis.text=element_text(face=\"bold\"), #set thickness of axis ticks axis.ticks=element_line(size=0.4), #remove plot background plot.background=element_blank(), #remove plot border panel.border=element_blank())  #save with dpi 150 and cairo type ggsave(filename=\"measles-mod1.png\",plot = p,dpi=150,type=\"cairo\")                           ggplot2 heatmap after customisation.                White border lines are added to each cell, x-axis and y-axis titles are removed, size of y.axis text is decreased, breaks on x-axis is changed to show every 10 years and export options use 150 dpi and type is set to cairo which produces better quality graphics.   I would prefer the y-axis labels (states) to be ordered ascending top-bottom. This means going back to our ‘long’ format data and refactoring the State variable in reverse. The fill variable here (Incidence) is a continuous variable and hence, ggplot by default uses the blue colour ramp. Here and in many cases, it might make better sense to bin the continuous data into levels and represent each level as a discrete colour. The cut() function in R allows to break and label a continuous variable. Perhaps, better colours can make the figure pop.   #reverse level order of state m3$State &lt;- factor(as.character(m3$State),levels=rev(levels(m3$State)))  #create a new variable from incidence m3$IncidenceFactor &lt;- cut(m3$Incidence, breaks = c(-1,0,1,10,100,500,1000,max(m3$Incidence,na.rm=T)), labels=c(\"0\",\"0-1\",\"1-10\",\"10-100\",\"100-500\",\"500-1000\",\"&gt;1000\"))  #change level order m3$IncidenceFactor &lt;- factor(as.character(m3$IncidenceFactor), levels=rev(levels(m3$IncidenceFactor)))   The Incidence variable is binned into 7 levels and saved as a new variable IncidenceFactor. The NAs remain as NA. Defining breaks in your variable depends on the type of data, the number of bins that make sense with the context or just trial and error. Too many bins are not good. Checking your variable using summary(x) or a boxplot(x) can reveal a lot about the data.   #define a colour for fonts textcol &lt;- \"grey40\"  #modified ggplot p &lt;- ggplot(m3,aes(x=Year,y=State,fill=IncidenceFactor))+ geom_tile()+ #redrawing tiles to remove cross lines from legend geom_tile(colour=\"white\",size=0.25, show_guide=FALSE)+ #remove axis labels, add title labs(x=\"\",y=\"\",title=\"Incidence of Measles in the US\")+ #remove extra space scale_y_discrete(expand=c(0,0))+ #custom breaks on x-axis scale_x_discrete(expand=c(0,0), breaks=c(\"1930\",\"1940\",\"1950\",\"1960\",\"1970\",\"1980\",\"1990\",\"2000\"))+ #custom colours for cut levels and na values scale_fill_manual(values=c(\"#d53e4f\",\"#f46d43\",\"#fdae61\", \"#fee08b\",\"#e6f598\",\"#abdda4\",\"#ddf1da\"),na.value=\"grey90\")+ #mark year of vaccination geom_vline(aes(xintercept = 36),size=3.4,alpha=0.24)+ #equal aspect ratio x and y axis coord_fixed()+ #set base size for all font elements theme_grey(base_size=10)+ #theme options theme( #remove legend title legend.title=element_blank(), #remove legend margin legend.margin = grid::unit(0,\"cm\"), #change legend text properties legend.text=element_text(colour=textcol,size=7,face=\"bold\"), #change legend key height legend.key.height=grid::unit(0.8,\"cm\"), #set a slim legend legend.key.width=grid::unit(0.2,\"cm\"), #set x axis text size and colour axis.text.x=element_text(size=10,colour=textcol), #set y axis text colour and adjust vertical justification axis.text.y=element_text(vjust = 0.2,colour=textcol), #change axis ticks thickness axis.ticks=element_line(size=0.4), #change title font, size, colour and justification plot.title=element_text(colour=textcol,hjust=0,size=14,face=\"bold\"), #remove plot background plot.background=element_blank(), #remove plot border panel.border=element_blank())                           Further customisation of the ggplot2 heatmap.                This is the final version of the figure. All font elements are coloured grey40.  Missing values (NAs) are coloured grey90. The year of the introduction of vaccination is indicated as a dark vertical stripe. If the y-axis labels are too many or too small, they can be dropped the same way as the x-axis. A custom colour palette was used from ColorBrewer based on the Spectral palette. Using the R package RColorBrewer() or using ggplot2 function scale_fill_brewer() opens up all the colour palettes from the ColorBrewer website. Here is an example below:   library(RColorBrewer)  #change the scale_fill_manual from previous code to below scale_fill_manual(values=rev(brewer.pal(7,\"YlGnBu\")),na.value=\"grey90\")+                           Heatmap with Colorbrewer colour palette.                Most of the plot customisation happens in the theme() section of the code. Refer to the ggplot index for all theme arguments. Based on where the figure will go next, the fonts may need to be changed. The extrafont() package comes handy. Another option is to export in vector format such as .svg or .pdf. Import into a vector editor such as Adobe Illustrator and add your own text. This will produce nice crisp results but does take some manual work. See the cover image of this post for example.   Plotting using base graphics   We will take a quick look at plotting using base graphics. The base function heatmap() and the enchanced heatmap.2() function from the gplots package uses a wide format data matrix as input. We start with the ‘long’ data that we prepared in section 1. We convert the data in ‘long’ format into ‘wide’ format using the dcast() function from package reshape2.   #convert long format to wide format m4 &lt;- dcast(m3,Year~State,value.var = \"Incidence\") #remove state names and create a matrix m5 &lt;- as.matrix(m4[,-1]) #assign state names as rownames of matrix rownames(m5) &lt;- m4$Year   The wide data is converted to a matrix after removal of non-numeric columns. The state names are reassigned as rownames of the matrix to be used as y-axis text.   png(filename = \"measles-base.png\",height = 20,width = 20,res = 150,units = \"cm\") heatmap(t(m5),Rowv = NA,Colv = NA,na.rm = T, scale = \"none\",col = terrain.colors(10),     xlab = \"\", ylab = \"\",main = \"Incidence of Measles in the US\") dev.off()                           Heatmap using the base function heatmap()                That’s pretty much what we can do with that. The enchanced heatmap.2() function from gplots() can do more. The legend for the colours is beyond terrible, so we will use the function gradient.rect() from plotrix package to create our own legend. And after fiddling around with numerous arguments and countless trial and errors later, we have as below.   library(gplots) #heatmap.2 library(plotrix) #gradient.rect() for legend  png(filename = \"measles-gplot.png\",height = 18,width = 30,res = 150,units = \"cm\",type=\"cairo\") #heatmap heatmap.2(t(m5),na.rm = T,dendrogram = \"none\",Rowv = NULL,Colv = \"Rowv\", trace = \"none\",scale = \"none\",offsetRow = 0.3, offsetCol = 0.3, breaks = c(-1,0,1,10,100,500,1000,max(m3$Incidence,na.rm = T)), #a separator line every 10 years colsep = which(seq(1928,2003) %% 10 == 0),margin = c(3,8), col = rev(c(\"#d53e4f\",\"#f46d43\",\"#fdae61\",\"#fee08b\",\"#e6f598\",\"#abdda4\",\"#ddf1da\")), xlab = \"\", ylab = \"\",key = F,lhei = c(0.1,0.9),lwid = c(0.2,0.8)) #legend gradient.rect(0.125,0.25,0.135,0.75,nslices = 7,border = F,gradient = \"y\", col = rev(c(\"#d53e4f\",\"#f46d43\",\"#fdae61\",\"#fee08b\",\"#e6f598\",\"#abdda4\",\"#ddf1da\"))) #legend text text(x = rep(0.118,7), y = seq(0.28,0.72,by = 0.07),adj = 1,cex = 0.8, labels = c(\"0\",\"0-1\",\"1-10\",\"10-100\",\"100-500\",\"500-1000\",\"&gt;1000\")) #title title(main = \"Incidence of Measles in the US\",line = 1,oma=T,adj=0.21) dev.off()                           Heatmap created using gplots function heatmap.2()                Ultimately, we get something useful. The white separator line is a neat feature. It can be used to group columns or rows as required. The while vertical lines above group decades. And below is the full undisrupted code.   # 2015 | Roy Mathew Francis # Heatmap R code  #install packages install.packages(pkgs = c(\"ggplot2\",\"reshape2\",\"plyr\",         \"Cairo\",\"gplots\",\"plotrix\"),dependencies = T)  #load packages library(ggplot2) #ggplot() for plotting library(reshape2) #melt(), dcast() for data reformatting library(plyr) #ddply() for data reformatting library(Cairo) #better aliasing of output images  #----------------------------------------------------------------------------  #DATA PREPARATION  #read csv file m &lt;- read.csv(\"measles_lev1.csv\", header = T,stringsAsFactors = F,skip = 2)  #convert data to long format m2 &lt;- melt(m,id.vars = c(\"YEAR\", \"WEEK\")) #rename column names colnames(m2) &lt;- c(\"Year\", \"Week\", \"State\", \"Value\")  #custom function to convert to camel case camelCase &lt;- function(string = NULL,separator = \"\\.\") {     if(is.null(string)) stop(\"No input string.n\")     s &lt;- strsplit(string, separator)     s &lt;- tolower(s[[1]])     paste(toupper(substring(s, 1, 1)), substring(s, 2),sep = \"\", collapse = \" \") }  #change variable to character, convert to camel case, #remove dot and change variable back to factor m2$State &lt;- factor(as.character(sapply(as.character(m2$State),camelCase)))  #change variable types m2$Year &lt;- factor(m2$Year) m2$Week &lt;- factor(m2$Week) #also converts '-' to NA m2$Value &lt;- as.numeric(m2$Value)  #custom sum function returns NA when all values in set are NA, #in a set mixed with NAs, NAs are removed and remaining summed. naSum &lt;- function(x) {     if(all(is.na(x))) val &lt;- sum(x,na.rm = F)     if(!all(is.na(x))) val &lt;- sum(x,na.rm = T)     return(val) }  #sums incidences for all weeks into one year m3 &lt;- ddply(m2,c(\"Year\",\"State\"),Incidence = round(naSum(Value),0),summarise)  #----------------------------------------------------------------------------  #GGPLOT2  #basic ggplot p &lt;- ggplot(m3,aes(x = Year,y = State,fill = Incidence))+         geom_tile()  #save plot to working directory ggsave(filename = \"measles-basic.png\",plot = p)  #modified ggplot p &lt;- ggplot(m3,aes(x = Year,y = State,fill = Incidence))+     geom_tile(colour = \"white\",size = 0.25)+     xlab(\"\")+     ylab(\"\")+     scale_y_discrete(expand = c(0,0))+     scale_x_discrete(expand = c(0,0),breaks = c(\"1930\",\"1940\",\"1950\",\"1960\",\"1970\",\"1980\",\"1990\",\"2000\"))+     coord_fixed()+     theme_grey(base_size = 8)+     theme(legend.position = \"right\",         axis.text = element_text(face = \"bold\"),         axis.ticks = element_line(size = 0.4),         plot.background = element_blank(),         panel.border = element_blank())  ggsave(filename = \"measles-mod1.png\",plot = p,dpi = 150,type = \"cairo\")  #reverse level order of state m3$State &lt;- factor(as.character(m3$State),levels = rev(levels(m3$State))) #create a new variable from incidence m3$IncidenceFactor &lt;- cut(m3$Incidence,                 breaks = c(-1,0,1,10,100,500,1000,max(m3$Incidence,na.rm = T)),                 labels = c(\"0\",\"0-1\",\"1-10\",\"10-100\",\"100-500\",\"500-1000\",\"&gt;1000\")) #change level order m3$IncidenceFactor &lt;- factor(as.character(m3$IncidenceFactor), levels = rev(levels(m3$IncidenceFactor)))  #assign text colour textcol &lt;- \"grey40\"  #further modified ggplot p &lt;- ggplot(m3,aes(x = Year,y = State,fill = IncidenceFactor))+     geom_tile()+     geom_tile(colour = \"white\",size = 0.25, show_guide = FALSE)+     labs(x = \"\",y = \"\",title = \"Incidence of Measles in the US\")+     scale_y_discrete(expand = c(0,0))+     scale_x_discrete(expand = c(0,0),breaks = c(\"1930\",\"1940\",\"1950\",\"1960\",\"1970\",\"1980\",\"1990\",\"2000\"))+     #library(RColorBrewer)     #scale_fill_manual(values = rev(brewer.pal(7,\"YlGnBu\")),na.value = \"grey90\")+     scale_fill_manual(values = c(\"#d53e4f\",\"#f46d43\",\"#fdae61\",\"#fee08b\",\"#e6f598\",\"#abdda4\",\"#ddf1da\"),na.value = \"grey90\")+     coord_fixed()+     theme_grey(base_size = 10)+     theme(legend.position = \"right\",legend.direction = \"vertical\",         legend.title = element_blank(),         legend.margin = grid::unit(0,\"cm\"),         legend.text = element_text(colour = textcol,size = 7,face = \"bold\"),         legend.key.height = grid::unit(0.8,\"cm\"),         legend.key.width = grid::unit(0.2,\"cm\"),         axis.text.x = element_text(size = 10,colour = textcol),         axis.text.y = element_text(vjust  =  0.2,colour = textcol),         axis.ticks = element_line(size = 0.4),         plot.background = element_blank(),         panel.border = element_blank(),         plot.title = element_text(colour = textcol,hjust = 0,size = 14,face = \"bold\"))  #export figure ggsave(filename = \"measles-mod3.png\",plot = p,dpi = 150,type = \"cairo\")  #----------------------------------------------------------------------------  #BASE GRAPHICS  #load package library(gplots) library(plotrix)  #convert from long format to wide format m4 &lt;- dcast(m3,Year~State,value.var = \"Incidence\") m5 &lt;- as.matrix(m4[,-1]) rownames(m5) &lt;- m4$Year  #base heatmap png(filename = \"measles-base.png\",height = 20,width = 20,res = 150,units = \"cm\") heatmap(t(m5),Rowv = NA,Colv = NA,na.rm = T, scale = \"none\",col = terrain.colors(100),         xlab = \"\", ylab =  \"\",main = \"Incidence of Measles in the US\") dev.off()  #gplots heatmap.2 png(filename = \"measles-gplot.png\",height = 18,width = 30,res = 150,units = \"cm\",type=\"cairo\") gplots::heatmap.2(t(m5),na.rm = T,dendrogram = \"none\",Rowv = NULL,Colv = \"Rowv\",trace = \"none\",scale = \"none\",offsetRow = 0.3, offsetCol = 0.3,                     breaks = c(-1,0,1,10,100,500,1000,max(m3$Incidence,na.rm = T)),colsep = which(seq(1928,2003) %% 10 == 0),                     margin = c(3,8),col = rev(c(\"#d53e4f\",\"#f46d43\",\"#fdae61\",\"#fee08b\",\"#e6f598\",\"#abdda4\",\"#ddf1da\")),                     xlab = \"\", ylab =  \"\",key = F,lhei = c(0.1,0.9),lwid = c(0.2,0.8)) gradient.rect(0.125,0.25,0.135,0.75,nslices = 7,border = F,gradient = \"y\",col = rev(c(\"#d53e4f\",\"#f46d43\",\"#fdae61\",\"#fee08b\",\"#e6f598\",\"#abdda4\",\"#ddf1da\"))) text(x = rep(0.118,7), y = seq(0.28,0.72,by = 0.07),adj = 1,cex = 0.8,labels = c(\"0\",\"0-1\",\"1-10\",\"10-100\",\"100-500\",\"500-1000\",\"&gt;1000\")) title(main = \"Incidence of Measles in the US\",line = 1,oma=T,adj=0.21) dev.off()  #------------------------------------------------------------------------------- #End of script   Conclusion   We have covered data prep and plotting heatmaps in R using base graphics and ggplot2. ggplot2 seems more consistent with code structure but the base graphics may be useful when combing multiple graphics in complicated ways. Hacking ggplot2 can be harder than fiddling with base graphics. I have not covered heatmaps with dendrograms because they are only useful in specific situations. I refer you to other articles such as this, this and this for more information.  ","categories": ["Code","Research"],
        "tags": ["Data Visualisation","Heatmap","R","Scientific Graphs"],
        "url": "/a-guide-to-elegant-tiled-heatmaps-in-r/",
        "teaser": null
      },{
        "title": "Björn the Norwegian forest cat",
        "excerpt":"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               This is Björn. He is a 6 month old Norwegian Forest Cat. Björn is energetic and playful. He is very fluffy and cuddly. He likes human company. He is one those strange cats that like belly rubs. His sleeping spot is the hammock on the play tower. But he can also be naughty around the house. Sometimes, he does go hyper and runs around the house. He attacks toilet paper rolls like there is no tomorrow. Sometimes he can be stupid enough to fall out through the window. Other times, he can be sharp enough to spot a fly on the window from meters away. He likes to help with the gardening. He also likes to help with things when not needed. But, he is generally good company to have around the house.  ","categories": ["Photography"],
        "tags": ["Cat","Forest"],
        "url": "/meet-bjorn-the-cat/",
        "teaser": null
      },{
        "title": "Structure 'Sort by Q' explained",
        "excerpt":"I currently use STRUCTURE 2.3.4 on Windows. A typical assignment output file for K=2 looks like below. Here is the structure file I use.                           Fig.1: A typical Structure assignment output.                We can use the plotting functionality within STRUCTURE to view the assignment results as a barplot. See fig below. The individuals are ordered in the same order as the input file when selecting the ‘Original order’ option. There is another option to sort individuals called ‘Sort by Q’. What does this actually do?                           Fig.2: Barplot in STRUCTURE software showing original order of individuals (top) and ‘Sort by Q’ order (bottom).                One might reasonably assume that the individuals are sorted by one of the assignment clusters. But, that is not the case. We will try to plot the data manually and investigate this option. The structure output file used can be downloaded here. This is the same as that linked earlier.   We use the R package pophelper to convert structure files to R dataframe, ggplot package for plotting and reshape2 package for data restructuring. The data is read into R as a dataframe with two columns Cluster1 and Cluster2 with assignment values.   #install pophelper library #library(devtools) #install_github('royfrancis/pophelper')  #load packages library(ggplot2) library(reshape2) library(pophelper)  #read data to dataframe df &lt;- readQ(\"structure-file.txt\")[[1]] head(df)   &gt; head(df) Cluster1 Cluster2 1 0.965 0.035 2 0.977 0.023 3 0.961 0.039 4 0.975 0.025 5 0.974 0.026 6 0.982 0.018   Now we create a function to create the plot.   #create function to generate plots plotfn &lt;- function(df=NULL, filename=NULL){ #reshape to long format df$num &lt;- 1:nrow(df) df1 &lt;- reshape2::melt(df, id.vars=\"num\") #reversing order for cosmetic reasons df1 &lt;- df1[rev(1:nrow(df1)), ]  #plot p &lt;- ggplot(df1, aes(x=num, y=value, fill=variable))+ geom_bar(stat=\"identity\", position=\"fill\", width=1, space=0)+ scale_x_continuous(expand = c(0, 0))+ scale_y_continuous(expand = c(0, 0))+ labs(x = NULL, y = NULL)+ theme_grey(base_size=7)+ theme(legend.position = \"none\", axis.ticks = element_blank(), axis.text.x = element_blank())  ggsave(filename=filename, plot=p, height=4, width=12, dpi=150, units=\"cm\") }  #plot unsorted plot plotfn(df=df, filename=\"bp-r-unsorted.png\")   Here is the assignment barplot in the original order.                           Fig.3: Assignment barplot recreated in R. Individuals are in original order.                Now we create two plots. One figure where the table is sorted by Cluster1 and second figure where the table is sorted by Cluster2.   #sort table by cluster1 df_c1 &lt;- df[order(df[, 1]), ] plotfn(df=df_c1, filename=\"bp-r-sortedc1.png\")  #sort table by cluster2 df_c2 &lt;- df[order(df[, 2]), ] plotfn(df=df_c2, filename=\"bp-r-sortedc2.png\")                           Fig.4: Assignment barplot in R sorted by cluster 1.                                        Fig.5: Assignment barplot in R sorted by cluster 2.                Both of these plots do not resemble the ‘Sort by Q’ option in Structure software. They look like mirror images only because it’s K=2. For K&gt;2, they would look quite different. Anyway, the ‘Sort by Q’ option does a bit more. For each individual, the max assignment value is picked to create a new column called ‘max’. The cluster number with the max assignment is created as a new column called ‘match’. The the whole table is sorted ascending by ‘match’ and descending by ‘max’. Here is the R code.   #pick max cluster, match max to cluster maxval &lt;- apply(df, 1, max) matchval &lt;- vector(length=nrow(df)) for(j in 1:nrow(df)) matchval[j] &lt;- match(maxval[j], df[j, ])  #add max and match to df df_q &lt;- df df_q$maxval &lt;- maxval df_q$matchval &lt;- matchval  #order dataframe ascending match and decending max df_q &lt;- df_q[with(df_q,  order(matchval, -maxval)), ]  #remove max and match df_q$maxval &lt;- NULL df_q$matchval &lt;- NULL  #plot plotfn(df=df_q, filename=\"bp-r-sortedq.png\")   And that gives us the plot we are looking for. The same plot created in the STRUCTURE software.                           Fig.6: Assignment barplot in R sorted by Q.                Here is also an Excel file with the calculations, if R is not your thing.                           Fig.7: Assignment barplot and ‘Sort by Q’ calculation in Excel.                You can always verify by checking the individual number (#) with the individual numbers in the STRUCTURE software (set to ‘Plot in multiple lines’).   That’s all for now. I hope this was useful for all those who were as confused as I was.  ","categories": ["Code","Research"],
        "tags": ["Population Genetics","R","Scientific Graphs"],
        "url": "/structure-sort-by-q-explained/",
        "teaser": null
      },{
        "title": "Polar adventures through Swedish Lapland",
        "excerpt":"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 I finally had the opportunity to go to Swedish Lapland this year. Lapland feels completely different between summer and winter. Winter is special for many reasons, firstly my fondness of snow (and loads of it), northern lights (oh yeah), snow animals (reindeer, huskies), snow sports (ski, snowboard) and lastly the alpine landscape.   A trip in the middle of February into the harsh unforgiving and unpredictable weather of  Swedish Lapland calls for some careful preparation and planning. This means subzero temperatures and flash snow-storms. This website has some useful tips and tricks for travelling in Sweden. Of course, we were prepared with extra heater, warm blankets, sleeping bags, fire-making equipment, snow shovels, reflectors etc. Perhaps the most surprising discovery was the size of Sweden. It is hard to imagine the size of Sweden on a Mercator projection map. It turns out that the distance from Copenhagen, Denmark to Narvik, Norway (2064 km) is the same distance as Copenhagen to Naples, Italy (2102 km). Yes, it’s 22 hours of driving one-way.   Pretty much all of Sweden from the Skåne up to Lapland is quite flat and featureless, except for the whole western border with Norway which is marked by the Scandinavian mountain range. The natural vegetation changes dramatically as we move from the South to North. The vegetation changes from mixed deciduous hardwood forests in the southern tip to boreal (taiga) forests in central Sweden to alpine vegetation in the very North. Personally, for me, the most exciting winter landscapes were the dense coniferous forests of Västerbotten.   Winter time in Northern Sweden is exciting for several activities which may not be possible elsewhere. Some of these include ice-fishing, cross-country skiing, snow-trekking, husky sledging, snowmobile expeditions, aurora-hunting and other winter sports. I was particularly interested in watching auroras and trying out snowmobiling. Of course, one wants to try out all activities if possible when there but these things are not cheap by any standards.   Jokkmokk lies on the Arctic circle and is sort of the gateway to Lapland and the Sami culture. I was lucky to to see northern lights for the first time in Jokkmokk. I kept checking on my mobile app constantly for high kp-index and clear skies. Driving out of the town centre into a dark remote lake side provided the best views. Auroras are not always easily visible to the naked eyes. A long-exposure shot in camera followed by some photo-editing brings out the strange shapes and colours. Luckily, I was able to see some bright intense auroras due to high kp-index (&gt;7). They usually tend to be green although I think I might’ve seen some reds too. My attempt at filming the aurora did not work as it was not bright enough to register at 30fps.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Auroras light up the night sky.          Kiruna is the northernmost major town in Sweden located about 145 km north of Jokkmokk/Arctic circle. Kiruna is essentially a mining town powered by the LKAB Iron-ore mine. Kiruna is also a popular tourist destination with winter activities and the famous Ice hotel. Kiruna also serves as an entry to the Abisko National Park. Kiruna also has a nice big tourist information centre with brochures and guides to the various attractions. There are numerous tours being organised here including trips to Mt. Kebnekaise (the highest mountain in Sweden) by snowmobiles, husky sledging or winter trekking. We took the basic guided snowmobile tour through the local wilderness. The snowmobiles are surprisingly easy to learn. The tour lasting about 3.5 hours took us through some enchanting snow covered woodlands and hills. We were also offered some grilled moose meat sausage.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         Views of the Ice hotel.          The Ice hotel is located at Jukkasjärvi, about 17 km east of Kiruna. The Ice hotel is constructed out of ice every year. The ice hotel comes with full hotel experience including the ice bar, ice chapel and ice bedrooms. The prices for the rooms are pretty insane but each room is a unique work of art. It was fun to look around but other than that there is not much to do. There is a Sami museum close to the Ice hotel but was closed when we got there.   Beyond Kiruna, the only two points of interest within Sweden are the Abisko national park and the Riksgränsen-Björkliden ski slopes. I decided to have a go at snowboarding for the first time. It is pretty hard to get started. The most difficult part was going uphill using the J-lift wearing the snowboard. After several failed attempts, I decided to do my learning at the bottom of the slope. I must’ve spent more time kissing the snow than on the board. After a painful 2 hours, I am soaking in sweat as I was wearing a down jacket which is, as I learned, totally inappropriate for active sports. All this sweat and -10°C doesn’t go well. I was pretty frozen by the time we decided to get off the slopes. We were in for another surprise when we learnt that the only fuel station after Kiruna was in Narvik, Norway. We had to take a detour from Riksgränsen all the way to Narvik. The route to Narvik was pretty fun with all the mountainous, spiralling roads which is quite typical for Norway.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            Considering wildlife, we were able to see reindeers, huskies and siberian jays. On the way back, we were lucky to spot some snowy grouse by the roadside. Nothing eventful on the way back except that we were caught in a blizzard for a while. I am looking forward to going north in the summer time.  ","categories": ["Photography","Travel"],
        "tags": ["Lapland","National Park","Nature","Road Trip","Sweden","Wildlife"],
        "url": "/swedish-lapland-trip/",
        "teaser": null
      },{
        "title": "City trip to Leeds, UK",
        "excerpt":"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 I spent the last couple of days in Leeds, UK. The weather was just awful. It was around 2-6 °C which was good and bad. Good in the sense that it was not too cold and therefore comfortable to go around. Bad in the sense that all the precipitation was rain and no snow.   Leeds is not the most interesting city in the UK. It’s mostly run by all the universities. Loads of students. It’s quite expensive and budget accommodation is lacking. Perhaps the best part of Leeds is the city center with its countless shops and shopping malls. The whole place is overrun by shoppers. You get a whole lot of shopping options from expensive designer brands to cheaper TK-Maxx and Primark. There are not many other attractions in Leeds except for art galleries, cathedrals and state houses which are completely uninteresting for me. The only two attractions that I thought was interesting is the Royal Armouries museum in the city and the National Coal Mine museum near Wakefield.   The NCM (National Coal Mine museum) is pretty cool. The entry is essentially free and you get to go on a free tour taking you 140m down into a defunct coal mine. The tour guide gives you some detailed insight into the experience of working in a coal mine throughout the ages. The tour guides are ex-miners and it can be a challenge to understand what they are saying through their heavy Yorkshire accent. Other than that, they are humorous and friendly chaps.                           The university of Leeds.                ","categories": ["Photography","Travel"],
        "tags": ["UK"],
        "url": "/leeds-uk/",
        "teaser": null
      },{
        "title": "The FortaCloud fraud",
        "excerpt":"Cloud computing is fun, flexible and relatively cheap. If you are a hobbyist like me, who likes to fiddle around with web apps, experiment with web hosting and such, standard established cloud services like Amazon AWS, Linode, Digital Ocean etc might a bit too pricey. So, in the search for cheaper alternatives, one comes across options such as Vultr, CloudSigma or FortaCloud. It is a gamble when one has to authenticate using credit cards and sometimes it can go wrong.   I came across FortaCloud in December 2015. They claim to provide cloud computing services and had a free-tier program. I thought it might a good idea to try it out and see how it goes. I created an account and signed up for their so called ‘Free Tier FortaCloud VPS’ on the 18-December-2015. There was an initial setup fee of 0.82 USD. I had to use the credit card to pay for this. I thought that was alright and perhaps it was a security measure to avoid misuse. Below is the email confirmation I received.                           Email of registration                Once my service was active, I set up a basic virtual server which was up and running. Below is the email I received once my server was running.                           Email of server creation                At this point, I was interesting in monitoring the performance and stability of the server. I set up a service called SeaLion. With SeaLion, I could monitor my server from the SeaLion dashboard, rather than logging on to my FortaCloud account or even SSH into my FortaCloud server. So far so good. Things were moving smoothly.   And then 3 days later, the FortaCloud server stopped responding to SeaLion. I could not SSH into the server or access the server in anyway. It seems like my server crashed for whatever reason. I tried to log into my account and that was not possible either. I tried to reset my password and it said my account did not exist and my email was not even registered. I tried to get support. That is when I realised the website does not have any support or proper contact details. A question ticket is charged 5 USD per question. Even after submitting the question, and paying for it, they do not respond. Then I came across their facebook page and found a post about the disruption in service. So, I posted a comment on their facebook page to which I received no response.                           Facebook                I tried for a few days and then gave up on it. One month passed by (Jan-2016), and in the middle of Feb-2016, I just happened to be going through my bank transactions and I noticed that my card was charged once in Jan and once in Feb by FortaCloud.   01-18 01-19 FORTACLOUD.CO xxxxxxxxxx USD 101,39 02-09 02-10 FORTACLOUD.CO xxxxxxxxxx USD 124,03   I have been charged 225 USD for a free service that did not work! I was shocked by this unauthorized transaction. I tried to log into my account again. I reset my password and tried again and it finally worked. Once I got into my account, I could see that the ‘Free Cloud VPS’ was active and I didn’t even know about it. So I tried to close it down and of course there is no option to close it or delete it. Digging around for a bit, I found a button to request cancellation. I submitted a cancellation request and now a week later, it is still active. The free tier server running is seen below.                           Dashboard                Scenario 1: It is a free service, and I should not be charged at all, except for the setup fee of 0.82 USD. Scenario 2: Let’s imagine ‘free’ was not really free, then they should be charging me 9.95 USD as shown above. I got charged 225 USD which doesn’t make any sense whichever scenario.   I then tried to remove my credit card details from the website, and as you guessed, it is not possible. I tried to close my account, and that is not possible either. And there is no one to contact. I finally figured that although I cannot remove my card details, I could change it. So I immediately changed the card details to a random card number found on the internet. And of course I blocked my credit card to avoid further use.   I started looking around about this company on their facebook page.                           Facebook page                To my horror, I came across numerous complaints from users which I hadn’t paid attention to before. It seems like I am not the first person to have this problem. Many other users had similar experience as me.                           Facebook complaints                                        Facebook complaints                Reading around a bit, there are many reports of the fraudulent activities of FortaCloud. Some examples are articles such as this, this and this. The credit card transactions are clearly fraudulent and this whole company is a scam.   I am currently working up a case with my credit card company to reclaim the transactions. Many others are going through the same process. I hope this has been helpful and should help you avoid being caught up in this mess.   Since I had mentioned two other services Vultr and CloudSigma in the beginning of this post, I need to clarify those. I have been using Vultr for the past several months. The service has been quite stable and reliable. CloudSigma is still in experimental phase and the service is not entirely stable. But poor service is one thing and downright fraud is another.  ","categories": ["Web"],
        "tags": ["Cloud Computing","Web Hosting"],
        "url": "/fortacloud-fraud/",
        "teaser": null
      },{
        "title": "Curious about a Bee Swarm",
        "excerpt":"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         A few years back, my advisor and myself came across a honey bee swarm on the young apple trees in the backyard. It was a nice bright summer day and the flowers were in full bloom. We decided to get some empty hives and give them a home before the swarm went around and ruined people’s day. My contribution was very little in the this whole activity. I had a camera and so I went around shooting videos and taking pictures. As a little Easter project, I decided to put together the footage lying around from that day. Happy Easter!    ","categories": ["Photography"],
        "tags": ["Bee"],
        "url": "/curious-about-a-bee-swarm/",
        "teaser": null
      },{
        "title": "Which file compression to use on Linux?",
        "excerpt":"Seven different compression formats (7z, bzip2, gzip, lrzip, lz4, xz and zip) are tested using ten different compression commands (7za, bzip2, lbzip2, lrzip, lz4, pbzip2, gzip, pigz, xz and zip) on five different file types (fastq, mp3 tar archive, mp4 movie file, random text file and a tiff stack) for compression ratio and time. bzip2 compression using the command lbzip2 and pbzip2 comes out as the winner due to high compression ratio, speed and multi-threading capabilities.     Five different data files were tested: a fastq text file, mp3 tar archive, an mp4 movie file, a randomly generated text file and a tiff image stack. Some properties of the files: fastq file (403 MB, 1.56 million reads), mp3 tar archive (390 MB, a tar archive composed of four tar archives each with 6 mp3 tracks of size 10MB to 32MB), mp4 file (340 MB), text file (400MB, created using (base64 /dev/urandom | head -c 419430400 &gt; text.txt) and tiff stack (404MB, 1380 frames, 640 x 480 px, sequence of zebrafish larvae swimming in a microtitre plate).  For clarity, fastq files are text files containing next generation sequencing data and tiff stacks are used for image analysis using ImageJ, for example.   Seven different compression formats were tested: 7z, bzip2, gzip, lrzip, lz4, xz and zip using ten different compression commands: 7za, bzip2, lbzip2, pbzip2, gzip, pigz, lrzip, xz and zip. For decompression, the same commands were used except for zip where unzip was used. The 7za command by default compresses to the 7z format but also allows exporting to bzip2, gzip and zip. lbzip2 and pbzip2 are multi-threaded versions of bzip2. Similarly, pigz is the multi-threaded version of gzip.   All commands were run on a Scientific Linux cluster with 8 cores and a total of 24GB of RAM. All commands used the default settings/switches. Most of these commands allow the user to set the level of compression but this was not changed. By running on 8 cores, the methods were free to utilize the 8 available cores if they could.                           Total time used by each format/command for each file type.                The x-axis shows the different file types. The y-axis shows the total time taken (in seconds) for compression plus decompression. The colour of the bar denotes format/command. The shorter the bars, the faster the command is. lz4 seems to be the fastest for all file types. lbzip2, pbzip2 and pigz are also extremely fast for all file types taking around 15 secs or so. 7z and xz seems to be pretty slow taking more than 2 mins for all file types.                           File size comparison between original file and compressed file for each format/command for each file type. A smaller value means higher compression.                The x-axis shows the different file types. The y-axis shows the percentage size difference between the original file and the compressed file. The colour of the bar denotes format/command. The shorter the bars, the smaller the compressed file is. The compression ratio is drastically different depending on the file type. fastq files show the highest compression with 7z, lrz and xz showing the best compression. For mp4 files, no compression was observed with any of the formats. The tiff stack had varying levels of compression with different formats but generally show high compression. Even for the same file types, compressed file size greatly varies with the content within the file.                           Scatterplot showing scaled times and scaled file size difference for each file type.                The x-axis shows scaled file size difference and the y-axis shows scaled time. Separate sub-plots show different file types. The different compression formats are shown in ten different colours. On both axes, smaller value is better. A format with shortest time and smallest compressed file size is desirable. The plot quadrant/area is coloured to indicate best, good, intermediate and bad formats. R function rescale() from scales package was used to rescale time and file size difference to values between 0 and 1 across all file types.                           Summary score comparison for each format/command for each file type. Score is an arbitrary value calculated from scaled time and scaled compression size difference. A larger value is better.                A summary score was calculated from the scaled total time and scaled file size difference. 1 was subtracted from all values to invert the values. Now a larger value is better than a smaller value. And then the inverted scaled value for each format was added to produce the summary score. The x-axis shows the different file types. The y-axis shows the summary score. The score takes into account speed and well as level of file compaction and is comparable across file types. A larger score is better.   Multi-threaded bzip2 commands such as lbzip2, pbzip2 have the best scores for all file types especially for text files. Parallel gzip command pigz also has high score. lrz shows a curiously high score for mp3 file. The mp4 movie file does not seem to reduce in size at all with any compression format, therefore, it might be pointless to compress them in the first place. 7z has generally low scores as they it is much slower than others. For tiff-stacks, 7z has a pretty high score due to the extremely high file size reduction.   The overall winner in this particular test would be bzip2 compression using the command lbzip2 or pbzip2 as it is extremely fast, multi-threaded, has high compression ratio and allows random access. A multi-threaded 7z tool might be a game changer. 7z tool for windows (7zip) supports multi-threading on windows. A multi-threaded version of xz called pixz is available but not tested here. bzip2 also has a parallel version specifically for clusters called the mpibzip2.   The versions of softwares used here were: 7za (9.20), bzip2 (1.0.5), gzip (1.3.12), lbzip2 (2.5), lrzip (0.621), lz4 (r131), pbzip2 (1.1.12), pigz (2.3.3), xz (4.999.9beta) and zip (3.0). The basic bash commands used to compress and decompress the files are shown below.   #compression 7za a -y -t7z file.7z file; bzip2 -qk file; gzip -cq file &gt; file.gz; lbzip2 -qkz file; lrzip -q file -o file.lrz; lz4 file file.lz4; pbzip2 -qkz file; pigz -kq file; xz -zkq file; zip -q file.zip file;  #decompression 7za x -y file.7z; bzip2 -dk file.bz2; gzip -qdc file.gz &gt; file; lbzip2 -dkq file.bz2; lrzip -dq file.lrz; lz4 -d file.lz4; pbzip2 -dk file.bz2; pigz -dk file.gz; xz -dk file.xz; unzip file.zip;   Of course this is by no means the bottom line. There are lot more aspects to dig into such as ability to compress streams, encryption, the underlying algorithm (LZMA etc), indexability, memory usage etc. The whole comparison can be repeated by changing the compression level switches within these commands which might change the outcome. Memory consumption was not measured or monitored in this test. The way a function is parallelized can be different. For instance, one file can be split to be run on multiple cores simultaneously (case 1) or if multiple files are provided, each core works on one single file in parallel (case 2). Given that information, if only one file is provided to a parallel function programmed as case 2, only one core is utilized and there is no added benefit from multiple cores. And of course, there are numerous other compression algorithms, formats and commands out there.  ","categories": ["Research"],
        "tags": ["Bash","Image Analysis","NGS"],
        "url": "/file-compression-use-linux/",
        "teaser": null
      },{
        "title": "Wild road trip through Scotland",
        "excerpt":"This spring, I decided to explore Scotland in a one-week trip. April-May is still a bit too early in the season, but I was able to experience all sorts of weather from sun to rain to snow and wind before the tourist season.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            Here is a route map:       Scotland is just stunningly beautiful. It has possibly some of the most spectacular landscapes in the world. Skye is just so dramatic in landscapes and weather patterns. I experienced snow and warm sunny weather on the same day in different parts of the island. We were able to visit the Talisker distillery, Fairy pools and Kilt rock, but missed Old man of Storr and many other attractions. I had limited time on Skye, but Skye has so much to offer. One could spend an entire holiday just exploring Skye.   Driving is a pleasure in Scotland due to the much reduced traffic at least compared to England. On mainland Scotland, major roads and highways are large and wide and well maintained. Smaller roads may be narrow with two-way traffic and not so well maintained. On Skye, all roads are narrow. Driving on Skye really tests your ability to keep the vehicle perfectly in a lane that is just as wide as the vehicle while driving 50-60 mph (80-90 kph). If driving from Skye towards Inverness and you have extra time, I can recommend the Church Road route offering some stunning views.   The only location I travelled north of Inverness was the Dunrobin castle in Golspie. It was in good time for the falconry session which was pretty exciting.   As far as wildlife goes, we spotted occasional deers along the road and seals along the harbour in Inverness. We were not lucky enough to spot eagles, whales or puffins. There is a lot more to be discovered in Scotland and I hope to return at some point.  ","categories": ["Photography","Travel"],
        "tags": ["Forest","National Park","Nature","Road Trip","Scotland","Wildlife"],
        "url": "/scotland/",
        "teaser": null
      },{
        "title": "Read counts of RNA-Seq Spike-ins using STAR and QoRTs",
        "excerpt":"In RNA-Seq analyses, adding pre-determined quantity of synthetic RNA sequences (spike-ins) to samples is a popular way to verify the experimental pipeline, determine quantification accuracy and for normalisation of differential expression. The most commonly used spike-ins are the ERCC spike-ins.   This post will cover the bioinformatic steps involved in obtaining read counts of spike-ins from a FASTQ file sequenced with spike-ins. The steps are namely creating a custom FASTA genome build incorporating the spike-in sequences, custom GTF file creation, mapping the reads to the custom genome, read counting and visualisation. This post will not be covering the wet lab part of adding spike-ins. I have a FASTQ data file (sample01.fq.gz) from single cell 50bp single-end Illumina reads with spike-ins that I am using for this workflow.   Go to the manufacturer’s website and obtain the Spike-in sequences. Mine was obtained from Thermofisher. Download the sequence file.   wget \"https://tools.thermofisher.com/content/sfs/manuals/cms_095047.txt\"   We can take a look at the file.   column -s, -t &lt; cms_095047.txt | less -#2 -N -S 1 ERCC_ID GenBank 5prime_assay 3prime_assay Sequence 2 ERCC-00002 DQ459430 Ac03459872_a1 Ac03459968_a1 TCCAGATTACTTCCATTTCCGCCCAAGCTGCTCACAGTATACGGGCGTC 3 ERCC-00003 DQ516784 Ac03459873_a1 Ac03459969_a1 CAGCAGCGATTAAGGCAGAGGCGTTTGTATCTGCCATTATAAAGAAGTT 4 ERCC-00004 DQ516752 Ac03459874_a1 Ac03459970_a1 TCTTGCTTCAACAATAACGTCTCTTTCAGAAGGCATTGGTATCTTTTCC 5 ERCC-00009 DQ668364 Ac03459876_a1 Ac03459972_a1 CAATGATAGGCTAGTCTCGCGCAGTACATGGTAGTTCAGCCAATAGATG 6 ERCC-00012 DQ883670 Ac03459877_a1 Ac03459973_a1 CGAGAGATGTTTGTAGGTGCGGAATGTGTGCGGTCTACCTTAGCTGTAG 7 ERCC-00013 EF011062 Ac03459878_a1 Ac03459974_a1 ACCGAGCTCAGATGTGAAGGATCTTCTTGGAGGATTAAAAAAATGATTA 8 ERCC-00014 DQ875385 Ac03459879_a1 Ac03459975_a1 AAAAGAAGAAGGAAAGTAGAGGAGATCAAGATGTCAAACGAAACAATTA 9 ERCC-00016 DQ883664 Ac03459880_a1 Ac03459976_a1 GTTGTAACGAATGTTAATTTAGGAGGCAAGAGTTTGTGGGCGCGGACTT 10 ERCC-00017 DQ459420 Ac03459881_a1 Ac03459977_a1 GAGAACTGAAAGTGAGTCCCAACGAGAGAGGTGCATCTGTCCAGTGAGA 11 ERCC-00019 DQ883651 Ac03459883_a1 Ac03459979_a1 CTTTAATGGTTGTACATACTTGACGGATTGCAGTGAGTGTATTCCCGTC 12 ERCC-00022 DQ855004 Ac03459884_a1 Ac03459980_a1 CCCGGGCCAATTGCCTCTATAACTAGAGCTGAGCCCACCATTAAAGCGA 13 ERCC-00024 DQ854993 Ac03459886_a1 Ac03459982_a1 ACGGAGGAGCTTTGGCATACTAGGCTAGCGAATCTGCAACTAACGCAAG 14 ERCC-00025 DQ883689 Ac03459887_a1 Ac03459983_a1 GGGGTCCATTATAGTGCAGGCGTGGTAAAGTAGCATTAGAACCCTACTA 15 ERCC-00028 DQ459419 Ac03459888_a1 Ac03459984_a1 AAGAAAGTATTCCATTCCGGCTCATGGTCCCGGCTAGACCTGCAAGATC 16 ERCC-00031 DQ459431 Ac03459889_a1 Ac03459985_a1 GGGAGAAAAGCTTCTTGCCTAGTTACAGCACGAAGATTGGGACCCATCG   It is a tabular format with two columns that we need: ERCC ID and Sequence.   Genome file   The first step is to convert this tabular file into FASTA format and then append that to our genome FASTA file. We convert this table into a FASTA format file named ercc.fa.   cat cms_095047.txt | awk -F \"\\t\" -v OFS=\"\" 'NR&gt;1{print \"&gt;\",$1,\" \",$2,\" \",$3,\" \",$4,\"\\n\",$5}' &gt; ercc.fa   I am working with (Zebrafish) Danio rerio GRCz10. Let’s take a look at the Zebrafish FASTA genome file.   tail -n 20 Danio_rerio.GRCz10.dna.toplevel.fa &gt;CACACACACACACACACACACACACACACACACACATACACACTCACACAACGCAAACGG ACATACAGTACACAAACACACTCAAGACACCACAGACCAAACACACACACACTCCAAACA CAATAGACAATATTGACTATAATAGCCAATAATCTGTCCCACAATCCCAAACACCGACAG CACTGCTGGATTACAGCCAATAATGCAACAACTCTGGAAGGAAAAGCCTGCAGGTCAGAG GCCTTTAATAATGTAGATCACACTTAACCCGGTTAGAGAGGCTAATGGGGTCAAACAGGC CTGTGTGGCATTTTACATCACCAGGAAACCACTGGAGAACATTTAACACACACACACACA CACACACACACACAAGCAGATGTGCACATATACACATACATACATTATATATGTATACAT ACATACAT &gt;KN150525.1 dna:scaffold scaffold:GRCz10:KN150525.1:1:650:1 REF GTGTATTATAAAACATACGAGACTTAAAGTATAATCTATCCAGCACCTATAGAGCCACTC GCGTTGCTTACGAAATTAGTATTTAGAAAGACCCACCACAAAGAAAAGAGATGTGTCTGT ATGTATAAAAAATATATGAGACAGGGTTATAACCACCCGACGCGCTTTGTCTAAATAGAA GTCGAGCACACACACACACACACACACACACACACACACACACACACACGGGTCAGCCAG AGGACAGCTAACCTGCACACACCAGTCAAACCATGAACGGCAGAGAAACACCTCCATAAT CCACACTCACCCCTGACGAGCACACACACACACACACACATATACACACACACACACACA CTATAGTAAACTGGGTGTGCGTGGATGTGTGTATGTGTGTGATGGACTAGGATGTGTGTG TGTGTGTGTGTGTGTGTGTGTGTGTGTGTGTGAGCTCACGTTCACTGAAGACGAAGTCTG CGTTGATGTCGAAGGGCTGGATCTGGATGTCGTAGATCGACACGGTGACGCCTTTCTGAT GGTCGCTGGAGATTAGAGTCAGCGTCTGCTGCGCCGCACACACATAGGAGCGCCCGGCCG GAGTCACCAGCGCCGAGAGACGGTGTGTGCTGGCAGTGTGTTTCCCCGCT   We make a copy of the genome.   cp Danio_rerio.GRCz10.dna.toplevel.fa Danio_rerio.GRCz10.dna.toplevel.ercc.fa   And add the contents of ercc.fa file to the bottom of the copied Zebrafish genome file.   cat ercc.fa &gt;&gt; Danio_rerio.GRCz10.dna.toplevel.ercc.fa   Now, our new genome FASTA file looks like so   tail -n 5 Danio_rerio.GRCz10.dna.toplevel.ercc.fa &gt;ERCC-00170 DQ516773 Ac03459966_a1 Ac03460062_a1 TATTGGTGGAGGGGCACAAGTTGCTGAAGTTGCGAGAGGGGCGATAAGTGAGGCAGACAG GCATAATATAAGAGGGGAGAGAATTAGCGTAGATACTCTTCCAATAGTTGGTGAAGAAAA TTTATATGAGGCTGTTAAAGCTGTAGCAACTCTTCCACGAGTAGGAATTTTAGTTTTAGC TGGCTCTTTAATGGGAGGGAAGATAACTGAAGCAGTTAAAGAATTAAAGGAAAAGACTGG CATTCCCGTGATAAGCTTAAAGATGTTTGGCTCTGTTCCTAAGGTTGCTGATTTGGTTGT TGGAGACCCATTGCAGGCAGGGGTTTTAGCTGTTATGGCTATTGCTGAAACAGCAAAATT TGATATAAATAAGGTTAAAGGTAGGGTGCTATAAAGATAATTTAATAATTTTTGATGAAA CCGAAGCGTTAGCTTTGGGTTATGAAACTCCATGATTTTCATTTAATTTTTTCCTATTAA TTTTCTCCTAAAAAGTTTCTTTAACATAAATAAGGTTAAAGGGAGAGCTCTATGATTGTC TTCAAAAATACAAAGATTATTGATGTATATACTGGAGAGGTTGTTAAAGGAAATGTTGCA GTTGAGAGGGATAAAATATCCTTTGTGGATTTAAATGATGAAATTGATAAGATAATTGAA AAAATAAAGGAGGATGTTAAAGTTATTGACTTAAAAGGAAAATATTTATCTCCAACATTT ATAGATGGGCATATACATATAGAATCTTCCCATCTCATCCCATCAGAGTTTGAGAAATTT GTATTAAAAAGCGGAGTTAGCAAAGTAGTTATAGACCCGCATGAAATAGCAAATATTGCT GGAAAAGAAGGAATTTTGTTTATGTTGAATGATGCCAAAATTTTAGATGTCTATGTTATG CTTCCTTCCTGTGTTCCAGCTACAAACTTAGAAACAAGTGGAGCTGAGATTACAGCAGAG AATATTGAAGAACTCATTCTTTAGATAATGTCTTAGGTTAAAAAAAAAAAAAAAAAAAAA AAA &gt;ERCC-00171 DQ854994 Ac03459967_a1 Ac03460063_a1 CTGGAGATTGTCTCGTACGGTTAAGAGCCTCCGCCCGTCTCTGGGACTATGGACGGGCAC GCTCATATCAGGCTATATTTGGTCCGGGTTATTATCGTCGCGGTTACCGTAATACTTCAG ATCAGTTAAGTAGGGCCATATGCCTCGGGAATAAGCTGACGGTGACAAGGTTTCCCCCTA ATCGAGACGCTGCAATAACACAGGGGCATACAGTAACCAGGCAAGAGTTCAATCGCTTAG TTTCGTGGCGGGATTTGAGGAAAACTGCGACTGTTCTTTAACCAAACATCCGTGCGATTC GTGCCACTCGTAGACGGCATCTCACAGTCACTGAAGGCTATTAAAGAGTTAGCACCCACC ATTGGATGAAGCCCAGGATAAGTGACCCCCCCGGACCTTGGAGTTTCATGCTAATCAAAG AAGAGCTAATCCGACGTAAAGTTGCGGCGTTGATTACGCAGGATTGCGACCAAAGAACGA GAAAAAAAAAAAAAAAAAAAAAAAA   Annotation file   Now we define the annotation file. Our original Ensembl GTF file looks like this   tail -n 5 Danio_rerio.GRCz10.84.gtf &gt;KN150216.1 ensembl exon 353 443 . - . gene_id \"ENSDARG00000099551\"; gene_version \"1\"; transcript_id \"ENSDART00000165750\"; transcript_version \"1\"; exon_number \"3\"; gene_name \"tgds\"; gene_source \"ensembl\"; gene_biotype \"protein_coding\"; transcript_name \"tgds-201\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; exon_id \"ENSDARE00001154426\"; exon_version \"1\"; KN150216.1 ensembl CDS 353 443 . - 0 gene_id \"ENSDARG00000099551\"; gene_version \"1\"; transcript_id \"ENSDART00000165750\"; transcript_version \"1\"; exon_number \"3\"; gene_name \"tgds\"; gene_source \"ensembl\"; gene_biotype \"protein_coding\"; transcript_name \"tgds-201\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; protein_id \"ENSDARP00000138327\"; protein_version \"1\"; KN150216.1 ensembl exon 267 280 . - . gene_id \"ENSDARG00000099551\"; gene_version \"1\"; transcript_id \"ENSDART00000165750\"; transcript_version \"1\"; exon_number \"4\"; gene_name \"tgds\"; gene_source \"ensembl\"; gene_biotype \"protein_coding\"; transcript_name \"tgds-201\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; exon_id \"ENSDARE00001169286\"; exon_version \"1\"; KN150216.1 ensembl CDS 267 280 . - 2 gene_id \"ENSDARG00000099551\"; gene_version \"1\"; transcript_id \"ENSDART00000165750\"; transcript_version \"1\"; exon_number \"4\"; gene_name \"tgds\"; gene_source \"ensembl\"; gene_biotype \"protein_coding\"; transcript_name \"tgds-201\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; protein_id \"ENSDARP00000138327\"; protein_version \"1\"; KN150216.1 ensembl five_prime_utr 938 942 . - . gene_id \"ENSDARG00000099551\"; gene_version \"1\"; transcript_id \"ENSDART00000165750\"; transcript_version \"1\"; gene_name \"tgds\"; gene_source \"ensembl\"; gene_biotype \"protein_coding\"; transcript_name \"tgds-201\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\";   We will make a copy of the GTF file.   cp Danio_rerio.GRCz10.84.gtf Danio_rerio.GRCz10.84.ercc.gtf   We will create similar GTF file from the ERCC tabular file. Each ERCC sequence is defined as a new chromosome. First we define them as genes with gene_id staring with G followed by the ERCC ID and biotype ‘ercc’. This is written to a file ercc.gtf   cat cms_095047.txt | awk -F \"\\t\" -v OFS=\"\" 'NR&gt;1{print $1,\"\\t\",\"ercc\",\"\\t\",\"gene\",\"\\t\",\"1\",\"\\t\",length($5),\"\\t\",\".\",\"\\t\",\"+\",\"\\t\",\".\",\"\\t\",\"gene_id \\\"G\",$1,\"\\\"; \",\"gene_version \\\"1\\\"; \", \"gene_name \\\"\",$1,\"\\\"; \",\"gene_source \\\"ercc\\\"; \",\"gene_biotype \\\"ercc\\\";\"}' &gt; ercc.gtf   The ercc.gtf now looks like this:   head ercc.gtf &gt;ERCC-00002 ercc gene 1 1062 . + . gene_id \"ERCC-00002\"; gene_version \"1\"; gene_name \"ERCC-00002\"; gene_source \"ercc\"; gene_biotype \"ercc\"; ERCC-00003 ercc gene 1 1024 . + . gene_id \"ERCC-00003\"; gene_version \"1\"; gene_name \"ERCC-00003\"; gene_source \"ercc\"; gene_biotype \"ercc\"; ERCC-00004 ercc gene 1 524 . + . gene_id \"ERCC-00004\"; gene_version \"1\"; gene_name \"ERCC-00004\"; gene_source \"ercc\"; gene_biotype \"ercc\"; ERCC-00009 ercc gene 1 985 . + . gene_id \"ERCC-00009\"; gene_version \"1\"; gene_name \"ERCC-00009\"; gene_source \"ercc\"; gene_biotype \"ercc\"; ERCC-00012 ercc gene 1 995 . + . gene_id \"ERCC-00012\"; gene_version \"1\"; gene_name \"ERCC-00012\"; gene_source \"ercc\"; gene_biotype \"ercc\"; ERCC-00013 ercc gene 1 809 . + . gene_id \"ERCC-00013\"; gene_version \"1\"; gene_name \"ERCC-00013\"; gene_source \"ercc\"; gene_biotype \"ercc\"; ERCC-00014 ercc gene 1 1958 . + . gene_id \"ERCC-00014\"; gene_version \"1\"; gene_name \"ERCC-00014\"; gene_source \"ercc\"; gene_biotype \"ercc\"; ERCC-00016 ercc gene 1 845 . + . gene_id \"ERCC-00016\"; gene_version \"1\"; gene_name \"ERCC-00016\"; gene_source \"ercc\"; gene_biotype \"ercc\"; ERCC-00017 ercc gene 1 1137 . + . gene_id \"ERCC-00017\"; gene_version \"1\"; gene_name \"ERCC-00017\"; gene_source \"ercc\"; gene_biotype \"ercc\"; ERCC-00019 ercc gene 1 645 . + . gene_id \"ERCC-00019\"; gene_version \"1\"; gene_name \"ERCC-00019\"; gene_source \"ercc\"; gene_biotype \"ercc\";   The ercc.gtf is appended to the bottom of the copied Zebrafish GTF file.   cat ercc.gtf &gt;&gt; Danio_rerio.GRCz10.84.ercc.gtf   Mapping   We create a genome index as normal. I am using STAR aligner.   star \\   --runMode genomeGenerate \\   --runThreadN 8 \\   --genomeFastaFiles /home/ngs/raw/Danio_rerio.GRCz10.dna.toplevel.ercc.fa \\   --genomeDir /home/ngs/index/starindex-ercc/   And then we map our single-end or paired-end FASTQ files using the new ERCC genome index and ERCC GTF file. Here I have a single-end read gzipped FASTQ file.   star \\   --genomeDir \"/home/ngs/index/starindex-ercc/\" \\   --runThreadN 12 \\   --readFilesCommand zcat \\   --readFilesIn \"sample01.fq.gz\" \\   --outSAMtype BAM SortedByCoordinate \\   --outFileNamePrefix \"sample01-\" \\   --sjdbGTFfile \"/home/ngs/raw/Danio_rerio.GRCz10.84.ercc.gtf\"   This should produce a BAM file that can be used in the next step.   Quantification   The reads can be quantified (read counts) using any of the standard workflows. Reads can be quantified from the BAM files using HTSeq count, featureCounts etc. or the FASTQ files can be used directly with KALLISTO. Here I use QoRTs. QoRTs is a versatile tool for gene and transcript quantification as well as RNA-Seq QC. Chromosome-level counts, gene-level counts, transcript counts, biotype counts and numerous other results can be obtained.   QoRTs can be run as below for single-end reads. Note that --maxReadLength must be specified for single-end reads. I had to specify --maxPhredScore because my phred scores exceed the value of 41. --minMAPQ is used to specify that only uniquely mapped reads must be counted. Max mapping quality (unique mapping) for STAR is the value 200.   java -Xmx20G -jar /home/programs/qorts/qorts-1.1.2.jar QC \\   --singleEnded \\   --generatePlots \\   --generateMultiPlot \\   --noGzipOutput \\   --verbose \\   --maxReadLength 43 \\   --minMAPQ 60 \\   --maxPhredScore 100 \\   \"sample01.bam\" \\   \"/home/raw/Danio_rerio.GRCz10.84.ercc.gtf\" \\   \"sample01-qorts\"   I have also run featureCounts for comparison.   subread-1.5.0-p2-Linux-x86_64/bin/featureCounts \\   -a \"/home/raw/Danio_rerio.GRCz10.84.ercc.gtf\" \\   -o \"sample01-featurecounts-gene.txt\" \\   -F \"GTF\" \\   -t \"exon\" \\   -g \"gene_id\" \\   -M \\   -p \\   -C \\   -B \\   -S rf \\   -T 6 \\   \"sample01.bam\"   Here are some of gene counts (QC.geneCounts.txt) for Spike-ins from the QoRTs results. The second column contains the counts we are looking for.   GENEID COUNT COUNT_CDS COUNT_UTR COUNT_AMBIG_GENE GERCC-00002 2062 0 2062 0 GERCC-00003 24 0 24 0 GERCC-00004 411 0 411 0 GERCC-00009 0 0 0 0 GERCC-00012 0 0 0 0 GERCC-00013 0 0 0 0 GERCC-00014 0 0 0 0 GERCC-00016 0 0 0 0   And here are some of the gene counts from featureCounts. The last column contains the counts we need.   Geneid Chr Start End Strand Length sample01.bam GERCC-00002 ERCC-00002 1 1062 + 1062 2062 GERCC-00003 ERCC-00003 1 1024 + 1024 24 GERCC-00004 ERCC-00004 1 524 + 524 413 GERCC-00009 ERCC-00009 1 985 + 985 0 GERCC-00012 ERCC-00012 1 995 + 995 0 GERCC-00013 ERCC-00013 1 809 + 809 0 GERCC-00014 ERCC-00014 1 1958 + 1958 0 GERCC-00016 ERCC-00016 1 845 + 845 0   Visualisation   We can confirm that everything has worked and the ERCC sequences are mapping to the custom genome build by inspecting the resulting BAM file. The BAM file can be visualised in a genome browser such as IGV. I am using SeqMonk. I created a new genome in SeqMonk using the customised ERCC FASTA genome file and the custom ERCC GTF file. I then imported my BAM file. Below is a preview from SeqMonk.                           Reads aligned to ERCC ‘genes’ visualised in SeqMonk. Red bars at top denote ERCC ID. Mapped reads are seen stacked underneath.                                        Zoomed-in view of the ERCC ID ‘ERCC-00003’ and the 24 reads mapping to this location.                The SeqMonk visualisation adds up well with the read counts from QoRTs and featureCounts. For example, the 24 reads mapping to ERCC-00003. The read counts are now ready to be used in downstream analyses such as differential gene expression using DESeq2, EdgeR or Limma. The ERCC ID counts will be considered just like any other gene.  ","categories": ["Code","Research"],
        "tags": ["NGS","RNA-Seq"],
        "url": "/read-counts-of-rna-seq-spike-ins/",
        "teaser": null
      },{
        "title": "Autumn in Fulufjället",
        "excerpt":"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           Autumn in Fulufjällets National park in western Sweden. Fulufjällets national park on the border with Norway is a great location for a weekend away in nature.  The place is charming with vast meadows, forests, lakes, streams, mountains and idyllic houses. Fulufjället is known for its well preserved primeval forests, wildlife and unique geology. Fulufjället is also home to Sweden’s highest waterfall; Njupeskär falls and oldest tree; Old Tjikko. The location is also popular for winter activities such as skiing, snowmobiling and ice climbing on the Njupeskär falls. The park has an attractive and informative visitor centre (naturum) with user information including maps and guides. The park has wind shelters and camping sites. Part of the national park is located in Norway.  ","categories": ["Photography","Travel"],
        "tags": ["Autumn","Forest","Hiking","National Park","Road Trip","Sweden"],
        "url": "/autumn-in-fulufjallet/",
        "teaser": null
      },{
        "title": "Golden colours of the Autumn",
        "excerpt":"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Autumn has these magical colours with shades of reds, yellows and orange. And Uppsala has quite a few spot to enjoy the autumn colours.  ","categories": ["Photography"],
        "tags": ["Autumn","Uppsala","Sweden"],
        "url": "/autumn-2016/",
        "teaser": null
      },{
        "title": "Humanity has never lived in better times",
        "excerpt":"It may seem like the world is descending into total chaos, violence, and destruction. War in Syria, Ukraine, Yemen, Islamic state, migrant crisis, Ebola, plane crashes, earthquakes, tsunamis and what-not. The more news you watch, the more worried you will be. This is because the news outlets tend to focus on spectacularly negative instances. Violence, atrocities, and hatred are thrown into the spotlight and into the lives of common people. With the ever increasing digital connectivity, it is easy to disseminate information and to absorb information at an unprecedented level. Relatively smaller incidents have a larger voice. As said by Ray Kurzwil, “The world isn’t getting worse, our information is getting better”. To appreciate the world we live in, we have to put things into a wider context.   The fact is that humanity has never lived in a better time than now in pretty much every aspect you look at; war, violence, diseases, poverty are all at the lowest it has ever been. Of course, there is still a long way to go, but this is the best it has been since the beginning of humankind. To prove my point, here we evaluate human progress using some real data and simple time-series plots. Most of the data and information was obtained from OurWorldInData.   Life expectancy   We look at life expection as a metric for human health. Life expectancy is the mean age that an individual is expected to live. Mean global life expectancy has increased from around 30 years to over 70 years in 240 years.                           Average global life expectancy from 1770 to 2015.                A person born in 1770 would on average live for 30 years while an individual born in 2015 on average is expected to live up to 70 years old. Life expectancy should not be confused with lifespan. The lifespan may have remained the same while factors that detrimentally affect human health such as child mortality, diseases, epidemics, wars, poverty, malnutrition etc has dropped drastically which allows people to live longer lives. Living longer means that ageing itself is a health risk due to increasing old-age related health disorders such as Alzheimers, various Cancers, Osteoarthritis etc. Perhaps increasing human lifespan may also be feasible with recent developments in biotechnology and ageing research. Read more about life expectancy here.   Child mortality   Child mortality is the percentage of deaths in children under the age of 5.                           Mean global child mortality from 1800 to 2010.                Child mortality is down from 40% in 1800 to 5% in 2015. This means that 95% of newborns in 2015 are born alive and survive at least first 5 years of their life. Child mortality is directly linked to life expectancy and is the largest contributing factor for increased life expectancy. Child mortality although indicating child survival is also an indicator of medical progress, maternal health etc. Go here to read more about child mortality, changes, and causes.   Vaccination   Vaccination is possibly one of the greatest achievements in medicine. The ability to tackle a diseases/pathogen before an outbreak/infection. Vaccines although discovered in the late 1700s, was globally implemented and coordinated by the World Health Organization in the 1960s. Smallpox was the first disease to be eradicated by vaccination by 1980. Vaccination was quickly introduced for a wide range of diseases.                           Percentage of the world population with access to DPT3 vaccination since 1980 to 2015.                As an example, the DTP3 (Diphtheria-Tetanus-Pertussis) vaccine in 2015 reaches more than 6 billion people (86%) compared to less than 1 billion people (21%). WHO is on the brink of eradicating the second disease Polio.   In recent years, a fraudulent Lancet article (later retracted) linking vaccination to autism in children has given rise to an anti-vaccination movement especially in the US with parents choosing to not vaccinate their children. This has led to recent outbreaks of Pertussis and Measles. This again demonstrates how critical vaccinations are to maintain herd immunity in the population.   Literacy   Education is the driving force behind progress. Education gives rise to scientific progress, freedom, tolerance and equality.                           Percentage of the world population with basic literacy from 1800 to 2010.                The percentage of people with basic literacy has increased from 10% to 85% over the last 200 years. Literacy rates below 30% are now confined to Afghanistan, Central Africa and Sub-Saharan Africa. Se here to read more about literacy.   Poverty   Next we look at poverty as a measure of standard of living.                           Percentage of the world population living in extreme poverty from 1820 to 2010.                Although one might come across reports of rising unemployment, increasing inflation and economic instability, the fact is that absolute poverty has dropped by more than 70% in the last 200 years. Alleviating poverty not only provides people with better health education and standard of life, it also creates satisfied and happy people.   Political Regimes   We also explore political regimes as a measure of human rights and freedom of expression. Human beings have always lived in a hierarchical system of classes and castes. The idea of considering all human beings to be equal is a new concept.                           Change in composition of the world population living in various political regimes from 1816 to 2010.                War   War is arguably detrimental to human progress or a melting pot of innovation due to necessity. We use war-related deaths as a measure of violence. War related deaths have been declining since the last major war in 1946. If the number of war deaths were controlled for population size, the decline is even more pronounced. We are living in the most peaceful era of our species’ existence as far as violence is concerned.                           Number of global war fatalities per 100,000 people from 1946 to 2013.                The matter of if war is on the decline is hard to answer. Perhaps war fatalities have declined but many are seriously wounded. Perhaps better body armour, better military evacuation and medical care may play an important role to reduce fatalities. War classification has changed to civil conflict, counter-terrorism, drug-violence, organised gangs etc. Major wars only happen a couple of times a century and perhaps sufficient time has not passed since the last war. Only time will tell.   Environment   If everything is getting better, something must be getting worse? It seems like we can’t have everything. All this human progress has come at great cost to the environment. Two centuries of industrial development has resulted in significant damage and destruction to our planet. Some of the notable environmental effects as a consequence of human activity are reduced forest cover, increased air pollution, dramatic increase in the number of extinct animal and plant species, changes in landscape composition and man-made catastrophes (oil spill etc).                           Total global carbon dioxide emissions since 1800 to 2010.                The carbon dioxide levels in the atmosphere is a commonly used metric to assess the consequence of human activity on the environment. Global Co2 emissions have been increasing faster and faster since the industrial revolution as the world gets richer and richer. Methane emissions have also increased dramatically especially with the advent of mass cattle farms.   Conclusion   Declinism and rising depression are worrying trends. Sure, the world is not anywhere close to perfect, but we have sophisticated tools and better opportunity to solve those problems. It is easy to miss out on the slow progress we make as a species while we fixate on the shortcomings. We have to be optimistic and keep the momentum of change rolling. We don’t see how good things are because we are not aware of how bad things were.  ","categories": ["Code","Research"],
        "tags": ["Data Analysis","Data Visualisation","R"],
        "url": "/humanity-never-lived-better-times/",
        "teaser": null
      },{
        "title": "Tropical break in India",
        "excerpt":"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         It was mostly visiting relatives, excessive eating and dealing with heat, humidity and traffic. On the bright side, the ayurvedic massages can be quite relaxing. This also presents opportunities to enjoy some local cuisine. I do enjoy Indian food, but South Indian food tend to be quite spicy and it can be challenging to find food that agrees with my taste buds. I usually eat a lot of non-veg, but here in Kerala, there is so much diversity in vegetarian food that I would happily become a vegetarian.   The mountainous regions of Kerala, such as Idukki district is one my favourite destinations to escape the heat and pollution of cities. The route is scenic and if you are lucky, you might even get to see some wildlife as a lot of Idukki is part of the Western ghats nature reserve and national parks. This is also an ideal location for stargazing as there is minimal light pollution. As a plus, there tends to be less mosquitoes in the mountains.  ","categories": ["Photography","Travel"],
        "tags": ["Forest"],
        "url": "/india-2017/",
        "teaser": null
      },{
        "title": "Multiple Statistical Testing",
        "excerpt":"In a standard statistical test, one assumes a null hypothesis, performs a statistical test and computes a p-value. The estimated p-value is compared to a predetermined threshold (usually 0.05). If the estimated p-value is greater than 0.05 (say 0.2), it means that there is a 20% chance of obtaining the current result if the null hypothesis is true. Since we decided our threshold as 5%, the 20% is too high to reject the null hypothesis and we accept the null hypothesis. Now, if the estimated p-value was less than 0.05 (say 0.02), there is a 2% probability of obtaining the observed result if the null hypothesis is true. Since 2% is a very low probability and it is below our threshold of 5%, we reject the null hypothesis and accept an alternative hypothesis.   The 5% threshold, although giving us high confidence, is an arbitrary value and does not absolutely guarantee an outcome. There is still the possibility that we are wrong 5% of the time. This is known as the probability of a Type I error. A Type I error occurs when a researcher falsely concludes that an observed difference is real, when in fact, there is no difference.   That was the story of a single statistical test. With large data, it is common for data analysts to do multiple statistical tests on the same data. Similar to a single test, each test in a multiple test has the 5% Type 1 error rate. And this accumulates for the number of tests.   The probability of accumulating type I (t1) errors can be shown as below:       P(making an t1 error) = a     P(not making a t1 error) = 1-a     P(not making a t1 error in m tests) = (1-a)^m     P(not making at least 1 t1 error in m tests) = 1-(1-a)^m      Probability of a t1 error with one test (m=1):     1 - (1 - 0.05)^1     0.05      Probability of a t1 error with 10 tests (m=10):     1 - (1 - 0.05)^10     0.40   So, for 10 tests, there is a 40% chance of observing at least one significant result even if there are none. If 100 genes are tested, 5% (5 genes) could be found to be significant by chance even if they are not. In datasets such as RNA-seq data where thousands of genes are tested, the type I error rate gets inflated massively. This is the reason why p-values from multiple testing needs to be corrected.   Methods for dealing with multiple testing usually involves adjusting p-value threshold in some way, so that the probability of observing at least one significant result due to chance remains below a desired significance level.   There are several categories of methods to controlling Type I errors:  Per Comparison Error Rate (PCER), Per Family Error Rate (PFER), Family-Wise Error Rate (FWER), False Discovery Rate (FDR), Positive False Discovery (pFDR) etc. We will only consider two categories of methods to correcting for multiple testing, namely, controlling the family-wise error rate (FWER) and controlling the false discovery rate (FDR).   Control Family-wise error rate (FWER)   Bonferroni correction   The classical single-step approach and the simplest method is the Bonferroni correction. It is also the most stringent method. All tested p-values are multiplied by the total number of tests.       Corrected p-value = p-value * num_tests   Example:       Let n=1000, p-value threshold=0.05     Rank Gene p-value     1 A 0.00002     2 B 0.00004     3 C 0.00009      Correction:     GeneA p-value = 0.00002 * 1000 = 0.02 (significant)     GeneB p-value = 0.00004 * 1000 = 0.04 (significant)     GeneC p-value = 0.00009 * 1000 = 0.09 (not significant)   Using a p-value of 0.05 means there is a 5% or less chance of 1 test being a false positive. Another definition is to expect 0.05 genes to be significant by chance. Using the Bonferroni correction brings the error probability back to 5% as shown below.       1-(1-(a/m))^m     1 - (1 - 0.05/10)^10     0.05   The Bonferroni correction has a few drawbacks of its own. It may be suitable when we want to avoid even a single false positive in the results. But in most cases, this is too extreme such that false negatives (type II errors) may be inflated. Thus truly important differences may be deemed non-significant.   Bonferroni step-down (Holm)   This is a modified Bonferroni method using sequential step-down and it is slightly less stringent than the Bonferroni method.           Take the p-value of each gene and rank them from the smallest to the largest.            Multiply the first p-value by the number of genes present in the gene list: if the end value is less than 0.05, the gene is significant. ` Corrected P-value= p-value * n`            Multiply the second p-value by the number of genes less 1: ` Corrected P-value= p-value * (n-1)`            Multiply the third p-value by the number of genes less 2: ` Corrected P-value= p-value * (n-2)` Follow that sequence until no gene is found to be significant.       Example:       Let n=1000, p-value threshold=0.05     Rank Gene p-value     1 A 0.00002     2 B 0.00004     3 C 0.00009      Correction:     GeneA p-value = 0.00002 * 1000 = 0.02 (significant)     GeneB p-value = 0.00004 * 999 = 0.039 (significant)     GeneC p-value = 0.00009 * 998 = 0.0898 (not significant)   Controlling FWER is generally too extreme and appropriate only in situations where a researcher does not want ANY false positives at the risk of missing out few true positives. A more relevant quantity to control is the FDR.   Control False discovery rate (FDR)   Controlling for False discovery rate (FDR) is currently the preferred approach to account for multiple testing. FDR sets the proportion of discoveries (significant results) that are actually false positives. For example, one could set the FDR to 0.05 and all tested p-values are adjusted. Of the tests that are still significant (below 0.05), 5% could be false positives. Depending on the situation, FDR could be set to 5%, 10%, 15% etc. The False Discovery Rate (FDR) control was developed by Benjamini and Hochberg in 1995.   Benjamini-Hochberg   This correction is the least stringent  and the most powerful option. Here is how it works:   Here is how it works:     Rank the p-value of each gene in order from the smallest to the largest.   Multiply the largest p-value by the number of genes in the test.   Take the second largest p-value and multiply it by the total number of genes in gene list divided by its rank. If less than 0.05, it is significant. ` Corrected p-value = p-value * (n/n-1)`   Take the third p-value and proceed as in step 3: ` Corrected p-value = p-value * (n/n-2)`   Example:       Let n=1000, p-value threshold=0.05     Rank Gene p-value     1000 A 0.1     999 B 0.04     998 C 0.01      Correction:     GeneA P-value &gt; 0.05 =&gt; gene D (not significant)     GeneB P-value = (1000/999) * 0.04 = 0.04004 (significant)     GeneC P-value = (1000/998) * 0.01 = 0.01002 (significant)   Practical example   We try out some examples of p-value correction with randomly generated data in R. We use the function p.adjust() from stats package and mt.rawp2adjp() from the multtest package.   We generate 10,000 random raw p-values between 0.00001 and 0.01. We assume these are p-values generated from 10,000 statistical tests. Since all p-values are below the threshold of 0.05, we will call this set the low set. Now we correct this dataset for multiple testing in R using six different methods: Benjamini-Hochberg (BH), Hommel, Benjamini-Yekutieli (BY), Holm, SidakSD and Bonferroni (B). The tests are roughly ordered in the level of stringency from least stringent to most stringent.   library(Cairo) library(ggplot2) library(tidyr) library(dplyr) library(multtest)  textcol &lt;- \"grey30\"  #create a vector of random p-values df1 &lt;- data.frame(num=as.character(1:10000),  Raw=runif(n = 10000, min=0.00001, max=0.01))  #create a dataframe of adjusted p-values df1 &lt;- df1 %&gt;% mutate(Holm=p.adjust(Raw, method=\"holm\")) %&gt;% mutate(Hommel=p.adjust(Raw, method=\"hommel\")) %&gt;% mutate(Bonferroni=p.adjust(Raw, method=\"bonferroni\")) %&gt;% mutate(BenjaminiHochberg=p.adjust(Raw, method=\"BH\")) %&gt;% mutate(BenjaminiYekutieli=p.adjust(Raw, method=\"BY\")) %&gt;% mutate(SidakSD=mt.rawp2adjp(Raw, proc=\"SidakSD\")$adjp[, 2])  df2 &lt;- df1 %&gt;% gather(key=method, value=measure, -num) %&gt;% mutate(method=factor(method, levels=c(\"Raw\", \"BenjaminiHochberg\", \"Hommel\", \"BenjaminiYekutieli\",  \"Holm\", \"SidakSD\", \"Bonferroni\"))) %&gt;% mutate(sig=ifelse(measure&lt;0.05, \"significant\", \"not-significant\")) %&gt;% mutate(sig=factor(sig, levels=c(\"significant\", \"not-significant\")))  # scatterplot p &lt;- ggplot(df2, aes(x=method, y=measure, colour=sig))+ #geom_point(alpha=0.8)+ geom_boxplot(alpha=0.8, size=0.25, position=\"identity\", outlier.size=1)+ geom_hline(yintercept=0.05, colour=textcol, size=0.6, alpha=0.4, linetype=\"21\")+ coord_flip()+ xlab(\"\")+ ylab(expression( italic(p)-value))+ labs(caption=\"Boxplots showing 10,000 random raw p-values (0.00001 - 0.01) and six methods of multiple testing correction.\")+ scale_colour_manual(values=c(\"#66c2a5\", \"#fc8d62\"), guide=guide_legend(nrow=1))+ theme_bw(base_size=10)+ theme( legend.position=\"top\",  legend.direction=\"horizontal\",  legend.justification=\"right\",  legend.title=element_blank(),  legend.margin=margin(0, 4, -10, 0),  legend.text=element_text(colour=textcol, size=6, face=\"plain\"),  legend.key.size=grid::unit(0.25, \"cm\"),  strip.background=element_rect(colour=\"white\", fill=\"grey90\"),  strip.text=element_text(face=\"bold\"),  axis.text.x=element_text(size=9, colour=textcol),  axis.text.y=element_text(vjust=0.2, colour=textcol),  axis.title=element_text(colour=textcol, size=8.5),  axis.ticks=element_blank(),  plot.title=element_text(colour=textcol, hjust=0, size=13, face=\"bold\"),  #plot.subtitle=element_text(colour=textcol, hjust=0, vjust=-1),  plot.caption=element_text(colour=textcol, hjust=0, size=5, lineheight=-1),  plot.margin=margin(10, 10, 5, 10),  panel.grid=element_line(colour=\"grey99\"),  panel.border=element_blank())  ggsave(\"pval-scatterplot-low.png\", p, height=6, width=13, units=\"cm\", dpi=300, type=\"cairo\")                           Boxplots showing 10,000 raw p-values between 0.00001 and 0.01 and corrected p-values using six methods. Data is shown as boxplots. Central vertical line indicates median and points indicate outliers. Dotted line across all data denotes the significance threshold of 0.05. Significant data is coloured in green and non-significant data is coloured in red.                We see that BH and Hommel correction keep all 10,000 p-values below the significance threshold of 0.05. BY is close but not significant. BY values can be made significant by raising the FDR from 5% to 12% or so. Holm, Sidak and B correct aggressively and most p-values are insignificant.   Now we try the same idea but start with a set of p-values that are above and below the threshold of 0.05. We generate 10,000 random raw p-values between 0.00001 and 0.1. We will call this the lowhigh set.   #create a vector of random p-values df1 &lt;- data.frame(num=as.character(1:10000),    Raw=runif(n = 10000, min=0.00001, max=0.1))                           Boxplots showing 10,000 raw p-values between 0.00001 and 0.1 and corrected p-values using six methods. Data is shown as boxplots. Central vertical line indicates median and points indicate outliers. Dotted line across all data denotes the significance threshold of 0.05. Significant data is coloured in green and non-significant data is coloured in red.                We see that the raw values are distributed below and above the 0.05 threshold. All correction methods push all values above the 0.05 threshold and none of the tests is significant.   So the starting distribution of p-values certainly has an effect on the outcome of the corrected p-values. Another critical factor is the number of samples/number of tests. The less the number of starting p-values, the less aggressive will be the correction. To demonstrate that, we generate a low set and a lowhigh set, but for each, we will generate five different sets of starting number (n=10,100,1000,10000,100000). Corrections using 100,000 data points can take a few minutes.   dlist &lt;- list(length=5) vec &lt;- c(10, 100, 1000, 10000, 100000) for(i in 1:length(vec)) {   print(vec[i])   #create a dataframe of adjusted p-values   dlist[[i]] &lt;- data.frame(num=as.character(1:vec[i]), Raw=runif(n = vec[i], min=0.00001, max=0.01)) %&gt;%   mutate(Holm=p.adjust(Raw, method=\"holm\")) %&gt;%   mutate(Hommel=p.adjust(Raw, method=\"hommel\")) %&gt;%   mutate(Bonferroni=p.adjust(Raw, method=\"bonferroni\")) %&gt;%   mutate(BenjaminiHochberg=p.adjust(Raw, method=\"BH\")) %&gt;%   mutate(BenjaminiYekutieli=p.adjust(Raw, method=\"BY\")) %&gt;%   mutate(SidakSD=mt.rawp2adjp(Raw, proc=\"SidakSD\")$adjp[, 2]) %&gt;%   gather(key=method, value=measure, -num) %&gt;%   mutate(method=factor(method, levels=c(\"Raw\", \"BenjaminiHochberg\", \"Hommel\", \"BenjaminiYekutieli\",    \"Holm\", \"SidakSD\", \"Bonferroni\"))) %&gt;%   mutate(sig=ifelse(measure&lt;0.05, \"significant\", \"not-significant\")) %&gt;%   mutate(sig=factor(sig, levels=c(\"significant\", \"not-significant\"))) %&gt;%   mutate(n=paste0(\"n=\", vec[i])) }  df3 &lt;- bind_rows(dlist) %&gt;%   mutate(n=factor(n))  # scatterplot p &lt;- ggplot(df3, aes(x=method, y=measure, colour=sig))+   geom_boxplot(alpha=0.8, size=0.25, position=\"identity\", outlier.size=0.85)+   geom_hline(yintercept=0.05, colour=textcol, size=0.4, alpha=0.4, linetype=\"21\")+   coord_flip()+   facet_wrap(~n, nrow=5)+   xlab(\"\")+   ylab(expression( italic(p)-value))+   labs(caption=\"Boxplots showing raw p-values and six adjustment methods using five different number of tests (n).\")+   scale_colour_manual(values=c(\"#66c2a5\", \"#fc8d62\"), guide=guide_legend(nrow=1))+   theme_bw(base_size=10)+   theme(   legend.position=\"top\",    legend.direction=\"horizontal\",    legend.justification=\"right\",    legend.title=element_blank(),    legend.margin=margin(0, 4, -10, 0),    legend.text=element_text(colour=textcol, size=6, face=\"plain\"),    legend.key.size=grid::unit(0.25, \"cm\"),    strip.background=element_rect(colour=\"white\", fill=\"grey96\"),    strip.text=element_text(face=\"bold\", size=6, lineheight=0.4),    axis.text.x=element_text(size=6.5, colour=textcol),    axis.text.y=element_text(size=6.5, vjust=0.2, colour=textcol),    axis.title=element_text(colour=textcol, size=8),    axis.ticks=element_blank(),    plot.title=element_text(colour=textcol, hjust=0, size=13, face=\"bold\"),    plot.caption=element_text(colour=textcol, hjust=0, size=4, lineheight=-1),    plot.margin=margin(10, 10, 5, 10),    panel.grid.major=element_line(colour=\"grey94\", size=0.25),    panel.grid.minor=element_line(colour=\"grey96\", size=0.15),    panel.border=element_blank())  ggsave(\"pval-n-scatterplot-low.png\", p, height=13, width=9, units=\"cm\", dpi=300, type=\"cairo\")                           Boxplots showing 10,000 raw p-values between 0.00001 and 0.01 and corrected p-values using six methods and six different number of tests. Data is shown as boxplots. Central vertical line indicates median and points indicate outliers. Dotted line across all data denotes the significance threshold of 0.05. Significant data is colored in green and non-significant data is coloured in red.                For the low set, as n increases from 10 to 100000, BH and Hommel methods keep all values below 0.05, while other methods drastically push the p-values further out towards 1. BY finds a compromise by keeping p-values low and significance can be achieved by increasing FDR.                           Boxplots showing 10,000 raw p-values between 0.00001 and 0.1 and corrected p-values using six methods and six different number of tests. Data is shown as boxplots. Central vertical line indicates median and points indicate outliers. Dotted line across all data denotes the significance threshold of 0.05. Significant data is colored in green and non-significant data is coloured in red.                For the lowhigh set, the corrections seem to be even stronger pushing the corrected p-values higher than 0.05 with all methods and all n values.   Conclusion   As a general guide, multiple testing may not be necessary for exploratory analyses for hypothesis generation. FDR control may be suitable for exploratory analyses or screening to minimise false positives for further investigation. FWER control may be necessary for endpoint confirmatory analyses to absolutely eliminate any false positives.   In conclusion, when performing multiple statistical tests on the same data, it is important to correct the resulting p-values. Controlling for FDR is the more preferred way to adjust p-values. The most popular method at least for genomics is the Benjamini-Hochberg.   References           ETH multiple testing guide            Biostathandbook multiple comparison            Benjamini, Y., &amp; Hochberg, Y. (1995). Controlling the false discovery rate: a practical and powerful approach to multiple testing. Journal of the royal statistical society. Series B (Methodological), 289-300.            Pollard, K. S., Dudoit, S., &amp; van der Laan, M. J. (2005). Multiple testing procedures: the multtest package and applications to genomics. In Bioinformatics and computational biology solutions using R and bioconductor (pp. 249-271). Springer New York.            Bender, R., &amp; Lange, S. (2001). Adjusting for multiple testing—when and how?. Journal of clinical epidemiology, 54(4), 343-349.            Noble, W. S. (2009). How does multiple testing correction work?. Nature biotechnology, 27(12), 1135-1137.            Reiner, A., D. Yekutieli and Y. Benjamini. (2003). Identifying differentially expressed genes using false discovery rate controlling procedures. Bioinformatics 19: 368-375.      ","categories": ["Code","Research"],
        "tags": ["R","Statistics"],
        "url": "/multiple-statistical-testing/",
        "teaser": null
      },{
        "title": "Running ImageJ on a Linux cluster",
        "excerpt":"I use ImageJ for many of my image analysis needs. My desktop computer runs Windows 7 and it has pretty solid specs with Core i7 processor and 16GB RAM. I recently had to handle some large tiff stacks (4-5gb) and it simply wouldn’t work on my desktop as I constantly ran into ‘out of memory’ errors. So I decided to run them on a computing cluster instead since I have access to one. Running on a cluster might be useful when handling data with large memory requirements or to perform computations on numerous files in parallel by distributing load to multiple cores. It took me a while to figure out how to get things to work, so I thought I would make a record of it. And this might hopefully be useful to others.      The computing cluster runs Linux OS CentOS 7. The cluster uses the SLURM system for job submissions. I have set up a directory structure is as follows:   home +--img/    +--from/    +--to/    +--work/    +--programs/    +--scripts/   I have downloaded the ImageJ Java executable for Linux and copied it to the programs directory.   Substacking tiff stacks   In this example, the script performs a simple basic operation: Open a tiff-stack from a source directory (from), subset the stack (skip frame1, keep frames 2 to end) and write the new file to a destination directory (to). And this will be performed in batch for all files in the source directory. In this example, the file will be processed one after the other (serial) and not in parallel.   I have two script files: a bash script file (tifftosubstack.sh) and an ImageJ macro file (tifftosubstack.ijm) in my scripts directory.   My tifftosubstack.sh bash script looks like below:   #!/bin/bash  #SBATCH -A accountname #SBATCH -p core #SBATCH -n 3 #SBATCH -t 2:00:00 #SBATCH -J tifftoss  echo \"Tiff to Substack\"  #run imagej script echo \"Running ImageJ ...\" /home/programs/fiji-linux-64/ImageJ-linux64 --headless -macro /home/img/scripts/tifftosubstack.ijm  echo \"End of Script\"   The batch script just runs the ImageJ executable along with the ImageJ macro file. The SBATCH lines are only relevant if the script is submitted as a SLURM job. It doesn’t matter where the bash script is run because there are no relative links. And my ImageJ macro script looks as below:   // ImageJ batch script to substack tiff stack // 2016 Roy Mathew Francis  //set directories dirsource = \"/home/img/from/\"; dirresult = \"/home/img/to/\";  setBatchMode(true); list = getFileList(dirsource); print(\"======================================================\"); print(\"~~ ImageJ Tiff Substack ~~\"); for (i=0; i&lt;list.length; i++)   {     print(\"------------------------------------------------------\");     print(\"Processing \"+(i+1)+\" of \"+list.length+\"...\");     spath = dirsource+list[i];      //open tiff image     print(\"Opening \"+spath);     open(spath);     title = getTitle();     imgorig = getImageID();      //substack slice 2 to end   selectImage(imgorig);   print(\"Slices: \"+nSlices);   run(\"Make Substack...\", \"  slices=2-\"+nSlices);    //save tiff file   imgsub = getImageID();   selectImage(imgsub);     saveAs(\"Tiff\", dirresult+title);     print(dirresult+title+\" saved.\");      //close all     close(\"*.tif\");   }  print(\"======================================================\"); eval(\"script\",\"System.exit(0);\");   The ijm script is a loop that reads each file in the source directory, creates a substack (frame 2 to end of stack) and writes it out to the destination directory. The ijm file can be modified to perform other kinds of ImageJ processing. Once the two scripts are ready, verify that the paths are correct and set execute permissions on them.   chmod +x tifftosubstack.sh   Then run the bash script.   ./tifftosubstack.sh   If your current directory is not scripts, then use the full path to execute the bash script. Once the execution is complete, the new tiff stack will be in the ‘to’ directory.   Substacking compressed tiff stacks   This is an example where the loop is created in the bash script instead. The operation is same as before except that we start with compressed tiff stacks (gzipped). The steps will be as follows: the file is copied from source directory (from) to a temporary working directory (work), uncompressed, processed in ImageJ and exported to destination directory (to), and finally compressed. This will be performed in batch for all files in the source directory and the files will be processed one after the other (serial) and not in parallel.   I have two script files: a bash script file (tifftosubstackcompress.sh) and an ImageJ macro file (tifftosubstackcompress.ijm).   My tifftosubstackcompress.sh bash script looks like below:   #!/bin/bash  #SBATCH -A accountname #SBATCH -p core #SBATCH -n 3 #SBATCH -t 2:00:00 #SBATCH -J tifftoss  echo \"Tiff to Substack Compress\"  cd /home/img/work/ for z in /home/img/from/*.gz   do      #copy file to work directory     echo \"Copying $z ...\"   cp $z .    #get filename from path and remove .gz   b=$(basename $z)   newfilename=\"${b/.*/}\"    #Decompress file   echo \"Decompressing $b ...\"     gunzip $b      #run imagej script   echo \"Running ImageJ ...\"   /home/programs/fiji-linux-64/ImageJ-linux64 --headless -macro /home/img/scripts/tifftosubstack.ijm    #compress new file   cd /home/img/to/   echo \"Compressing $newfilename.tiff ...\"   gzip \"$newfilename.tif\"      #remove temporary file   cd /home/img/work/   echo \"Removing $newfilename.tiff\"   rm -rf \"$newfilename.tiff\"   done  echo \"End of Script\"   The bash script loop copies each gzipped tiff from the source to work directory, decompresses it, runs ImageJ, and compresses the output tiff.   And my ImageJ macro script looks as below:   // ImageJ batch script to substack tiff stack // 2016 Roy Mathew Francis  //set directories dirsource = \"/home/img/work/\"; dirresult = \"/home/img/to/\";  setBatchMode(true); list = getFileList(dirsource); print(\"======================================================\"); print(\"~~ ImageJ Tiff Substack ~~\"); for (i=0; i&lt;list.length; i++)   {     print(\"------------------------------------------------------\");     print(\"Processing \"+(i+1)+\" of \"+list.length+\"...\");     spath = dirsource+list[i];      //open tiff image     print(\"Opening \"+spath);     open(spath);     title = getTitle();     imgorig = getImageID();      //substack slice 2 to end   selectImage(imgorig);   print(\"Slices: \"+nSlices);   run(\"Make Substack...\", \"  slices=2-\"+nSlices);    //save tiff file   imgsub = getImageID();   selectImage(imgsub);     saveAs(\"Tiff\", dirresult+title);     print(dirresult+title+\" saved.\");      //close all     close(\"*.tif\");   }  print(\"======================================================\"); eval(\"script\",\"System.exit(0);\");   The only line that has changed is the source directory path (dirsource). Set execute permissions and execute the bash script as before. After the script has executed, a new compressed tiff file will be present in the to directory.   These are simple examples to get started. You can obviously build on this for all sorts of image analysis operations.  ","categories": ["Code","Research"],
        "tags": ["Image Analysis","ImageJ"],
        "url": "/running-imagej-linux-cluster/",
        "teaser": null
      },{
        "title": "Exploring cosmic landscapes in Iceland",
        "excerpt":"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Iceland is a nature lover’s paradise. Same goes for a landscape photographer. The nature in Iceland is just out of this world. It’s just remarkable how diverse the topography is, for a relatively small island. It’s also a great location if you want to get away from people. You could be driving for hours without meeting a single car. In this post I want to talk about some of the experience and some of the practicalities of travelling in Iceland.     Pre-travel planning   I might be a bit crazy about planning trips, but this was my first trip to Iceland and I had planned this trip at least a year in advance to meticulous detail. I had booked flights, booked rental car and accommodation. I mapped out all interesting locations to visit, routes, fuel stations and fuel prices (See map below). I planned for about 2 weeks allowing some flexibility depending on best flight dates. I picked June as the weather should be ideal and most roads, attractions, restaurants, museums, and ferry routes are open. Also, you get to see the midnight sun. Flights, car and accommodation is obvious to book in advance as you can get the best deal. I bought flight tickets from IcelandAir as there were not many options to start with the price didn’t vary much. I had explored numerous car rental options. Ultimately, I rented an SUV (Toyota RAV4) from IceRental4x4. I was happy with the rental company, as they had a wide range of SUVs to choose from, relatively good prices and they pick you up directly from the airport. The car was small enough to be fuel efficient and large enough for luggage and two persons to sleep. I planned 3 nights of stay in hotels from booking.com. If you are camping in a car or tent, it is a good idea to book a proper hotel at least a couple of times in the trip to organise and refresh yourself.   Route                           Where to go next?                I usually research about the destination and pick the spots I want to visit. This could be landscapes, museums, hotels, camping grounds, sight-seeing etc. I then plan the best route that connects these points. I then decide on how much distance can be covered each day leaving enough time for exploration. In Iceland, one would usually start from Reykjavik, go around on the ring road and return back. I decided to go anti-clockwise on the ring road for no particular reason. I haven’t come across any good arguments for whether you should go clockwise or anti-clockwise.   Below is a Google MyMaps map with lots of information that I have collected over time.     The map contains interesting locations that include landscapes (cliffs and mountains), beaches, waterfalls, bird-watching spots, hot pools, camping locations, fuel stations and sights in Reykjavik. This is by no means a complete list. Locations such as landscapes, sights and wildlife was manually collected from various Iceland guides. The fuel stations were obtained from Gasvaktin, camp sites from camping.info and hot pools from hotpoticeland. Our route is indicated in blue. The hotels I stayed in are also marked. Some of the items have not been grouped into layers due to layer limit.   Weather and clothing                           The Icelandic weather greets me at the airport.                The weather can be unpredictable and can change at a moment’s notice even in the middle of summer. Rain can be heavy and relentless. I had rain for 2 days continuous without a break. Even water resistant clothing is pretty useless in such rain. Now combine that with harsh wind and it can be frustrating even to open the car doors to look for something or set-up tent for camping etc. So I had to buy water-proof ponchos. These are not breathable but they cover your whole body and can even protect your camera equipment. It is also a good idea to bring an extra pair of shoes, socks, jacket and trousers so you have something to fall back on. Most of the coastal regions are extremely windy. Although the temperature is pleasant, the strong winds can make it uncomfortable and chilly. Some windproof jacket and headgear can be quite helpful. Also, extreme wind can damage the doors of vehicles and is not covered by the insurance. Parking against the wind direction can help. It is also a good idea to take a camera cleaning kit due to the weather.  Generally speaking, it is best to plan well and take everything you need with you as it can be pretty expensive to buy stuff.   Midnight sun   You are probably aware of the phenomenon called midnight sun, where, in polar latitudes, the sun doesn’t completely set and you end up with bright nights. This can be a pain if you trying to sleep in a car or a room with no blinds. So an eye-mask is a good thing to have. Midnight sun also has many positive aspects to it. No straining night driving. You are not limited to exploring during the day time hours. Light is available even through the night, so you could be visiting locations such as waterfalls with minimal tourists around. If you have clear skies, you get really long extended sunrise and sunsets. This is great if you are a photographer chasing the narrow elusive golden hour. But note that working hours or opening hours still apply. No one is going to sell you tickets or groceries in the middle of the night.   Camping and food                           Icelandic grocery chain ‘Bonus’.                Camping is an excellent option in Iceland. Camp sites are frequent and most of them are pretty decent. Camping is only allowed at designated campsites. Camping is not allowed at parking lots, rest stops, within city limits etc. Campsites usually have access to showers and toilets and some even a kitchen and wifi. You can buy a small gas canister at an outdoor store, which is good for preparing hot meals and can last you two weeks or so if used carefully. I always bought some groceries (water, milk, eggs, bread etc) and stored them in the car. Get a cool bag or an electronic cool box for perishables. Buying groceries at Bonus stores is the cheapest. Camping + buying groceries is significantly cheaper than staying at hotels/hostel and eating out. Having said that, you should probably try out some of the local cuisine while you are there, especially the seafood.. nom nom.   Is a SUV necessary?                                           I picked the Toyota RAV4 SUV for my trip after reading many guides and blogs. In retrospect, I am wondering if it was necessary. The RAV had sufficient space for two adults to sleep which was nice for camping. Other arguments in favour would be that an SUV is ideal for off-road driving and necessary for driving on F-roads. But, for a first time traveller, it is unlikely that you will even go anywhere off-road. There is plenty to see and explore just staying on the ring road and paved roads. Apart from being able to sleep inside, and storage space, a small car would have been just fine. A regular car would be cheaper on rental as well as fuel.   Hot springs                                           Iceland is famous for its hot springs due to its location along geothermal regions. Hot springs are supposed to be scattered across the country. The most popular and possibly the largest is the blue Lagoon at Keflavik. This is best to avoid as it is crowded and overpriced. It is best to find smaller natural isolated pools. This can be challenging to find as they can be difficult to reach or is private. I spent a some time trying to find some. I did come across a few. But they are either poorly marked on maps or is usually filled with people already. Maybe driving away from the ring road can give you access to better options.   Shopping   You are most certainly going to be buying stuff in Iceland for export. This can include souvenirs, clothing, wool, chocolates, handcrafted items etc. Remember to group purchases to at least 6000 ISK, so that you are eligible for tax-refund. If you buy items separately at a lower value, you will not be able to claim for tax-refund. And also get the completed tax refund form the seller at the time of purchase. Another point is that duty-free at Keflavik airport can be a good place to buy goods for cheap tax-free. So if you plan on buying souvenirs, clothing from Icewear, cosmetics etc, it is a good idea to do so at the airport.   Reykjavik sights                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Reykjavik, the capital of Iceland is a fairly small city that you could probably cover in a day. Accomodation is expensive and streets are narrow. Vandalism here are stunning works of art. The city is quite cute and has some interesting little places to explore. The tourist centre is a good place to start. There are coffee shops, wool shops, souvenir shops and pubs everywhere. There are probably a bit too many ‘Americanised’ shops to cater to the swarms of American tourists. Americans seem to bring with them bad habits such as tipping. Tipping is not a custom anywhere in Scandinavia because waiters are paid a good fixed minimum wage and do not have to depend on petty tips. Moving on from the rant, there are some neat museums too. My personal favourite attraction was the Perlan Museum. This is probably one the most modern and innovative museums I have ever visited. Reykjavik has taken some boring water storage tanks and converted it into an attraction. The musuem covers the natural wonder of Iceland such as its geology, volanoes, geothermal systems, glaciers, ice caves and northern lights. There is even an artifically created ice cave in the musueum for guests to experience. Perlan also has a neat observation deck where you can get a good view over the city. Another one of my favourite attractions was the Whales of Iceland musuem. If you want to get a feel for the size of whales and learn about whales, this is the place to go.   I guess that’s about it. I hope this has been helpful and wish you a fantastic trip!                                           ","categories": ["Photography","Travel"],
        "tags": ["Hiking","Iceland","National Park","Road Trip","Wildlife"],
        "url": "/iceland/",
        "teaser": null
      },{
        "title": "Road trip through Texas, USA",
        "excerpt":"This was my first trip to the USA, mostly to attend a conference in Austin, Texas. And then I took a week to go around and explore the place. I managed to visit Houston and San Antonio as well.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              As a first time visitor, the first thing you notice is just how massive everything is. The cities are huge, so are the streets and houses and people and portions of meals. You can’t just get out of your hotel and decide to walk to get somewhere. You won’t get anywhere. I feel like things are spaced far apart. If you need to get 5 things from 5 different stores, it could easily take you a whole day if you took the bus. So having a car is almost a necessity. So I rented a car to make my life easier. And then God said, let there be traffic.   I was in Austin for my stay due to the conference. Texas gets seriously hot in summer. The day time temperature was around 37°C. But I am still going to complain about how cold it was inside J. W. Mariott. Austin has loads of good places to eat and reasonably priced. Street food like burgers and tacos are simply delicious. I visited Houston for one and only one reason: Johnson space centre. And I was not disappointed. Yes, you get to see loads of real space memorabilia including a space shuttle. This place is just awesome if you are into space stuff. Houston and Austin both seem to have bridges with high density of mexican free-tailed bats. Watching hundreds or thousands of bats take off into the dusk is quite spectacular. San Antonio is even more interesting. The river walk promenade is exceptionally beautiful. You could just hang around and watch the river taxis go about for hours. The whole city centre is lively and vibrant. I also visited the limestone caves at Natural Bridge Caverns and  Medina lake. The limestone caves were pretty cool but it’s a shame that you can’t explore by yourself. You have to be in a group with a guide.   I stayed at Motel 6 in both Houston and San Antonio and I was quite pleased. They are not as shady as they sound. The rooms were clean and equipped. I used to think that outlets were places where brands got rid of last year’s models or unsold stuff. Now I understand that outlets can be a whole business model. I have come across entire shopping malls with just outlets. They are not really cheaper than regular prices and even if they are, the items are manufactured for the outlets. Outlets are such a scam. I ended up buying a lot of electronics, because BestBuy is like heaven. The fuel prices in the US are dirt cheap which is probably why most Texans drive ridiculously large vehicles for no good reason. Back in Sweden, my 4 km 15 min taxi ride from the station to my apartment cost close to my whole 800 km fuel cost in Texas.  ","categories": ["Photography","Travel"],
        "tags": ["Texas","USA"],
        "url": "/texas/",
        "teaser": null
      },{
        "title": "Scraping Instagram and choosing hashtags",
        "excerpt":"If you are an Instagram user, at some point, you care going to be interested in the various metrics such as followers, number of posts by a certain user etc. You might want to compare these metrics between different users or to find out the number of posts with a certain hashtag etc. The casual way to do it is to go the relevant Instagram page and look at the metric and write it down somewhere, and go to next and so on. Clearly this is not ideal strategy if you want to look at a few hundred pages. It would be neat to get this data in an automated manner.     The Instagram API exists to allow automated access to underlying data. The R package instaR is based on this idea. You can go make a developer account and after all the hassle, what you get is only data from your own account. You do not really get information about any other internal metrics or publicly available information. Why publicly available info is not available through the API is perplexing. You can go these pages and write down this information, but it cannot be made available programmatically. One possible explanation is that Instagram has restricted this information because it wants users to buy this information or rely on 3rd party tools (who in turn pays Instagram). Income from ADs, and internal data is not good enough that they have to sell public data as well.   So the only option we are left with is web scraping or crawling or whatever it is called. I used the httr package to get the webpage and then do some string matching using grep and stringr to extract the information I am looking for. I packaged the code into a little R package here.   In this example, I am exploring which hashtags to use for my aerial/drone photographs and how to pick them in a non-random informed manner. I have picked some usernames/accounts in the aerial/drone category with curated user content to collect some data. I have picked usernames that use the same hashtag.   # a vector of valid username/hashtags names &lt;- c(\"abstractaerialart\", \"aerial_aesthetics\", \"all_things_drone\",  \"aworldfromabove\", \"discovererdrone\", \"drnland\", \"drohnengeek\", \"drone_feed\",  \"dronedesire\", \"dronefluence\", \"dronefly\", \"dronelads\", \"dronenature\",  \"droneofficials\", \"dronepals\", \"dronephotooftheday\", \"dronephotosdaily\",  \"dronewise\", \"flyresponsibly\", \"fromwhereidrone\", \"gameofdronez\",  \"guidetodrone\", \"hypepools\", \"igdroneshots\", \"instadroned\", \"mavicimages\",  \"nordicsfromabove\", \"top_drones\", \"topdroneshots\", \"twenty4sevendrones\",  \"worlddroneshots\", \"amzdronepics\")   And then I run the loop below with functions from iscrape to collect the data.   library(iscrape)  len &lt;- length(names) klist &lt;- vector(\"list\", length=len) for(i in 1:len) {   cat(paste0(\"\\nRunning \", i, \" of \", len, \"; \", names[i], \"; \"))   pu &lt;- get_page_user(names[i])   pcount &lt;- get_count_post(pu)   cat(pcount, \" \")   fcount &lt;- get_count_follower(pu)   cat(fcount, \" \")   focount &lt;- get_count_following(pu)   cat(focount, \"; \")    ph &lt;- get_page_hashtag(names[i])   hcount &lt;- get_count_hashtag(ph)   cat(hcount, \";\")    klist[[i]] &lt;- data.frame(name=names[i], posts=pcount, followers=fcount, following=focount,hashtagcounts=hcount, stringsAsFactors=F)    # variation in timing page request   Sys.sleep(sample(1:6, 1, replace=T)) }  dfr &lt;- dplyr::bind_rows(klist)   Now, we have a data frame with user names, post count, follower count, following count and hashtag counts.   &gt; head(dfr)                 name posts followers following hashtagcounts 1 abstractaerialart   179      3955       978          1366 2 aerial_aesthetics   749     47300      5910         12700 3  all_things_drone   642     24500      1943           597 4   aworldfromabove   172      4155        36           495 5   discovererdrone   170     59400      1852          7997 6           drnland   386     18800      1242         11900   The data is plotted out as interactive scatterplot using R package highcharter.   library(highcharter)  hchart(dfr, \"bubble\", hcaes(x=followers, y=hashtagcounts, size=posts)) %&gt;%   hc_xAxis(type=\"logarithmic\", title=list(text=\"Subscribers\"), crosshair=TRUE) %&gt;%   hc_yAxis(type=\"logarithmic\", title=list(text=\"Hastag Count\"), crosshair=TRUE) %&gt;%   hc_chart(zoomType=\"xy\") %&gt;%   hc_tooltip(useHTML=TRUE, positioner=list(list(x=0, y=0)),                formatter = JS(\"function(){                             return('&lt;br&gt;&lt;b&gt;Hashtag: &lt;/b&gt;'+this.point.name+                             '&lt;br&gt;&lt;b&gt;Subscribers: &lt;/b&gt;'+this.x+                             '&lt;br&gt;&lt;b&gt;Hashtag Count: &lt;/b&gt;'+this.y)}\"))                           Scatterplot of followers vs hashtag counts for account names. Size denotes number of posts.                In this case, the interactive plot may be better because it is for interactive use and you don’t have to worry about overplotting. Here, I have plotted the number of followers for an account vs the number of public posts with this username. There is not much data points here (n=32), but we can see some trend showing that accounts with more followers are also hashtagged more. As an account becomes more popular with more followers, people tend to use that account name as hashtag more and more. We can also roughly see that accounts with more followers and more hashtag counts have more posts. This data fits our expectations. My guess for cases where data points deviate from this expectation is due to account holder deleting poorly performing posts.   This plot may help us in picking suitable hashtags. Hashtags with hundreds of thousands of posts can give your post more exposure but quickly gets displaced by incoming posts. Hashtags with few hundred posts may only cater to a niche handful of subscribers, but your post will stay around on the top for days or even weeks. I suppose the reasonable approach is to pick hashtags from across the board to maximise exposure for your content.   One can go further with this idea and gather more data (get more account names from diverse niches) and also categorise accounts into various groups, give them arbitrary rating etc and plot them out.                           Scatterplot of followers vs hashtag counts. Colours denote categories. Size denotes an arbitrary rating. Red line is a reference line for 50,000 posts.                Here, there are lot more accounts and it becomes easier to visualise the linear relationship. I have coloured accounts into categories and the size of points denote my personal rating. Now I can also choose hashtags from category other than aerial that may be relevant to my content.   Of course, this is too simplistic and I don’t even know if this works, it’s just theory. There are numerous other factors which influences the organic reach of a post. I’ve even read that repeatedly using the same hashtags or using too many or too few hashtags are frowned upon by Instagram.  ","categories": ["Code","Research","Web"],
        "tags": ["Data Analysis","Instagram","R","Scientific Graphs","Web Scraping"],
        "url": "/scraping-instagram-choosing-hashtags/",
        "teaser": null
      },{
        "title": "Photo calender using R",
        "excerpt":"I had this idea of using some of my travel photos to create a photo calendar. I would normally go about it using Adobe Photoshop or Adobe Illustrator. But, that would involve a lot of manual work placing dates and days for each month. I would also like to mark some public holidays and friend’s birthdays. So, I wondered if it might be possible to do it with R. After fiddling about with it over the weekend, I managed to make it work. It went better than I expected. And here I am recreating the calendar using some stock photos. All stock photos are royalty-free from Pexels. For the impatient ones, the whole code and images are available at this Github repository. For detailed guide, keep reading.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            The calendar is designed in A3 size format. Twelve photos are selected, labelled 1.jpg to 12.jpg. The photos are first sized to exactly A3 size using an image editing software and placed in a directory labelled ‘images’. A text file is prepared with important dates and description such as holidays, birthdays etc. Another text file is prepared with month and photo caption.   The packages required are loaded.   # load libraries library(ggplot2) # for plotting library(stringr) # for text processing library(Cairo) # for graphics library(extrafont) # for custom fonts library(png) # read png library(jpeg) # read jpeg   I use the font Gidole for the calendar. This was installed on my system and imported using the package ‘extrafont’. I follow the workflow for Windows OS. Mac and linux users might need different commands for configuring font.   # prepare fonts font_import(pattern=\"Gidole\", prompt=FALSE) loadfonts(device=\"win\") windowsFonts(sans=\"Gidole\")   Export image size is set as A3. An extra 6 mm bleed edges is added to aid in trimming the print.   img_bleed &lt;- 6 img_height &lt;- 297+(img_bleed*2) img_width &lt;- 420+(img_bleed*2) img_dpi &lt;- 300 img_units &lt;- \"mm\" export &lt;- \"png\"   The raw calendar dates are prepared.   # PREPARE CALENDAR DATA.FRAME dates &lt;- seq(as.Date(\"2018-01-01\"), as.Date(\"2018-12-31\"), \"days\") df_raw &lt;- data.frame(fulldate=as.character(dates),                    date=as.numeric(strftime(dates, format=\"%d\")),                    month=as.numeric(strftime(dates, format=\"%m\")),                    year=strftime(dates, format=\"%Y\"),                    day=strftime(dates, format=\"%a\"),                    week=strftime(dates, format=\"%W\"),                   stringsAsFactors=F)   Marked dates with description are read in.   # read marked dates and description marked &lt;- read.delim(\"marked.txt\", sep=\"\\t\", header=T, stringsAsFactors=F)&lt;br&gt;   &gt; head(marked)     date                 description           type 1 Jan 01              New Year's Day public holiday 2 Mar 20              Spring Equinox           none 3 Mar 25 Daylight Saving Time Starts           none 4 Mar 30                 Good Friday public holiday 5 Mar 31               Holy Saturday   half holiday 6 Apr 01                  Easter Day public holiday   Multi word descriptions are split to two lines. New line character ‘\\n’ is added to the last space in the description. Character dates is converted to actual R dates.   # add newline to last space in the description marked$description &lt;- str_replace(marked$description, \" [:alnum:]+$\", str_replace(str_extract(marked$description, \" [:alnum:]+$\"), \" \", \"\\n\")) # create R dates marked$fulldate &lt;- as.character(strftime(as.Date(str_replace_all(paste0(marked$date, \" 2018\"), \" \", \"-\"), format=\"%b-%d-%Y\"), format=\"%Y-%m-%d\")) # identify duplicate dates if any marked[which(duplicated(marked$fulldate)), ] # remove character date marked$date &lt;- NULL  &gt; head(marked)                     description           type   fulldate 1              New Year's\\nDay public holiday 2018-01-01 2              Spring\\nEquinox           none 2018-03-20 3 Daylight Saving Time\\nStarts           none 2018-03-25 4                 Good\\nFriday public holiday 2018-03-30 5               Holy\\nSaturday   half holiday 2018-03-31 6                  Easter\\nDay public holiday 2018-04-01   Marked dates are merged with the calendar. Empty descriptions ‘NA’ are marked as ‘not important’. Weekends and public holidays are assigned the font colour red and birthdays are assigned the font colour blue. Start of week is marked. Based on the newline character added previously, descriptions are split into two for plotting on two lines.   # merge calendar and marked dates df_cal &lt;- merge(df_raw, marked, by=\"fulldate\", all.x=T) # mark NA descriptions as not important df_cal$type[is.na(df_cal$type)] &lt;- \"not important\" # assign red colour to public holidays and weekends df_cal$holiday &lt;- ifelse(df_cal$day==\"Sat\"|df_cal$day==\"Sun\"|df_cal$type==\"public holiday\", \"#e41a1c\", \"grey20\") # assign blue colour to birthdays df_cal$holiday[df_cal$type==\"birthday\"] &lt;- \"steelblue\" # mark start of week df_cal$week &lt;- ifelse(df_cal$day==\"Mon\", df_cal$week, NA) # alpha of background shape for week df_cal$weekbgalpha &lt;- ifelse(is.na(df_cal$week), 0, 1) # split description into two lines df_cal$description &lt;- paste0(df_cal$description, \"\\nx\") df_cal$description1 &lt;- sapply(strsplit(df_cal$description, \"\\n\"), \"[[\", 1) df_cal$description2 &lt;- sapply(strsplit(df_cal$description, \"\\n\"), \"[[\", 2) df_cal$description1[df_cal$description1==\"NA\"] &lt;- NA df_cal$description2[df_cal$description2==\"x\"] &lt;- NA  &gt; head(df_cal)     fulldate date month year day week        description           type 1 2018-01-01    1     1 2018 Mon   01 New Year's\\nDay\\nx public holiday 2 2018-01-02    2     1 2018 Tue &lt;NA&gt;              NA\\nx  not important 3 2018-01-03    3     1 2018 Wed &lt;NA&gt;              NA\\nx  not important 4 2018-01-04    4     1 2018 Thu &lt;NA&gt;              NA\\nx  not important 5 2018-01-05    5     1 2018 Fri &lt;NA&gt;              NA\\nx  not important 6 2018-01-06    6     1 2018 Sat &lt;NA&gt;              NA\\nx  not important   holiday weekbgalpha description1 description2 1 #e41a1c           1   New Year's          Day 2  grey20           0         &lt;NA&gt;         &lt;NA&gt; 3  grey20           0         &lt;NA&gt;         &lt;NA&gt; 4  grey20           0         &lt;NA&gt;         &lt;NA&gt; 5  grey20           0         &lt;NA&gt;         &lt;NA&gt; 6 #e41a1c           0         &lt;NA&gt;         &lt;NA&gt;   Image captions and caption colour file is read in.   # prepare image captions cap &lt;- read.delim(\"captions.txt\", header=T, stringsAsFactors=F)  &gt; head(cap)   month               description textcol 1     1           Snowy landscape  grey20 2     2    Snowfall on the leaves  grey20 3     3 Blue waters and mountains  grey20 4     4              Flower macro  grey20 5     5         Lady in the woods  grey20 6     6           Field of barley  grey20   Positions of the various text elements are defined bottom-up. ‘basepos’ is an arbitrary base line position which helps to move all text up or down. ‘markpos1’ defines second line of marked date description, ‘markpos2’ defines first line of marked date description. ‘datepos’ defines the row of date numbers. ‘daypos’ defines the row of day text. ‘weekpos’ defines the row of week numbers. ‘monthpos’ defines the position of the month.   # set text position variables basepos &lt;- 0.35 df_cal$markpos1 &lt;- basepos+0.08 df_cal$markpos2 &lt;- df_cal$markpos1[1]+0.12 df_cal$datepos &lt;- df_cal$markpos2[1]+0.20 df_cal$daypos &lt;- df_cal$datepos[1]+0.25 df_cal$weekpos &lt;- df_cal$daypos[1]+0.25 df_cal$monthpos &lt;- df_cal$weekpos[1]+0.26  &gt; head(df_cal) fulldate date month year day week        description           type 1 2018-01-01    1     1 2018 Mon   01 New Year's\\nDay\\nx public holiday 2 2018-01-02    2     1 2018 Tue &lt;NA&gt;              NA\\nx  not important 3 2018-01-03    3     1 2018 Wed &lt;NA&gt;              NA\\nx  not important 4 2018-01-04    4     1 2018 Thu &lt;NA&gt;              NA\\nx  not important 5 2018-01-05    5     1 2018 Fri &lt;NA&gt;              NA\\nx  not important 6 2018-01-06    6     1 2018 Sat &lt;NA&gt;              NA\\nx  not important   holiday weekbgalpha description1 description2 markpos1 markpos2 1 #e41a1c           1   New Year's          Day     0.43     0.55 2  grey20           0         &lt;NA&gt;         &lt;NA&gt;     0.43     0.55 3  grey20           0         &lt;NA&gt;         &lt;NA&gt;     0.43     0.55 4  grey20           0         &lt;NA&gt;         &lt;NA&gt;     0.43     0.55 5  grey20           0         &lt;NA&gt;         &lt;NA&gt;     0.43     0.55 6 #e41a1c           0         &lt;NA&gt;         &lt;NA&gt;     0.43     0.55   datepos daypos weekpos monthpos 1    0.75      1    1.25     1.51 2    0.75      1    1.25     1.51 3    0.75      1    1.25     1.51 4    0.75      1    1.25     1.51 5    0.75      1    1.25     1.51 6    0.75      1    1.25     1.51   The loop reads the image for each month and converts to a grob. A ggplot object is created and the calendar along with the markings are printed. The ggplot object and the image grob are combined and exported.   # EXPORT LOOP # loop to create and export monthwise levs &lt;- levels(factor(df_cal$month)) for(i in 1:length(levs)) {   fullmonth &lt;- month.name[i]   message(paste0(\"Running \", fullmonth, \" ...\"))    # subset one month   single &lt;- subset(df_cal, df_cal$month==i)    # read background image   #pic &lt;- readPNG(paste0(\"./images/\", i, \".png\"))   pic &lt;- readJPEG(paste0(\"./images/\", i, \".jpg\"))   pic1 &lt;- grid::rasterGrob(pic)   pic1$width &lt;- unit(1, \"npc\")   pic1$height &lt;- unit(1, \"npc\")    # create final image   p &lt;- ggplot(single)+     # background image     annotation_custom(pic1)+     # background rectangle 1     annotate(\"rect\", xmin=-Inf, xmax=Inf, ymin=0, ymax=single$weekpos[1], alpha=1, fill=\"white\")+     # background rectangle 2     annotate(\"rect\", xmin=-Inf, xmax=Inf, ymin=single$weekpos[1], ymax=single$weekpos[1]+0.4, alpha=0.6, fill=\"white\")+     # month title text faded with year     annotate(\"text\", x=1, y=single$monthpos[1], label=paste0(fullmonth, \" 2018\"), size=7, hjust=0, col=\"grey40\", fontface=\"bold\")+     # month title text     annotate(\"text\", x=1, y=single$monthpos[1], label=fullmonth, size=7, hjust=0, col=\"grey20\", fontface=\"bold\")+     # image description text     annotate(\"text\", x=single$date[nrow(single)], y=single$monthpos[1], label=cap$description[i], size=4, hjust=1, col=cap$textcol[i])+     # week background circle     geom_point(aes(x=date, y=weekpos), size=11, colour=\"grey90\", alpha=single$weekbgalpha, shape=20)+     # week text     geom_text(aes(x=date, y=weekpos), label=single$week, size=3, col=\"grey20\", fontface=\"bold\")+     # day text     geom_text(aes(x=date, y=daypos), label=single$day, size=4.8, col=single$holiday)+     # date text     geom_text(aes(x=date, y=datepos), label=single$date, size=7.5, col=single$holiday)+     # important information text line 1     geom_text(aes(x=date, y=markpos1), label=single$description2, size=2.8, col=\"grey20\", vjust=1)+     # important information text line 2     geom_text(aes(x=date, y=markpos2), label=single$description1, size=2.8, col=\"grey20\", vjust=1)+     # limit scales     scale_x_continuous(limits=c(0, max(single$date)+1), expand=c(0, 0.6))+     scale_y_continuous(limits=c(0, 10), expand=c(0, 0))+     labs(x=NULL, y=NULL)+     theme_bw(base_family=\"Gidole\")+     # remove graph elements     theme(plot.background=element_rect(fill=\"transparent\", colour=NA),            plot.margin = margin(c(0, 0, 0, 0)),            axis.title = element_blank(),            axis.text = element_blank(),            axis.ticks = element_blank(),            axis.ticks.length = unit(0, \"pt\"),            axis.line = element_blank(),            panel.grid = element_blank(),            panel.border = element_blank(),            panel.background=element_rect(fill=\"transparent\", colour=NA),            panel.ontop=TRUE)    # export image   if(export==\"png\")   {     png(filename=paste0(single$month[1], \".png\"), height=img_height,            width=img_width, res=img_dpi, units=img_units, family=\"Gidole\",            bg=\"transparent\", type=\"cairo\")     print(p)     dev.off()     message(paste0(\"Exported \", single$month[1], \".png\"))   }    if(export==\"tiff\")   {     tiff(filename=paste0(single$month[1], \".tiff\"), height=img_height,                width=img_width, res=img_dpi, units=img_units, family=\"Gidole\",                compression=\"lzw\", type=\"cairo\", bg=\"transparent\")     print(p)     dev.off()     message(paste0(\"Exported \", single$month[1], \".tiff\"))   }    if(export==\"pdf\")   {     pdf(file=paste0(single$month[1], \".pdf\"), height=round(img_height*0.039, 0),          width=round(img_width*0.039, 0), family=\"Gidole\")     print(p)     dev.off()     message(paste0(\"Exported \", single$month[1], \".pdf\"))   } }   I created a cover page as well. If you are interested in the font, the script font is Bira and the slim sans-serif font is Gidole.                               The images are exported in png format here. But for print, PNG is not an ideal format. Instead, JPEG or pdf might be better. I imported all the PNGs into Adobe Indesign and created a single print quality PDF. You can also set 6 mm of bleed edges in Indesign. I sent the PDF off to print and the results were glorious. Go forth and let your creativity flow. Make yourself a calendar for next year.   The full code along with images are available at this Github repository.  ","categories": ["Code","Craft","Design","Photography"],
        "tags": ["Calendar","Data Visualisation","Design","R"],
        "url": "/photo-calender-using-r/",
        "teaser": null
      },{
        "title": "Snowy Landscapes",
        "excerpt":"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 I had a few days off and took the opportunity to go on a short getaway to snowy winter wonderland. The snow in Uppsala and most of Stockholm and southern Sweden is rubbish. To find some, good quality snow, one needs to drive 5 hours west to Norway. Regions north of Mora and places such as Sälen and Stöten seem to have a decent quality snow. I stayed at an STF hostel called Sälens Vandrarhem, close to Stöten. Nice, cozy place and the owner is a fun dude. At this time of the year, the hostel is literally empty and it was all mine. I drove a bit into Norway to the little town of Ljørdalen. The town centre is basically 2 houses, one of which is a supermarket and the other is a church. The snow in Norway is much nicer. Heavier and more consistent. The landscape is absolutely amazing.I also saw a whole herd of Moose wandering about. Unfortunately, they got away before I could get my camera. This place is excellent for winter photography.  ","categories": ["Photography","Travel"],
        "tags": ["Forest","Norway","Road Trip","Sweden","Winter"],
        "url": "/snowy-landscapes/",
        "teaser": null
      },{
        "title": "Mini terrariums",
        "excerpt":"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         As a little holiday project, I got to working on mini terrariums. I had a few different ideas that didn’t all work together. So I ended up making three different variants. For one, I have some leafy plants and an animal prop. For the second one, I have moss and an toadstools. And finally cactus/dry succulents. I wanted a tiny sign post prop for the cacti but didn’t manage to get one. I bought the ~10cm glass balls from eBay, and the props from toy stores. The glass balls were filled with a small amount of charcoal followed by soil. The exposed soil on the top is covered with a thin layer of fine gravel. in hindsight, these glass were too small to work well. Also finding suitable plants of the right size can be tricky. Propagating cacti can be a bit of a pain.  ","categories": ["Craft","Design"],
        "tags": ["Gardening","Plants"],
        "url": "/mini-terrariums/",
        "teaser": null
      },{
        "title": "Customising ggplot2",
        "excerpt":"Customising a ggplot2 plot using the theme function. In this post, we are going to explore how to adjust various ggplot plot elements. What can be adjusted, what they are called and how they can be adjusted.   First, load the packages that will be used. The extrafont package is used to import custom fonts and is completely optional.   library(ggplot2) library(dplyr) library(RColorBrewer) library(Cairo) library(ggpubr)  library(extrafont) # for custom font loadfonts(device=\"win\")   Base plot   We create a basic sample plot to play around with.I am using the ‘diamonds’ datasets. The dataset is filtered to reduce the number of data points. Note that the pipe operator %&gt;% from the magrittr package is used to connect subsequent commands. The default theme is set to theme_bw. The base_family=\"Gidole\" only need to be set if adjusting the default font.   p &lt;- diamonds %&gt;%   filter(cut==\"Fair\"|cut==\"Good\", color==\"D\"|color==\"E\") %&gt;%   droplevels() %&gt;%   ggplot(aes(carat, price, alpha=color, shape=cut))+   geom_point()+   labs(title=\"Title\", subtitle=\"subtitle\")+   facet_grid(cut~color)+   theme_bw(base_family=\"Gidole\") p                           Base ggplot2 plot                All non-data related plot elements are adjusted through the theme() function. All possible arguments to theme that can be adjusted can be inspected using args(theme). Explanations for these arguments are available through ?theme.   Text elements   First, we explore the text elements in the plot that can be adjusted. All text elements are adjusted through the element_text() function.   args(element_text) function (family = NULL, face = NULL, colour = NULL, size = NULL,           hjust = NULL, vjust = NULL, angle = NULL, lineheight = NULL,           color = NULL, margin = NULL, debug = NULL, inherit.blank = FALSE)   Here, I am setting each text element to a unique colour so that it can be identified on out sample plot.   # change colour of text elements cols &lt;- brewer.pal(7, \"Dark2\")  p1 &lt;- p +   labs(title=\"Text\", subtitle=\"Coloured text elements\")+   theme(axis.title=element_text(color=cols[1]),    axis.text=element_text(color=cols[2]),    plot.title=element_text(color=cols[3]),    plot.subtitle=element_text(color=cols[4]),    legend.text=element_text(color=cols[5]),    legend.title=element_text(color=cols[6]),    strip.text=element_text(color=cols[7]))   Here, I am creating a legend plot to be placed on the right side.   dfr &lt;- data.frame(value=rep(1, 7), label=c(\"axis.title\", \"axis.text\", \"plot.title\", \"plot.subtitle\", \"legend.text\", \"legend.title\", \"strip.text\"), stringsAsFactors=FALSE) %&gt;%   mutate(label=factor(label, levels=c(\"axis.title\", \"axis.text\", \"plot.title\", \"plot.subtitle\", \"legend.text\", \"legend.title\", \"strip.text\")))  q &lt;- ggplot(dfr, aes(x=label, y=value, fill=label))+   geom_bar(stat=\"identity\")+   labs(x=\"\", y=\"\")+   coord_flip()+   scale_fill_manual(values=cols)+   theme_minimal(base_size=20, base_family=\"Gidole\")+   theme(     legend.position=\"none\",      axis.text.x=element_blank(),      axis.ticks=element_blank(),      panel.grid=element_blank())  ggarrange(p1, q, nrow=1, ncol=2, widths=c(0.7, 0.3))                           ggplot2 plot with customised text elements.                Rectangular elements   Now, we explore the rectangle elements. All rectangle elements are adjusted through the element_rect() function.   args(element_rect) function (fill = NULL, colour = NULL, size = NULL, linetype = NULL,color = NULL, inherit.blank = FALSE)  # change fill colour of rectangular elements cols &lt;- brewer.pal(6, \"Pastel1\")  p1 &lt;- p +   labs(title=\"Rect\", subtitle=\"Fill coloured rect elements\")+   theme(plot.background=element_rect(fill=cols[1]),    panel.background=element_rect(fill=cols[2]),    panel.border=element_rect(fill=NA, size=3, color=cols[3]),    legend.background=element_rect(fill=cols[4]),    legend.box.background=element_rect(fill=cols[5]),    strip.background=element_rect(fill=cols[6]))   Now the legend.   dfr &lt;- data.frame(value=rep(1, 6), label=c(\"plot.background\", \"panel.background\", \"panel.border\", \"legend.background\", \"legend.box.background\", \"strip.background\"), stringsAsFactors=FALSE) %&gt;%   mutate(label=factor(label, levels=c(\"plot.background\", \"panel.background\", \"panel.border\", \"legend.background\", \"legend.box.background\", \"strip.background\")))  q &lt;- ggplot(dfr, aes(x=label, y=value, fill=label))+   geom_bar(stat=\"identity\")+   labs(x=\"\", y=\"\")+   coord_flip()+   scale_fill_manual(values=cols)+   theme_minimal(base_size=20, base_family=\"Gidole\")+   theme(     legend.position=\"none\",      axis.text.x=element_blank(),      axis.ticks=element_blank(),      panel.grid=element_blank())  ggarrange(p1, q, nrow=1, ncol=2, widths=c(0.6, 0.4))                           ggplot2 plot with customised rectangular elements.                Line elements   Here we have the line elements. All line elements are adjusted through the element_line() function.   args(element_line) function (colour = NULL, size = NULL, linetype = NULL, lineend = NULL, color = NULL, arrow = NULL, inherit.blank = FALSE)  # change line elements cols &lt;- brewer.pal(4, \"Set1\")  p1 &lt;- p +   labs(title=\"Line\", subtitle=\"Coloured line elements\")+   theme(axis.ticks=element_line(colour=cols[1], size=1),    axis.line=element_line(colour=cols[2], size=1),    panel.grid.major=element_line(colour=cols[3]),    panel.grid.minor=element_line(colour=cols[4]))   Now the legend.   dfr &lt;- data.frame(value=rep(1, 4), label=c(\"axis.ticks\", \"axis.line\", \"panel.grid.major\", \"panel.grid.minor\"), stringsAsFactors=FALSE) %&gt;%   mutate(label=factor(label, levels=c(\"axis.ticks\", \"axis.line\", \"panel.grid.major\", \"panel.grid.minor\")))  q &lt;- ggplot(dfr, aes(x=label, y=value, fill=label))+   geom_bar(stat=\"identity\")+   labs(x=\"\", y=\"\")+   coord_flip()+   scale_fill_manual(values=cols)+   theme_minimal(base_size=20, base_family=\"Gidole\")+   theme(     legend.position=\"none\",      axis.text.x=element_blank(),      axis.ticks=element_blank(),      panel.grid=element_blank())  ggarrange(p1, q, nrow=1, ncol=2, widths=c(0.65, 0.35))                           ggplot2 plot with customised line elements.                Margins and spacings   Margins are set using margin() and spacing is set using grid::unit().   Below shows the effect of increasing margins on the left and right of legend (red), legend.box (blue) and plot (green).   cols &lt;- brewer.pal(4, \"Pastel1\")  p1 &lt;- p + labs(title=\"Margin\", subtitle=\"Default\") p2 &lt;- p + theme(   legend.background=element_rect(fill=cols[1]),    legend.box.background=element_rect(fill=cols[2]),    plot.background=element_rect(fill=cols[3]))+   labs(title=\"Margin\", subtitle=\"Default + Fill colour\")  p3 &lt;- p + theme(   legend.margin=margin(r=20, l=20),    legend.box.margin=margin(r=20, l=20),    plot.margin=margin(r=20, l=20))+   labs(title=\"Margin\", subtitle=\"Increased margin\")  p4 &lt;- p + theme(   legend.background=element_rect(fill=cols[1]),    legend.box.background=element_rect(fill=cols[2]),    plot.background=element_rect(fill=cols[3]),    legend.margin=margin(r=20, l=20),    legend.box.margin=margin(r=20, l=20),    plot.margin=margin(r=20, l=20))+   labs(title=\"Margin\", subtitle=\"Increased margin + Fill colour\")  ggarrange(p, p2, p3, p4, nrow=2, ncol=2)                           ggplot2 plot showing elements with customised margins                Left and right both shows the same effect of adjusting margins. Right is coloured. Below we adjust spacing.   cols &lt;- brewer.pal(4, \"Pastel1\")  p1 &lt;- p + labs(title=\"Spacing\", subtitle=\"Default\")  p2 &lt;- p + theme(   legend.background=element_rect(fill=cols[1]),    legend.box.background=element_rect(fill=cols[2]),    plot.background=element_rect(fill=cols[3]))+   labs(title=\"Spacing\", subtitle=\"Default + Fill colour\")  p3 &lt;- p + theme(   legend.spacing=grid::unit(1, \"cm\"),    legend.box.spacing=grid::unit(1, \"cm\"),    panel.spacing=grid::unit(1.5, \"cm\"))+   labs(title=\"Spacing\", subtitle=\"Increased spacing\")  p4 &lt;- p + theme(   legend.background=element_rect(fill=cols[1]),    legend.box.background=element_rect(fill=cols[2]),    plot.background=element_rect(fill=cols[3]),    legend.spacing=grid::unit(1, \"cm\"),    legend.box.spacing=grid::unit(1, \"cm\"),    panel.spacing=grid::unit(1.5, \"cm\"))+   labs(title=\"Spacing\", subtitle=\"Increased spacing + Fill colour\")  ggarrange(p1, p2, p3, p4, nrow=2, ncol=2)                           ggplot2 plot showing customised spacing.                Position   Some of the plot elements can be moved around and repositioned. For example; legend.position moves the legend to the ‘left’, ‘right’, ‘top’,’bottom’ or ‘none’. legend.justification specifies how the legend is aligned horizontally, ie; ‘left’, ‘center’ etc. legend.direction specifies the layout of the legend, if it is ‘vertical’ or ‘horizontal’.   p1 &lt;- p + theme(legend.position=\"right\")+                 labs(title=\"Legend\", subtitle=\"Right position\") p2 &lt;- p + theme(legend.position=\"left\", legend.direction=\"horizontal\")+                 labs(title=\"Legend\", subtitle=\"Left position Horizontal direction\") p3 &lt;- p + theme(legend.position=\"top\", legend.justification=\"center\")+                 labs(title=\"Legend\", subtitle=\"Top position Center justified\") p4 &lt;- p + theme(legend.position=\"bottom\", legend.justification=\"right\")+                 labs(title=\"Legend\", subtitle=\"Bottom position Right justified\")  ggarrange(p1, p2, p3, p4, nrow=2, ncol=2)                           ggplot2 plot with customised legend position. Top left shows default. Bottom left shows legend on the top. Top right shows legend on the left and horizontal. Bottom right shows legend on the bottom and right justified.                Remove elements   Any of the theme elements can be removed/not displayed by setting that argument to element_blank(). Here, I am creating a minimal looking plot by removing unnecessary elements.   p1 &lt;- p +   labs(title=\"Blank\", subtitle=\"Removing theme elements\")+   theme(panel.border=element_blank(),    panel.grid.major.x=element_blank(),    panel.grid.minor=element_blank(),    axis.ticks=element_blank(),    strip.background=element_blank())  print(p1)                           ggplot2 plot with removed plot elements.                Saving themes   Any changes made to the ggplot2 theme elements can be saved for future use. There are at least two ways to do this.   One approach is simply to save it as a variable and apply it to a ggplot.   custom &lt;- theme(panel.border=element_blank(),    panel.grid.major.x=element_blank(),    panel.grid.minor=element_blank(),    axis.ticks=element_blank(),    strip.background=element_blank())  p + custom   Another approach is to create a new theme function either from scratch or by modifying an existing theme. And then apply it as a theme function.   theme_custom &lt;- function (basesize=14) {     theme_bw(base_size=basesize) %+replace%         theme(panel.border=element_blank(),            panel.grid.major.x=element_blank(),            panel.grid.minor=element_blank(),            axis.ticks=element_blank(),            strip.background=element_blank()) }  p + theme_custom()   Session   devtools::session_info()  Below is the session info showing R version and versions of all loaded packages.  Session info ---------------------------------------------------------------------   setting  value                          version  R version 3.4.3 (2017-11-30)   system   x86_64, mingw32                ui       RStudio (1.1.442)              language (EN)                           collate  English_United Kingdom.1252   tz       Europe/Berlin                  date     2018-05-04                    Packages -------------------------------------------------------------------------   package      * version date       source           assertthat     0.2.0   2017-04-11 CRAN (R 3.4.1)   backports      1.1.2   2017-12-13 CRAN (R 3.4.3)   base         * 3.4.3   2017-12-06 local            bindr          0.1.1   2018-03-13 CRAN (R 3.4.3)   bindrcpp     * 0.2     2017-06-17 CRAN (R 3.4.1)   Cairo        * 1.5-9   2015-09-26 CRAN (R 3.4.0)   cli            1.0.0   2017-11-05 CRAN (R 3.4.3)   colorspace     1.3-2   2016-12-14 CRAN (R 3.4.1)   compiler       3.4.3   2017-12-06 local            cowplot        0.9.2   2017-12-17 CRAN (R 3.4.3)   crayon         1.3.4   2017-09-16 CRAN (R 3.4.3)   datasets     * 3.4.3   2017-12-06 local            devtools       1.13.5  2018-02-18 CRAN (R 3.4.3)   digest         0.6.15  2018-01-28 CRAN (R 3.4.3)   dplyr        * 0.7.4   2017-09-28 CRAN (R 3.4.3)   evaluate       0.10.1  2017-06-24 CRAN (R 3.4.1)   extrafont    * 0.17    2014-12-08 CRAN (R 3.4.0)   extrafontdb    1.0     2012-06-11 CRAN (R 3.4.0)   gdtools      * 0.1.7   2018-02-27 CRAN (R 3.4.3)   ggplot2      * 2.2.1   2016-12-30 CRAN (R 3.4.1)   ggpubr       * 0.1.6   2017-11-14 CRAN (R 3.4.3)   glue           1.2.0   2017-10-29 CRAN (R 3.4.3)   graphics     * 3.4.3   2017-12-06 local            grDevices    * 3.4.3   2017-12-06 local            grid           3.4.3   2017-12-06 local            gridExtra    * 2.3     2017-09-09 CRAN (R 3.4.1)   gtable         0.2.0   2016-02-26 CRAN (R 3.4.1)   htmltools      0.3.6   2017-04-28 CRAN (R 3.4.1)   knitr          1.20    2018-02-20 CRAN (R 3.4.3)   labeling       0.3     2014-08-23 CRAN (R 3.4.0)   lazyeval       0.2.1   2017-10-29 CRAN (R 3.4.3)   magrittr     * 1.5     2014-11-22 CRAN (R 3.4.1)   memoise        1.1.0   2017-04-21 CRAN (R 3.4.1)   methods      * 3.4.3   2017-12-06 local            munsell        0.4.3   2016-02-13 CRAN (R 3.4.1)   pillar         1.2.1   2018-02-27 CRAN (R 3.4.3)   pkgconfig      2.0.1   2017-03-21 CRAN (R 3.4.1)   plyr           1.8.4   2016-06-08 CRAN (R 3.4.1)   pophelper      2.2.6   2018-05-01 local            purrr          0.2.4   2017-10-18 CRAN (R 3.4.3)   R6             2.2.2   2017-06-17 CRAN (R 3.4.1)   RColorBrewer * 1.1-2   2014-12-07 CRAN (R 3.4.0)   Rcpp           0.12.16 2018-03-13 CRAN (R 3.4.3)   reshape2       1.4.3   2017-12-11 CRAN (R 3.4.3)   rlang          0.2.0   2018-02-20 CRAN (R 3.4.3)   rmarkdown      1.9     2018-03-01 CRAN (R 3.4.3)   rprojroot      1.3-2   2018-01-03 CRAN (R 3.4.3)   rsconnect      0.8.8   2018-03-09 CRAN (R 3.4.3)   rstudioapi     0.7     2017-09-07 CRAN (R 3.4.3)   Rttf2pt1       1.3.6   2018-02-22 CRAN (R 3.4.3)   scales         0.5.0   2017-08-24 CRAN (R 3.4.1)   stats        * 3.4.3   2017-12-06 local            stringi        1.1.7   2018-03-12 CRAN (R 3.4.3)   stringr      * 1.3.0   2018-02-19 CRAN (R 3.4.3)   svglite        1.2.1   2017-09-11 CRAN (R 3.4.3)   tibble         1.4.2   2018-01-22 CRAN (R 3.4.3)   tidyr          0.8.0   2018-01-29 CRAN (R 3.4.3)   tools          3.4.3   2017-12-06 local            utf8           1.1.3   2018-01-03 CRAN (R 3.4.3)   utils        * 3.4.3   2017-12-06 local            withr          2.1.1   2017-12-19 CRAN (R 3.4.3)   yaml           2.1.18  2018-03-08 CRAN (R 3.4.3)  ","categories": ["Code","Research"],
        "tags": ["Data Visualisation","ggplot2","R"],
        "url": "/customising-ggplot2/",
        "teaser": null
      },{
        "title": "Calendar plot with ggplot2",
        "excerpt":"So, someone asked me about my summer time availability at work. I realized that my availability was a bit complex and perhaps it was easier to sent a figure/diagram rather than explaining it in 200 words. That’s where I thought of the idea of creating an availability calendar using ggplot2. The idea is to show basically three categories: available, not available and a limited period (where I am not available in person, but can read emails).   First, we prepare the general date range as a data.frame. Character dates are converted to date format. Then day, week, month and date are extracted from that as separate columns for ease of plotting.   # prepare date range dfr &lt;- data.frame(date=seq(as.Date('2018-06-11'), as.Date('2018-08-12'), by=1)) dfr$day &lt;- factor(strftime(dfr$date, format=\"%a\"), levels=rev(c(\"Mon\", \"Tue\", \"Wed\", \"Thu\", \"Fri\", \"Sat\", \"Sun\"))) dfr$week &lt;- factor(strftime(dfr$date, format=\"%V\")) dfr$month &lt;- factor(strftime(dfr$date, format=\"%B\"), levels=c(\"June\", \"July\", \"August\")) dfr$ddate &lt;- factor(strftime(dfr$date, format=\"%d\"))  &gt; head(dfr)         date day week month ddate 1 2018-06-11 Mon   24  June    11 2 2018-06-12 Tue   24  June    12 3 2018-06-13 Wed   24  June    13 4 2018-06-14 Thu   24  June    14 5 2018-06-15 Fri   24  June    15 6 2018-06-16 Sat   24  June    16   Then we add our tracks to define availability into a new column named comment. We first set all comments to “Available” and then overwrite the ranges based on our availability. I have also added weekend as a category so it can be coloured differently.   # add date tracks dfr$comment &lt;- \"Available\" dfr$comment[dfr$date&gt;=as.Date('2018-06-11') &amp; dfr$date&lt;=as.Date('2018-06-20')] &lt;- \"Limited\" dfr$comment[dfr$date&gt;=as.Date('2018-06-21') &amp; dfr$date&lt;=as.Date('2018-06-29')] &lt;- \"Not available\" dfr$comment[dfr$date&gt;=as.Date('2018-07-09') &amp; dfr$date&lt;=as.Date('2018-07-13')] &lt;- \"Not available\" dfr$comment[dfr$date&gt;=as.Date('2018-08-06') &amp; dfr$date&lt;=as.Date('2018-08-10')] &lt;- \"Not available\" dfr$comment[dfr$day==\"Sat\" | dfr$day==\"Sun\"] &lt;- \"Weekend\"   Then I convert the comment column into a factor with custom level order. This order determines the order in which the labels will be shown in the plot legend.   dfr$comment &lt;- factor(dfr$comment, levels=c(\"Available\", \"Limited\", \"Not available\", \"Weekend\"))   Finally, the plotting. Columns week and day are used to define x and y axes. geom_tile() creates the blocks and colours for each day. facet_grid()  is used to separate months. scale_fill_manual() is used to assign custom colours. theme_bw() and theme() are used for fine adjustment of plot layout.   # plot p &lt;- ggplot(dfr, aes(x=week, y=day))+  geom_tile(aes(fill=comment))+  geom_text(aes(label=ddate))+  scale_fill_manual(values=c(\"#8dd3c7\", \"#ffffb3\", \"#fb8072\", \"#d3d3d3\"))+  facet_grid(~month, scales=\"free\", space=\"free\")+  labs(x=\"Week\", y=\"\")+  theme_bw(base_size=10)+  theme(legend.title=element_blank(),         panel.grid=element_blank(),         panel.border=element_blank(),         axis.ticks=element_blank(),         strip.background=element_blank(),         legend.position=\"top\",         legend.justification=\"right\",         legend.direction=\"horizontal\",         legend.key.size=unit(0.3, \"cm\"),         legend.spacing.x=unit(0.2, \"cm\"))   And then save the plot.   ggsave(\"calendar.png\", p, height=10, width=14, units=\"cm\", dpi=300, type=\"cairo\")                           Availability calendar plot in R using ggplot2.                And we have our finished calendar. The custom font I used in the plot is called Gidole. To use custom fonts, import the font using library extrafont . Then in theme_bw() , set base_family=\"Gidole\" . Also in geom_text() , set family=\"Gidole\" .  ","categories": ["Code"],
        "tags": ["Calendar","Data Visualisation","ggplot2","R"],
        "url": "/calendar-plot-with-ggplot2/",
        "teaser": null
      },{
        "title": "Calendar plot shiny app and dynamic UI",
        "excerpt":"In my previous post, I created a calendar planner plot using ggplot2. I wanted to take it further and create a shiny app for it. This was more of a shiny learning exercise rather than a purpose-driven app. Specifically, creating dynamic user inputs and retrieving data from these inputs.   The app looks like this:                           App preview                So I had my core plotting code from the previous post, and then I created user interfaces around it. The layout is simple; there is a panel on the left with input widgets and the panel on the right displays the plot. The app takes in user input for a date span (start and end duration) and a data.frame is created expanding that span to a sequence of dates.   I wanted the start date duration and end date duration widgets to be side-by-side in one row, and therefore, I have used custom divs. For date input, I have set the default start date to be today’s date and end date to be 30 days from today.   div(class=\"row\",     div(class=\"col-md-6\", dateInput(\"in_duration_date_start\", \"From\", value=format(as.Date(Sys.time(), \"%Y-%m-%d\", tz=\"Europe/Stockholm\"), \"%Y-%m-%d\"))),     div(class=\"col-md-6\", dateInput(\"in_duration_date_end\",\"To\", value=format(as.Date(Sys.time(), \"%Y-%m-%d\", tz=\"Europe/Stockholm\")+30, \"%Y-%m-%d\"))) )   Then the user adds more date spans to define various activities. I call these tracks. Rather than setting up a pre-defined number of tracks, I wanted to let the user choose the number of tracks. And each track will have some track parameters. I have defined a track name, start date, end date and a colour for the track. This called for a bit more complex code. The track inputs are dynamically created in a loop depending on the number of tracks set by the user.   The code that generates the track input user-interface (UI) dynamically looks like this:   output$tracks &lt;- renderUI({   num_tracks &lt;- as.integer(input$in_num_tracks)    # create date intervals   dseq &lt;- seq(as.Date(input$in_duration_date_start), as.Date(input$in_duration_date_end), by=1)   r1 &lt;- unique(as.character(cut(dseq, breaks=num_tracks+1)))    lapply(1:num_tracks, function(i) {      div(class=\"row\",         div(class=\"col-xs-3\", style=list(\"padding-right: 3px;\"), textInput(paste0(\"in_track_name_\", i), label=\"Name\", value=paste0(\"Text\", i), placeholder=\"Available\")),         div(class=\"col-xs-3\", style=list(\"padding-right: 3px; padding-left: 3px;\"),  dateInput(paste0(\"in_track_date_start_\", i), label=\"From\", value=as.Date(r1[i], \"%Y-%m-%d\"))),         div(class=\"col-xs-3\", style=list(\"padding-right: 3px; padding-left: 3px;\"), dateInput(paste0(\"in_track_date_end_\", i), label=\"To\", value=as.Date(r1[i+1], \"%Y-%m-%d\")-1)),          div(class=\"col-xs-3\", style=list(\"padding-left: 3px;\"), colourpicker::colourInput(paste0(\"in_track_colour_\", i), label=\"Colour\", palette=\"limited\", allowedCols=cols, value=cols[i]))     )   }) })   This looks like this in action:                               Based on the number of input tracks, the date duration span is split into equal intervals by default. The start and end dates of the intervals are used as the default spans for track dates. Default colours are also selected by looping through a pre-defined colour vector.   A core function takes in all the plot related input to create the plot and returns a ggplot2 object. This function is called by the renderImage() function that creates the preview plot and also by the download plot handler. Inside the core function, a loop is also used to read the input track parameters back. The relevant part of the code looks like this:   num_tracks &lt;- input$in_num_tracks track_cols &lt;- vector(length=num_tracks) tracks &lt;- vector(length=num_tracks) for(i in 1:num_tracks) {   temp_start_date &lt;- as.Date(as.character(eval(parse(text=paste0(\"input$in_track_date_start_\", i)))))   temp_end_date &lt;- as.Date(as.character(eval(parse(text=paste0(\"input$in_track_date_end_\", i)))))   temp_track_name &lt;- as.character(eval(parse(text=paste0(\"input$in_track_name_\", i))))   temp_track_col &lt;- as.character(eval(parse(text=paste0(\"input$in_track_colour_\", i)))) }   I have only focused on the exciting bit of code that is new to me. The full code is available at this GitHub repo. A demo of this app is available at this shinyapps.io link. This link may not always be active.  ","categories": ["Code"],
        "tags": ["Calendar","Data Visualisation","ggplot2","R","Shiny App"],
        "url": "/calendar-plot-shiny-app-and-dynamic-ui/",
        "teaser": null
      },{
        "title": "Using google maps for aerial photography",
        "excerpt":"If you are into aerial photography, especially using a drone, then you know that it’s not easy to imagine how the world looks like from above. Composition from the air adds an extra dimension to usual level of photo composition difficulty. The most obvious choice would be to find a high point (hill, building etc) to get a sense of the landscape. When that is not good enough, we have Google Earth. Now for those of you who were paying attention to google maps, you might have noticed that google maps is now available in 3D. At least in some areas and major cities. For those areas, that are available, the level of detail is stunning. The views are far superior than the blocky 3D seen previously in Google Earth. I mean, just look at this:                           Aerial view in Google maps 3D looks like an open-world game.                This is so impressive. It looks like an open-world game. Every tree is rendered independently and probably has several thousand polygons. Now with 3D control, I can survey the area I want to photograph and compose my scene just the way I want. To show you some examples, below are some of the scenes composed in google maps and then actually photographed using a drone.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Examples showing google maps preview and actual drone photograph.          This is brilliant because I don’t need to use up a whole battery on my drone just to find a composition. The limitation is that this detailed 3D model is not available everywhere and also does not reflect the current lighting and weather conditions. It would also be great if google maps showed us the height of the viewpoint/camera when in 3D. Another useful additional would be to somehow incorporate SunCalc into this to get the position and trajectory of the sun.  ","categories": ["Photography","Travel"],
        "tags": ["Drone","Uppsala","Sweden"],
        "url": "/using-google-maps-for-aerial-photography/",
        "teaser": null
      },{
        "title": "Fast LD computation from VCF files using Tomahawk",
        "excerpt":"Install tomahawk from here. Following commands are run on the Linux command-line.   Thinning VCF   For LD computation, pairwise comparisons are performed between all SNPs, therefore it is essential to thin down the number of SNPs. My dataset has about 18 million SNPs. My aim is to bring it down to 20,000 or so. Use vcftools to thin down a VCF file.   vcftools --vcf snp.vcf --recode --recode-INFO-all --thin 50000 --out \"snp-thin\"  VCFtools - 0.1.15 (C) Adam Auton and Anthony Marcketta 2009  Parameters as interpreted:         --vcf snp.vcf         --recode-INFO-all         --thin 50000         --out snp-thin         --recode  After filtering, kept 64 out of 64 Individuals Outputting VCF file... After filtering, kept 26205 out of a possible 18793942 Sites Run Time = 135.00 seconds  mv \"snp-thin.recode.vcf\" \"snp-thin.vcf\"   The argument --thin 50000 means to keep 1 SNP every 50,000 bases. Depending on the number of SNPs in the VCF file, you want to increase or decrease this. I am aiming for about 20,000 or so SNPs in the thinned file.   VCF to BCF   Convert VCF to BCF.   bcftools view snp.vcf -O b -o snp.bcf   Tomahawk   Instructions on using tomahawk is detailed here. Convert BCF to the tomahawk format. Import filters filter out sites with missingness &gt; 20% and HW p-value &lt; 0.001.   tomahawk import -i snp.bcf -o snp -m 0.2 -h 0.001   Tomahawk is used to compute LD.   Arguments: -u (unphased), -d (progress), -f (fast computation), -i (input file), -o (output file name prefix), -r (min R2 cut-off). See tomahawk help for more descriptions.   tomahawk calc -udfi snp.twk -o snp -a 5 -r 0.1 -P 0.1 -C 1 -t 1   The .two output is binary and needs to be exported to a text file. The -I arguments restricts the output to a specified chromosome. So, -I chr1 (UCSC) of -I 1 (Ensembl) exports only chromosome 1. It is probably better to have separate files for each chromosome as the files are a more managable size to read into R. The LD metrics are computed between SNPs within chromosomes anyway.   tomahawk view -i snp.two &gt; snp.ld tomahawk view -i snp.two -I 1 &gt; snp-1.ld   Plotting   Now we move into R. First we load some libraries and define some plotting functions to be used later.   # load libraries library(dplyr) library(ggplot2)  # define plotting functions #' @title plotPairwiseLD #' @description Plots R2 heatmap across the chromosome (like Haploview) #' @param dfr A data.frame with minimum CHROM_A, POS_A, CHROM_B, POS_B and R2. #' An output from tomahawk works. #' @param chr A chromosome name. #' @param xlim A two number vector specifying min and max x-axis limits. Any one or both can be defaulted by specifying NA. #' @param ylim A two number vector specifying min and max y-axis limits. Any one or both can be defaulted by specifying NA. #' @param minr2 A value between 0 and 1. All SNPs with R2 value below this #' value is excluded from plot. #' plotPairwiseLD &lt;- function(dfr, chr, xlim = c(NA, NA), ylim = c(NA, NA), minr2) {   if (missing(dfr)) stop(\"Input data.frame 'dfr' missing.\")    if (!missing(chr)) {     ld &lt;- filter(ld, CHROM_A == get(\"chr\") &amp; CHROM_B == get(\"chr\"))   }   ld &lt;- filter(ld, POS_A &lt; POS_B)    if (!missing(minr2)) {     ld &lt;- filter(ld, R2 &gt; get(\"minr2\"))   }    ld &lt;- ld %&gt;% arrange(R2)    ld$x &lt;- ld$POS_A + ((ld$POS_B - ld$POS_A) / 2)   ld$y &lt;- ld$POS_B - ld$POS_A   ld$r2c &lt;- cut(ld$R2, breaks = seq(0, 1, 0.2), labels = c(     \"0-0 - 0.2\", \"0.2 - 0.4\",     \"0.4 - 0.6\", \"0.6 - 0.8\",     \"0.8 - 1.0\"   ))   ld$r2c &lt;- factor(ld$r2c, levels = rev(c(     \"0-0 - 0.2\", \"0.2 - 0.4\",     \"0.4 - 0.6\", \"0.6 - 0.8\",     \"0.8 - 1.0\"   )))    ggplot(ld, aes(x = x, y = y, col = r2c)) +     geom_point(shape = 20, size = 0.1, alpha = 0.8) +     scale_color_manual(values = c(\"#ca0020\", \"#f4a582\", \"#d1e5f0\", \"#67a9cf\", \"#2166ac\")) +     scale_x_continuous(limits = xlim) +     scale_y_continuous(limits = ylim) +     guides(colour = guide_legend(title = \"R2\", override.aes = list(shape = 20, size = 8))) +     labs(x = \"Chromosome (Bases)\", y = \"\") +     theme_bw(base_size = 14) +     theme(       panel.border = element_blank(),       axis.ticks = element_blank()     ) %&gt;%     return() }   #' @title plotDecayLD #' @description Plots R2 heatmap across the chromosome (like Haploview) #' @param dfr A data.frame with minimum CHROM_A, POS_A, CHROM_B, POS_B and R2. #' An output from tomahawk works. #' @param chr A chromosome name. #' @param xlim A two number vector specifying min and max x-axis limits. Any one or both can be defaulted by specifying NA. #' @param ylim A two number vector specifying min and max y-axis limits.  Any one or both can be defaulted by specifying NA. #' @param avgwin An integer specifying window size. Mean R2 is computed within windows. #' @param minr2 A value between 0 and 1. All SNPs with R2 value below this #' value is excluded from plot. #' plotDecayLD &lt;- function(dfr, chr, xlim = c(NA, NA), ylim = c(NA, NA), avgwin = 0, minr2) {   if (missing(dfr)) stop(\"Input data.frame 'dfr' missing.\")    if (!missing(chr)) {     ld &lt;- filter(ld, CHROM_A == get(\"chr\") &amp; CHROM_B == get(\"chr\"))   }   ld &lt;- filter(ld, POS_A &lt; POS_B)    if (!missing(minr2)) {     ld &lt;- filter(ld, R2 &gt; get(\"minr2\"))   }    ld &lt;- ld %&gt;% arrange(R2)    ld$dist &lt;- ld$POS_B - ld$POS_A    if (avgwin &gt; 0) {     ld$distc &lt;- cut(ld$dist, breaks = seq(from = min(ld$dist), to = max(ld$dist), by = avgwin))     ld &lt;- ld %&gt;%       group_by(distc) %&gt;%       summarise(dist = mean(dist), R2 = mean(R2))   }    ggplot(ld, aes(x = dist, y = R2)) +     geom_point(shape = 20, size = 0.15, alpha = 0.7) +     geom_smooth() +     scale_x_continuous(limits = xlim) +     scale_y_continuous(limits = ylim) +     labs(x = \"Distance (Bases)\", y = expression(LD ~ (r^{       2     }))) +     theme_bw(base_size = 14) +     theme(       panel.border = element_blank(),       axis.ticks = element_blank()     ) %&gt;%     return() }   Then we read the file.   ld &lt;- read.delim(\"snp.ld\", sep=\"\\t\", comment.char=\"#\")   The output looks like below:   &gt; head(ld)   FLAG CHROM_A   POS_A CHROM_B   POS_B   REF_REF      REF_ALT   ALT_REF 1   10       1     359       1  100777 103.00000 8.526513e-14  5.000000 2   10       1  100777       1     359 103.00000 8.526513e-14  5.000000 3    2       1     359       1  452615  97.48765 5.512354e+00 10.512354 4    2       1  452615       1     359  97.48765 5.512354e+00 10.512354 5    6       1     359       1 1454410  49.07769 5.392231e+01  1.922308 6    6       1 1454410       1     359  49.07769 5.392231e+01  1.922308     ALT_ALT          D    Dprime         R        R2            P ChiSqModel 1 20.00000 0.12573242 1.0000000 0.8734775 0.7629629 4.440203e-19   2.296495 2 20.00000 0.12573242 1.0000000 0.8734775 0.7629629 4.440203e-19   2.296495 3 14.48765 0.08266716 0.6574848 0.5742982 0.3298184 5.573234e-08   1.993300 4 14.48765 0.08266716 0.6574848 0.5742982 0.3298184 5.573234e-08   1.993300 5 23.07769 0.06280179 0.8070154 0.3235735 0.1046998 1.876711e-04   7.703488 6 23.07769 0.06280179 0.8070154 0.3235735 0.1046998 1.876711e-04   7.703488   ChiSqTable 1   97.65926 2   97.65926 3   42.21675 4   42.21675 5   13.40157 6   13.40157   Plot pairwise LD plot.   plotPairwiseLD(ld)                           Pairwise LD plot across a chromosome.                plotPairwiseLD(ld, minr2=0.2)                           Pairwise LD plot with a min R2 limit of 0.2.                plotPairwiseLD(ld, minr2=0.2, ylim=c(0, 10^7))                           Pairwise LD plot with a min R2 limit of 0.2 and y-axis max limit set to 10^7.                plotDecayLD(ld, minr2=0.25)                           LD decay plot with a min R2 limit of 0.25.                plotDecayLD(ld, minr2=0.25, avgwin=100000)                           LD decay plot with a min R2 limit of 0.25 and R2 values averaged within 10 Kb windows.                ","categories": ["Code","Research"],
        "tags": ["Genetics","Genomics","Linkage Disequilibirium","NGS","Population Genetics","R","SNP"],
        "url": "/fast-ld-computation-from-vcf-files-using-tomahawk/",
        "teaser": null
      },{
        "title": "Beautiful circos plots in R",
        "excerpt":"Circos plots are popular for displaying huge amounts of data in a relatively small visual space. This is especially relevant for genomic data. This post explores creating Circos-style genomic data plots in R using R package circlize.   The first step is plan your figure. How many genomes to include? Or perhaps, it’s one genome with multiple chromosomes. What sort of information do you want to include? In my case, I have three bacterial genomes for which I would like to show the sequencing depth, GC content, locations of important genes and finally rearrangements between the three genomes. You need to have all this data prepared first. Unfortunately I am unable to provide the raw files, but I can give an idea of the file formats.   Data   My three genomes are named pb_501_001, pb_501_002 and pb_501_003.   Coverage was computed on BAM files using bedtools genomecov. The data looked like this:   pb_501_001 1 24 pb_501_001 2 58 pb_501_001 3 62 ...   with columns: sample, base and depth. This was read into R and converted to median coverage within a window size of 1000 bases creating a BED format. This is file named cov later in the code.   chr        start end  value1 pb_501_001 1     1000 145 pb_501_001 1001  2000 154 pb_501_001 2001  3000 158 ...   The GC content was computed using bedtools nuc. The data is a typical BED format when read into R looks like below. This is the file named gc later in the code.   chr        start end  value1 pb_501_001 1     1000 145.0 pb_501_001 1001  2000 154.0 pb_501_001 2001  3000 158.0 ...   The gene annotation file looks like below. This file is called ann later in the code.   chr        start  end    value1 gene pb_501_001 325210 326376 1      yxeP pb_501_001 397301 397795 1      sorB pb_501_001 399088 400035 1      sorC ...   The value1 is a random value (to maintain bed format, the actual value is not used).   The rearrangements can be obtained from Mauve or Nucmer (Mummer). The data is for pairs of genomes/chromosomes and it looks like:   pb_501_001 1     59986   pb_501_002 1       62503 pb_501_001 60196 1183032 pb_501_002 4682296 3560259 ...   with columns: name of genome/chr 1, start position in 1, end position in 1, name of genome/chr 2, start position in 2, end position in 2.   This is split up into two files which I call nuc1 and nuc2 later in the code.   nuc1   pb_501_001 1       59986 pb_501_001 60196   1183032 pb_501_001 1182899 4708778 ...   nuc2   pb_501_002 1       62503 pb_501_002 4682296 3560259 pb_501_002 67566   3555254 ...   Plotting   We are now ready to start plotting. We will first look at the default code for each step and then also explore the customised code . Plots are showed side by side to compare default and customised results.   Start by loading the library and initialising the layout.   library(circlize) circos.clear() circos.initialize(factors=c(\"pb_501_001\", \"pb_501_002\", \"pb_501_003\"),  xlim=matrix(c(rep(0, 3), ref$V2), ncol=2))   The xlim matrix defines the start and stop for each genome/chr. Essentially the genome or chr sizes. The matrix looks like this:   &gt; matrix(c(rep(0, 3), ref$V2), ncol=2)         [,1] [,2] [1,] 0    5167156 [2,] 0    5138029 [3,] 0    5080212   Genome   At this point, the plotting window is still blank because nothing as actually been drawn. Then we create the genome backbone.   # genomes circos.track(ylim=c(0, 1), panel.fun=function(x, y) { chr=CELL_META$sector.index xlim=CELL_META$xlim ylim=CELL_META$ylim circos.text(mean(xlim), mean(ylim), chr) })   Now, we can see the plot showing the three genomes and their relative lengths. They are labelled using the circos.text() command. We can customise this by first setting some custom defaults using circos.par(). gap.degree is increased to increase gap width so as to fit in our y-axis labels later. cell.padding is set to zero to reduce empty space around the plots. Then, we resize the panel height, adjust text size and background colour.   circos.clear() col_text &lt;- \"grey40\" circos.par(\"track.height\"=0.8, gap.degree=5, cell.padding=c(0, 0, 0, 0)) circos.initialize(factors=ref$V1, xlim=matrix(c(rep(0, 3), ref$V2), ncol=2))  # genomes circos.track(ylim=c(0, 1), panel.fun=function(x, y) { chr=CELL_META$sector.index xlim=CELL_META$xlim ylim=CELL_META$ylim circos.text(mean(xlim), mean(ylim), chr, cex=0.5, col=col_text,  facing=\"bending.inside\", niceFacing=TRUE) }, bg.col=\"grey90\", bg.border=F, track.height=0.06)   Now we can compare how our raw and custom plots look like.                           Genome backbone seen using default setting (left) and customised settings (right).                Let’s add an x-axis to the plot:   # genomes x axis circos.track(track.index = get.current.track.index(), panel.fun = function(x, y) { circos.axis(h=\"top\") })   We can see the x-axis added (see below). We are going to add custom breaks, change label colour, size, direction and thickness of the axis line.   # genomes x axis brk &lt;- c(0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5)*10^6 circos.track(track.index = get.current.track.index(), panel.fun=function(x, y) { circos.axis(h=\"top\", major.at=brk, labels=round(brk/10^6, 1), labels.cex=0.4,  col=col_text, labels.col=col_text, lwd=0.7, labels.facing=\"clockwise\") }, bg.border=F)                           Genome backbone with x-axis seen using default setting (left) and customised settings (right).                Coverage   Let’s add the coverage information and a y-axis for the coverage:   # coverage circos.genomicTrack(data=cov, panel.fun=function(region, value, ...) { circos.genomicLines(region, value) }) # coverage y axis circos.yaxis()   The coverage along the genome is displayed as a line. The y-axis is used to show a reference for the values that are plotted. We will change properties of the line such as colour and thickness. We add extra segments to act as gridlines. And we remove all y-axis elements except the text.   # coverage circos.genomicTrack(data=cov, panel.fun=function(region, value, ...) { circos.genomicLines(region, value, type=\"l\", col=\"grey50\", lwd=0.6) circos.segments(x0=0, x1=max(ref$V2), y0=100, y1=100, lwd=0.6, lty=\"11\", col=\"grey90\") circos.segments(x0=0, x1=max(ref$V2), y0=300, y1=300, lwd=0.6, lty=\"11\", col=\"grey90\") #circos.segments(x0=0, x1=max(ref$V2), y0=500, y1=500, lwd=0.6, lty=\"11\", col=\"grey90\") }, track.height=0.08, bg.border=F) circos.yaxis(at=c(100, 300), labels.cex=0.25, lwd=0, tick.length=0, labels.col=col_text, col=\"#FFFFFF\")                           Circlize plot showing coverage using default setting (left) and customised settings (right).                GC content   Let’s add GC content and it’s y-axis:   # gc content circos.track(factors=gc$chr, x=gc$x, y=gc$value, panel.fun=function(x, y) { circos.lines(x, y) }) # gc y axis circos.yaxis()   Similar to the previous step, GC content is displayed as lines. The customisation is similar to the previous step. The gridline positions and values have obviously changed.   # gc content circos.track(factors=gc$chr, x=gc$x, y=gc$value, panel.fun=function(x, y) { circos.lines(x, y, col=\"grey50\", lwd=0.6) circos.segments(x0=0, x1=max(ref$V2), y0=0.3, y1=0.3, lwd=0.6, lty=\"11\", col=\"grey90\") circos.segments(x0=0, x1=max(ref$V2), y0=0.5, y1=0.5, lwd=0.6, lty=\"11\", col=\"grey90\") circos.segments(x0=0, x1=max(ref$V2), y0=0.7, y1=0.7, lwd=0.6, lty=\"11\", col=\"grey90\") }, ylim=range(gc$value), track.height=0.08, bg.border=F) # gc y axis circos.yaxis(at=c(0.3, 0.5, 0.7), labels.cex=0.25, lwd=0, tick.length=0, labels.col=col_text, col=\"#FFFFFF\")                           Circlize plot showing GC content graph using default setting (left) and customised settings (right).                Gene labels   Then we add gene labels:   # gene labels circos.genomicLabels(ann, labels.column=5)   The gene labels are added along with connection lines to indicate their exact position in the genome. This is too crowded without some manual customisation. We adjust the label size, colour and the length, thickness and colour of the connection lines. The track height is set to a custom value.   # gene labels circos.genomicLabels(ann, labels.column=5, cex=0.25, col=col_text, line_lwd=0.5, line_col=\"grey80\",  side=\"inside\", connection_height=0.05, labels_height=0.04)                           Circlize plot showing gene positions and gene labels using default setting (left) and customised settings (right).                Rearrangements   Finally, we add the rearrangements using circos.genomicLink() function.   # rearrangements circos.genomicLink(nuc1, nuc2)   If you are running these steps interactively, this may not be visible as there is no space. But, if you run this inside a png() device, you can at least see it. Since there is no transparency, overlapping ribbons are not visible. We will assign green colour to rearrangements in the same direction and red colour to inversions. And we set 0.4 opacity for better visibility of overlapping ribbons.   # rearrangements rcols &lt;- scales::alpha(ifelse(sign(nuc1$start - nuc1$end) != sign(nuc2$start - nuc2$end), \"#f46d43\", \"#66c2a5\"), alpha=0.4) circos.genomicLink(nuc1, nuc2, col=rcols, border=NA)   And here we have the complete plot. Track heights for every track has been reduced to leave space for the rearrangement link figure in the middle.                           Circlize plot using default setting (left) and customised settings (right).                For the purpose of captioning, I have further added some numbers in an image editing software. This will be useful to describe each part of the figure.                           Circos plot with custom labels added for captions.                ","categories": ["Code","Research"],
        "tags": ["Data Visualisation","Genomics","NGS","R"],
        "url": "/beautiful-circos-plots-in-r/",
        "teaser": null
      },{
        "title": "A guide to elegant tiled heatmaps in R [2019]",
        "excerpt":"This is an update to an older post from 2015 on the same topic. This covers the exact same thing but using the latest R packages and coding style using the pipes (%&gt;% ) and tidyverse packages.   This was inspired by the disease incidence rate in the US featured on the Wall Street Journal. The disease incidence dataset was originally used in this article in the New England Journal of Medicine. In this post, I am using the measles level 1 incidence (cases per 100,000 people) dataset obtained as a .csv file from Project Tycho. Download the .csv file here.   In this post, we will look into creating a neat, clean and elegant heatmap in R. No clustering, no dendrograms, no trace  lines, no bullshit. We will go through some basic data cleanup, reformatting and finally plotting. We go through this step by step. For the whole code with minimal explanations, scroll to the bottom of the page.   I am running R version 3.5.2 64-bit on Ubuntu 18.04 64-bit. Packages in use are ggplot2 (3.1.0), dplyr (0.7.8), tidyr (0.8.2), stringr (1.3.1) and for base plots I am using gplots (3.0.1.1) and plotrix (3.7-4). Install the necessary packages if not already installed and load them.   # install packages # install.packages(pkgs = c(\"ggplot2\",\"dplyr\",\"tidyr\",\"stringr\",\"gplots\",\"plotrix\"),dependencies = T)  # load packages library(ggplot2) # ggplot() for plotting library(dplyr) # data reformatting library(tidyr) # data reformatting library(stringr) # string manipulation  1. Data preparation   Read in the .csv file and inspect the data. The first two rows with non-table data is skipped.   # read csv file m &lt;- read.csv(\"measles_lev1.csv\",header=T,stringsAsFactors=F,skip=2)  # inspect data head(m) str(m) table(m$YEAR) table(m$WEEK)   The head() function shows us the header names and the first 6 rows of the data. The str() function shows that YEAR and WEEK columns are stored as integers and the incidences as characters. The incidences although numeric have been read in as character because missing values are coded as “-“.  The table() function is used to check for any missing years or weeks. The data is currently stored in so called ‘wide’ format which we will convert to ‘long’ format. The ggplot2 plotting package prefers long format. The wide format is shown below.   &gt; head(m) YEAR WEEK ALABAMA ALASKA ARIZONA ARKANSAS CALIFORNIA COLORADO 1 1928    1    3.67      -    1.90     4.11       1.38     8.38 2 1928    2    6.25      -    6.40     9.91       1.80     6.02 3 1928    3    7.95      -    4.50    11.15       1.31     2.86 4 1928    4   12.58      -    1.90    13.75       1.87    13.71 5 1928    5    8.03      -    0.47    20.79       2.38     5.13 6 1928    6    7.27      -    6.40    26.58       2.79     8.09   The YEARand WEEK variables are kept as is and all incidence values are collapsed into a variable and value column. The column names are changed to lower case for convenience. The year and week variables are converted to factors and value variable is converted to a number.   m2 &lt;- m %&gt;%   # convert data to long format   gather(key=\"state\", value=\"value\", -YEAR, -WEEK) %&gt;%   # rename columns   setNames(c(\"year\", \"week\", \"state\", \"value\")) %&gt;%   # convert year to factor   mutate(year=factor(year)) %&gt;%   # convert week to factor   mutate(week=factor(week)) %&gt;%   # convert value to numeric (also converts '-' to NA, gives a warning)   mutate(value=as.numeric(value))   The output in long format is shown below.   &gt; head(m2)   year week   state value 1 1928    1 Alabama  3.67 2 1928    2 Alabama  6.25 3 1928    3 Alabama  7.95 4 1928    4 Alabama 12.58 5 1928    5 Alabama  8.03 6 1928    6 Alabama  7.27   The state variable is currently in all caps and multi-word states have dot separators. I prefer to have them in Title Case and separated by space to be shown on the plot later. A custom function is used to change the states into Title Case. Multi-word states are split at the dot separator and each word is changed into Title Case using the function str_to_title() and they are pasted back together.   # removes . and change states to title case using custom function fn_tc &lt;- function(x) paste(str_to_title(unlist(strsplit(x, \"[.]\"))), collapse=\" \") m2$state &lt;- sapply(m2$state, fn_tc)   Now, I would like to plot the heatmap with the year on the x-axis and state on the y-axis, which means we have to deal with the week variable in some way. We will sum all the incidences from all weeks for each year and discard the week variable. The dplyr compliant way to do this is to use group_by() followed by summarise() using a function.   The way sum() handles NAs is a bit strange. By default, it returns NA if one or more elements in the input vector is NA. If we set argument na.rm=TRUE, then NAs are removed and the remaining numbers are summed. But if all the elements are NA, the sum is returned as zero. This is weird and undesirable in this situation. Therefore, I have a custom sum function named na_sum() to remove NAs and return the remaining sum or return NA if all elements are NA. We then use this custom function inside summarise() function to summarise the data by year and state while getting rid of week.   # custom sum function returns NA when all values in set are NA, # in a set mixed with NAs, NAs are removed and remaining summed. na_sum &lt;- function(x) {   if(all(is.na(x))) val &lt;- sum(x, na.rm=F)   if(!all(is.na(x))) val &lt;- sum(x, na.rm=T)   return(val) }  # sum incidences for all weeks into one year m3 &lt;- m2 %&gt;%   group_by(year, state) %&gt;%   summarise(count=na_sum(value)) %&gt;%   as.data.frame()   Now our data looks like this without the week variable. The values are summed by year for each state to create a new variable count.   &gt; head(m3)    year      state  count 1 1928    Alabama 334.99 2 1928     Alaska   0.00 3 1928    Arizona 200.75 4 1928   Arkansas 481.77 5 1928 California  69.22 6 1928   Colorado 206.98   At this point, the data preparation is essentially over. The data is in ‘long’ format, the x, y and z variables for the plot are available and are of the correct type: factor, factor and numeric. If your data is already in this format, it is easy to jump straight into visualisation. But depending on what sort of data you start off with, the data preparation and reformatting can be complicated and tedious.   2. Plotting   I prefer to use the ggplot2 plotting package to plot graphs in R due to its consistent code structure. I will focus mostly on ggplot2 code here. But, just for the sake of completeness, I will also include some heatmap code using base graphics.   2.1 ggplot2   Once the data is in the right format, plotting the data is rather simple code in ggplot2. The ggplot2 index page has the code syntax and arguments.   #basic ggplot p &lt;- ggplot(m3, aes(x=year, y=state, fill=count))+       geom_tile()  #save plot to working directory ggsave(p, filename=\"measles-basic.png\")                           The basic plot produced from ggplot2.                The default output from ggplot2 is quite decent- There are several aspects of the basic plot that can be modified and tweaked. We add tile borders, custom x-axis breaks and custom text sizes. The ggplot code is modified below.   #modified ggplot p &lt;- ggplot(m3, aes(x=year, y=state, fill=count))+   #add border white colour of line thickness 0.25   geom_tile(colour=\"white\", size=0.25)+   #remove x and y axis labels   labs(x=\"\", y=\"\")+   #remove extra space   scale_y_discrete(expand=c(0, 0))+   #define new breaks on x-axis   scale_x_discrete(expand=c(0, 0),                     breaks=c(\"1930\", \"1940\", \"1950\", \"1960\", \"1970\", \"1980\", \"1990\", \"2000\"))+   #set a base size for all fonts   theme_grey(base_size=8)+   #theme options   theme(     #bold font for legend text     legend.text=element_text(face=\"bold\"),     #set thickness of axis ticks     axis.ticks=element_line(size=0.4),     #remove plot background     plot.background=element_blank(),     #remove plot border     panel.border=element_blank()   )  #save with dpi 200 ggsave(p, filename=\"measles-mod1.png\", height=5.5, width=8.8, units=\"in\", dpi=200)                           ggplot2 heatmap after customisation.                I would prefer the y-axis labels (states) to be ordered ascending top-bottom. This means going back to our ‘long’ format data and refactoring the state variable in reverse. The fill variable here (count) is a continuous variable and hence, ggplot by default uses the blue colour ramp. Here and in many cases, it might make better sense to bin the continuous data into levels and represent each level as a discrete colour. The cut() function in R allows to break and label a continuous variable.   The count variable is binned into 7 levels and saved as a new variable countfactor. The NAs remain as NA. Defining breaks in your variable depends on the type of data, the number of bins that make sense with the context or just trial and error. Too many bins are not good. Checking your variable using summary(x) or a boxplot(x) can reveal a lot about the data.   m4 &lt;- m3 %&gt;%   # convert state to factor and reverse order of levels   mutate(state=factor(state, levels=rev(sort(unique(state))))) %&gt;%   # create a new variable from count   mutate(countfactor=cut(count, breaks=c(-1, 0, 1, 10, 100, 500, 1000, max(count, na.rm=T)),                           labels=c(\"0\", \"0-1\", \"1-10\", \"10-100\", \"100-500\", \"500-1000\", \"&gt;1000\"))) %&gt;%   # change level order   mutate(countfactor=factor(as.character(countfactor), levels=rev(levels(countfactor))))   Now we are ready to plot the final dataset.   # assign text colour textcol &lt;- \"grey40\"  # further modified ggplot p &lt;- ggplot(m4, aes(x=year, y=state, fill=countfactor))+   geom_tile(colour=\"white\", size=0.2)+   guides(fill=guide_legend(title=\"Cases per\\n100,000 people\"))+   labs(x=\"\", y=\"\", title=\"Incidence of Measles in the US\")+   scale_y_discrete(expand=c(0, 0))+   scale_x_discrete(expand=c(0, 0), breaks=c(\"1930\", \"1940\", \"1950\", \"1960\", \"1970\", \"1980\", \"1990\", \"2000\"))+   scale_fill_manual(values=c(\"#d53e4f\", \"#f46d43\", \"#fdae61\", \"#fee08b\", \"#e6f598\", \"#abdda4\", \"#ddf1da\"), na.value = \"grey90\")+   #coord_fixed()+   theme_grey(base_size=10)+   theme(legend.position=\"right\", legend.direction=\"vertical\",         legend.title=element_text(colour=textcol),         legend.margin=margin(grid::unit(0, \"cm\")),         legend.text=element_text(colour=textcol, size=7, face=\"bold\"),         legend.key.height=grid::unit(0.8, \"cm\"),         legend.key.width=grid::unit(0.2, \"cm\"),         axis.text.x=element_text(size=10, colour=textcol),         axis.text.y=element_text(vjust=0.2, colour=textcol),         axis.ticks=element_line(size=0.4),         plot.background=element_blank(),         panel.border=element_blank(),         plot.margin=margin(0.7, 0.4, 0.1, 0.2, \"cm\"),         plot.title=element_text(colour=textcol, hjust=0, size=14, face=\"bold\")       )  #export figure ggsave(p, filename=\"measles-mod3.png\", height=5.5, width=8.8, units=\"in\", dpi=200)                           Further customisation of the ggplot2 heatmap.                This is the final version of the figure. All font elements are coloured grey40.  Missing values (NAs) are coloured grey90. The year of the introduction of vaccination is indicated as a dark vertical stripe. If the y-axis labels are too many or too small, they can be dropped the same way as the x-axis. A custom colour palette was used from ColorBrewer based on the Spectral palette. Using the R package RColorBrewer() or using ggplot2 function scale_fill_brewer() opens up all the colour palettes from the ColorBrewer website. Here is an example below:   library(RColorBrewer)  #change the scale_fill_manual from previous code to below scale_fill_manual(values=rev(brewer.pal(7, \"YlGnBu\")), na.value=\"grey90\")+                                       Heatmap with Colorbrewer colour palette.                Most of the plot customisation happens in the theme() section of the code. Refer to the ggplot index for all theme arguments. Based on where the figure will go next, the fonts may need to be changed. The extrafont() package comes handy. Another option is to export in vector format such as .svg or .pdf. Import into a vector editor such as Adobe Illustrator and add your own text. This will produce nice crisp results but does take some manual work. See the cover image of this post for example.   2.2 Base graphics   We will take a quick look at plotting using base graphics. The base function heatmap() and the enchanced heatmap.2() function from the gplots package uses a wide format data matrix as input. We start with the ‘long’ data that we prepared in section 1. We convert the data in ‘long’ format into ‘wide’ format using the spread() function from package tidyr. The wide data is converted to a matrix after removal of non-numeric columns. The state names are reassigned as rownames of the matrix to be used as y-axis text.   # load package library(gplots) # heatmap.2() function library(plotrix) # gradient.rect() function  # convert from long format to wide format m5 &lt;- m3 %&gt;% spread(key=\"state\", value=count) m6 &lt;- as.matrix(m5[ ,-1]) rownames(m6) &lt;- m5$year  #base heatmap png(filename=\"measles-base.png\", height=5.5, width=8.8, res=200, units=\"in\") heatmap(t(m6), Rowv=NA, Colv=NA, na.rm=T, scale=\"none\", col=terrain.colors(100),   xlab=\"\", ylab=\"\", main=\"Incidence of Measles in the US\") dev.off()                           Heatmap using the base function heatmap().                That’s pretty much what we can do with that. The enchanced heatmap.2() function from gplots package can do more. The legend for the colours is not the best, so we will use the function gradient.rect() from plotrix package to create our own legend. And after fiddling around with numerous arguments and countless trial and errors later, we have as below.   # gplots heatmap.2 png(filename=\"measles-gplot.png\", height=6, width=9, res=200, units=\"in\") par(mar=c(2, 3, 3, 2)) gplots::heatmap.2(t(m6), na.rm=T, dendrogram=\"none\", Rowv=NULL, Colv=\"Rowv\", trace=\"none\", scale=\"none\", offsetRow=0.3, offsetCol=0.3, breaks=c(-1, 0, 1, 10, 100, 500, 1000, max(m4$count, na.rm=T)), colsep=which(seq(1928, 2003)%%10==0), margin=c(3, 8), col=rev(c(\"#d53e4f\", \"#f46d43\", \"#fdae61\", \"#fee08b\", \"#e6f598\", \"#abdda4\", \"#ddf1da\")),  xlab=\"\", ylab=\"\", key=F, lhei=c(0.1, 0.9), lwid=c(0.2, 0.8)) gradient.rect(0.125, 0.25, 0.135, 0.75, nslices=7, border=F, gradient=\"y\", col=rev(c(\"#d53e4f\", \"#f46d43\", \"#fdae61\", \"#fee08b\", \"#e6f598\", \"#abdda4\", \"#ddf1da\"))) text(x=rep(0.118, 7), y=seq(0.28, 0.72, by=0.07), adj=1, cex=0.8, labels=c(\"0\", \"0-1\", \"1-10\", \"10-100\", \"100-500\", \"500-1000\", \"&gt;1000\")) text(x=0.135, y=0.82, labels=\"Cases per\\n100,000 people\", adj=1, cex=0.85) title(main=\"Incidence of Measles in the US\", line=1, oma=T, adj=0.21) dev.off()                           Heatmap created using gplots function heatmap.2().                Ultimately, we get something useful. The white separator line is a neat feature. It can be used to group columns or rows as required. The while vertical lines above group decades. And below is the full undisrupted code.   # 2019 | Roy Mathew Francis # Heatmap R code  #load packages library(ggplot2) # ggplot() for plotting library(dplyr) # data reformatting library(tidyr) # data reformatting library(stringr) # string manipulation  # DATA PREPARATION -------------------------------------------------------------  #read csv file m &lt;- read.csv(\"measles_lev1.csv\", header=T, stringsAsFactors=F, skip=2)  m2 &lt;- m %&gt;%   # convert data to long format   gather(key=\"state\", value=\"value\", -YEAR, -WEEK) %&gt;%   # rename columns   setNames(c(\"year\", \"week\", \"state\", \"value\")) %&gt;%   # convert year to factor   mutate(year=factor(year)) %&gt;%   # convert week to factor   mutate(week=factor(week)) %&gt;%   # convert value to numeric (also converts '-' to NA, gives a warning)   mutate(value=as.numeric(value))  # removes . and change states to title case using custom function fn_tc &lt;- function(x) paste(str_to_title(unlist(strsplit(x, \"[.]\"))), collapse=\" \") m2$state &lt;- sapply(m2$state, fn_tc)  # custom sum function returns NA when all values in set are NA, # in a set mixed with NAs, NAs are removed and remaining summed. na_sum &lt;- function(x) {   if(all(is.na(x))) val &lt;- sum(x, na.rm=F)   if(!all(is.na(x))) val &lt;- sum(x, na.rm=T)   return(val) }  # sum incidences for all weeks into one year m3 &lt;- m2 %&gt;%   group_by(year, state) %&gt;%   summarise(count=na_sum(value)) %&gt;%   as.data.frame()  m4 &lt;- m3 %&gt;%   # convert state to factor and reverse order of levels   mutate(state=factor(state, levels=rev(sort(unique(state))))) %&gt;%   # create a new variable from count   mutate(countfactor=cut(count, breaks=c(-1, 0, 1, 10, 100, 500, 1000, max(count, na.rm=T)),                           labels=c(\"0\", \"0-1\", \"1-10\", \"10-100\", \"100-500\", \"500-1000\", \"&gt;1000\"))) %&gt;%   # change level order   mutate(countfactor=factor(as.character(countfactor), levels=rev(levels(countfactor))))  # GGPLOT -----------------------------------------------------------------------  # assign text colour textcol &lt;- \"grey40\"  # further modified ggplot p &lt;- ggplot(m4, aes(x=year, y=state, fill=countfactor))+   geom_tile(colour=\"white\", size=0.2)+   guides(fill=guide_legend(title=\"Cases per\\n100,000 people\"))+   labs(x=\"\", y=\"\", title=\"Incidence of Measles in the US\")+   scale_y_discrete(expand=c(0, 0))+   scale_x_discrete(expand=c(0, 0), breaks=c(\"1930\", \"1940\", \"1950\", \"1960\", \"1970\", \"1980\", \"1990\", \"2000\"))+   scale_fill_manual(values=c(\"#d53e4f\", \"#f46d43\", \"#fdae61\", \"#fee08b\", \"#e6f598\", \"#abdda4\", \"#ddf1da\"), na.value = \"grey90\")+   #coord_fixed()+   theme_grey(base_size=10)+   theme(legend.position=\"right\", legend.direction=\"vertical\",         legend.title=element_text(colour=textcol),         legend.margin=margin(grid::unit(0, \"cm\")),         legend.text=element_text(colour=textcol, size=7, face=\"bold\"),         legend.key.height=grid::unit(0.8, \"cm\"),         legend.key.width=grid::unit(0.2, \"cm\"),         axis.text.x=element_text(size=10, colour=textcol),         axis.text.y=element_text(vjust=0.2, colour=textcol),         axis.ticks=element_line(size=0.4),         plot.background=element_blank(),         panel.border=element_blank(),         plot.margin=margin(0.7, 0.4, 0.1, 0.2, \"cm\"),         plot.title=element_text(colour=textcol, hjust=0, size=14, face=\"bold\")       )  # export figure ggsave(p, filename=\"measles-mod3.png\", height=5.5, width=8.8, units=\"in\", dpi=200)  # BASE GRAPHICS ----------------------------------------------------------------  # load package library(gplots) # heatmap.2() function library(plotrix) # gradient.rect() function  # convert from long format to wide format m5 &lt;- m3 %&gt;% spread(key=\"state\", value=count) m6 &lt;- as.matrix(m5[ ,-1]) rownames(m6) &lt;- m5$year  # base heatmap png(filename=\"measles-base.png\", height=5.5, width=8.8, res=200, units=\"in\") heatmap(t(m6), Rowv=NA, Colv=NA, na.rm=T, scale=\"none\", col=terrain.colors(100),         xlab=\"\", ylab=\"\", main=\"Incidence of Measles in the US\") dev.off()  # gplots heatmap.2 png(filename=\"measles-gplot.png\", height=6, width=9, res=200, units=\"in\") par(mar=c(2, 3, 3, 2)) gplots::heatmap.2(t(m6), na.rm=T, dendrogram=\"none\", Rowv=NULL, Colv=\"Rowv\", trace=\"none\", scale=\"none\",offsetRow=0.3, offsetCol=0.3, breaks=c(-1, 0, 1, 10, 100, 500, 1000, max(m4$count, na.rm=T)), colsep=which(seq(1928, 2003)%%10==0), margin=c(3, 8), col=rev(c(\"#d53e4f\", \"#f46d43\", \"#fdae61\", \"#fee08b\", \"#e6f598\", \"#abdda4\", \"#ddf1da\")), xlab=\"\", ylab=\"\", key=F, lhei=c(0.1, 0.9), lwid=c(0.2, 0.8)) gradient.rect(0.125, 0.25, 0.135, 0.75, nslices=7, border=F, gradient=\"y\", col=rev(c(\"#d53e4f\", \"#f46d43\",\"#fdae61\", \"#fee08b\", \"#e6f598\", \"#abdda4\", \"#ddf1da\"))) text(x=rep(0.118, 7), y=seq(0.28, 0.72, by=0.07), adj=1, cex=0.8, labels=c(\"0\", \"0-1\", \"1-10\", \"10-100\", \"100-500\", \"500-1000\", \"&gt;1000\")) text(x=0.135, y=0.82, labels=\"Cases per\\n100,000 people\", adj=1, cex=0.85) title(main=\"Incidence of Measles in the US\", line=1, oma=T, adj=0.21) dev.off()  # End of script ----------------------------------------------------------------  3. Conclusion   We have covered data prep and plotting heatmaps in R using base graphics and ggplot2. ggplot2 seems more consistent with code structure but the base graphics may be useful when combing multiple graphics in complicated ways. Hacking ggplot2 can be harder than fiddling with base graphics. I have not covered heatmaps with dendrograms because they are only useful in specific situations. Perhaps another useful function for heatmaps in R is pheatmap demonstrated here.  ","categories": ["Code","Research"],
        "tags": ["Data Analysis","Data Visualisation","ggplot2","Heatmap","R"],
        "url": "/a-guide-to-elegant-tiled-heatmaps-in-r-2019/",
        "teaser": null
      },{
        "title": "Road trip through Portugal",
        "excerpt":"I attended the Elixir hands-on meeting in Lisbon, Portugal this summer. This gave me an opportunity to visit Lisbon and take a road trip through Portugal.   I had never been to Lisbon before, so I wasn’t exactly sure what to expect. But, Lisbon was absolutely lovely. The weather was not too hot and the cute winding streets were charming and bustling with people. Parts of the city are hilly and scenic viewpoints are everywhere. The city is full of historic buildings, museums, restaurants and cafes. I thought the Maritime museum was very interesting. The food was reasonably priced and delicious. I also happened to be in perfect time for the Lisbon street festivals which meant that there were parties and dancing in the streets almost all through the city.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           Views of Lisbon.          After the conference, I rented a car and drove down South to Algarve exploring the countryside.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           Views from the countryside.          There are several beautiful sandy beaches along the west coast as well as the South. Few of them that I visited were Praia da Zambujeira do Mar, Praia de Odeceixe, Praia da Cordoama and Praia do Zavial.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           Views of various beaches.          Cape St. Vincent (Farol do Cabo de São Vicente) is the Southwestern tip of Portugal, which also happens to be the most Southwestern point of continental Europe. The lighthouse and rocky cliffs are magnificent. I arrived here in the evening just before sunset. The golden hour was amazing and then out of nowhere, the fog came in. The heavy fog drifting across the cliffs and over the peninsula was an amazing sight from the drone.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Views around Cape St. Vicente.          The coastal areas around Lagos, Portimão and Faro have rather interesting rock formations and cave systems. Some of them are Benagil caves (Algar de Benagil), Marinha beach (Praia da Marinha), Arco de Albandeira and Algar Seco. I stayed in Albufeira for a couple of days to check out the attractions. I also went on a dolphin watching trip which was exciting.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             Views from Algarve.          If you are into pirates, you should check out the Pirate shop in Sagres and the Pirate themed B&amp;B in Charneca de Caparica.  ","categories": ["Photography","Travel"],
        "tags": ["Portugal","Road Trip"],
        "url": "/portugal/",
        "teaser": null
      },{
        "title": "Exploring the Swedish west coast",
        "excerpt":"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Views from the west coast of Sweden.          The west coast of Sweden is well known for it’s rocky archipelago and scenic beauty. Also known for seafood, sea kayaking and floating saunas. I took a trip through Lidköping and Trollhättan to Lysekil, Fjällbacka and Grebbestad.   In Trollhättan, the Saab car museum, Trollhättan canal and Olidan power station were interesting. About 30 mins north of Lysekil is the animal park Nordens ark. It’s a fun place if you are into animals.   I slowly drove around between Lysekil and Grebbestad stopping at every scenic spot. There are also many camping spots around the area which makes it convenient. I stayed at Fjällbacka camping for one of the nights was a really cute campsite with glorious views over the archipelago.  ","categories": ["Photography","Travel"],
        "tags": ["Sweden","Road Trip"],
        "url": "/sweden-west-coast/",
        "teaser": null
      },{
        "title": "Scatterplot annotation",
        "excerpt":"In this post, we are exploring ideas to mark clusters of points on a scatterplot for labelling purposes. Or in other words, how to draw polygons around scatterplots. We use ggplot2 for plotting and few different functions to generate the markings. The required packages are shown below. A custom ggplot2 theme is used to simplify the plot.   library(dplyr) library(ggplot2) library(ggforce) library(ggalt) data(iris)  theme_custom &lt;- function (basesize=14) {   theme_bw(base_size=basesize) %+replace%     theme(       panel.border=element_blank(),       panel.grid=element_blank(),       legend.position=\"top\",       legend.direction=\"horizontal\",       legend.justification=\"right\",       strip.background=element_blank(),       axis.ticks=element_blank(),       axis.title=element_blank(),       axis.text=element_blank()     ) }   We are using the iris dataset using the three different species as groups/clusters.   iris %&gt;%   ggplot(aes(Petal.Length, Petal.Width, colour=Species))+   geom_point()+   theme_custom()                               Base   First we start by using base functions.   Ellipse   Perhaps the easiest approach is drawing confidence ellipse. All points are not guaranteed to be inside the ellipse. The default level is 0.95. Increasing this includes more of the points, but the ellipses may get too large. The nice thing about this approach is that it is not strongly affected by outliers. This is well suited for clusters where points have a large spread and overlap a lot.   ggplot(data=iris, aes(Petal.Length, Petal.Width, colour=Species))+   geom_point()+   stat_ellipse(level=0.95)+   theme_custom()                           Fig. 1: Clusters marked by 95% confidence ellipse.                For most part, this is the easiest approach and good enough.   Convex hull   The outermost points in a point cloud can be connected to form a polygon referred to as the convex hull of a point cloud. The convex hull can be computed using the base function chull(). This can be used in a dplyr pipeline as shown below.   hull &lt;- iris %&gt;%   group_by(Species) %&gt;%  slice(c(chull(Petal.Length, Petal.Width),          chull(Petal.Length, Petal.Width)[1]))   The information to draw polygons around all three clusters is contained in the data.frame hull. The points and the polygons can be plotted as shown below.   ggplot()+   geom_point(data=iris, aes(Petal.Length, Petal.Width, colour=Species))+   geom_path(data=hull, aes(Petal.Length, Petal.Width, colour=Species))+   theme_custom()                           Fig. 2: Clusters marked by convex hull.                This is a good start, but it would be nicer if the edges of the polygon had smooth edges rather than sharp edges. While we could use some custom script to smooth it, the easier approach is to use from existing packages.   Density and contours   Before we move on, another idea from the base functions is to compute a 2D kernal density from the point cloud and draw countour lines.   iris %&gt;%   ggplot(aes(Petal.Length, Petal.Width, colour=Species))+   geom_point()+   geom_density_2d()+   theme_custom()                           Fig. 3: Contour lines.                ggalt  geom_encircle   We use the function geom_encircle() from package ggalt to draw smoothed lines around clusters.   ggplot(data=iris, aes(Petal.Length, Petal.Width, colour=Species))+   geom_point()+   ggalt::geom_encircle(size=1.6)+   theme_custom()                           Fig. 4: Clusters encircled using geom_encircle.                ggforce   Another useful package for marking clusters is ggforce. Below is a demonstration of four functions for marking clusters. The four functions are geom_mark_rect(), geom_mark_circle(), geom_mark_ellipse() and geom_mark_hull(). The shapes they create is self-explanatory.   p1 &lt;- ggplot(data=iris, aes(Petal.Length, Petal.Width, colour=Species))+   geom_point()+   geom_mark_rect()+   theme_custom()  p2 &lt;- ggplot(data=iris, aes(Petal.Length, Petal.Width, colour=Species))+   geom_point()+   geom_mark_circle()+   theme_custom()  p3 &lt;- ggplot(data=iris, aes(Petal.Length, Petal.Width, colour=Species))+   geom_point()+   geom_mark_ellipse()+   theme_custom()  p4 &lt;- ggplot(data=iris, aes(Petal.Length, Petal.Width, colour=Species))+   geom_point()+   geom_mark_hull()+   theme_custom()  ggpubr::ggarrange(p1, p2, p3, p4, nrow=2, ncol=2, common.legend=T)                           Fig. 5: Clusters demarcated using rectangles (top-left), circles (top-right), ellipses (bottom-left) and smoothed convex hull (bottom-right).                geom_mark_hull() has extra arguments to control concavity, radius and size (expand) of the hull shape. geom_mark_* functions also support an argument label to label the the marks. In this case, the plot legend can be turned off.   ggplot(data=iris, aes(Petal.Length, Petal.Width, colour=Species))+   geom_point()+   geom_mark_ellipse(aes(label=Species), label.colour=\"grey30\")+   theme_custom()+   theme(legend.position=\"none\")                           Fig. 6: Clusters marked and labelled using geom_mark_ellipse().                It is not necessary to mark all the groups. A particular group of interest can be marked alone.   ggplot()+   geom_point(data=iris, aes(Petal.Length, Petal.Width))+   geom_mark_ellipse(data=filter(iris, Species==\"setosa\"), aes(Petal.Length, Petal.Width, label=Species), label.colour=\"grey30\", description=\"This is a brief description about this annotation.\")+   theme_custom()+   theme(legend.position=\"none\")   For more control over text annotation, see package ggrepel.   The ggforce package also has a function called geom_shape() to draw polygons but with some neat features. We can adjust the corner radius of the polygon to smooth the edges. We can also increase or decrease the overall size of the polygon. In this case, we use the convex hull object hull which we computed earlier.   p1 &lt;- ggplot()+   geom_point(data=iris, aes(Petal.Length, Petal.Width, colour=Species))+   geom_shape(data=hull, aes(Petal.Length, Petal.Width, colour=Species), fill=NA)+   theme_custom() p2 &lt;- ggplot()+   geom_point(data=iris, aes(Petal.Length, Petal.Width, colour=Species))+   geom_shape(data=hull, aes(Petal.Length, Petal.Width, colour=Species), fill=NA, radius=0.05)+   theme_custom() p3 &lt;- ggplot()+   geom_point(data=iris, aes(Petal.Length, Petal.Width, colour=Species))+   geom_shape(data=hull, aes(Petal.Length, Petal.Width, colour=Species), fill=NA, radius=0.05, expand=0.04)+   theme_custom() ggpubr::ggarrange(p1, p2, p3, nrow=2, ncol=2, common.legend=T)                           Fig. 8: Clusters marked using manually computed convex hull. geom_shape from ggforce is used. Standard convex hull (top-left), smoothed edges (top-right) and smoothed+enlarged (bottom-left).                The package ggforce has many more cool features. You should check out their website.  ","categories": ["Code","Research"],
        "tags": ["Data Visualisation","Heatmap","R","Scientific Graphs"],
        "url": "/scatterplot-annotation/",
        "teaser": null
      },{
        "title": "Exploring history and culture in Thessaloniki",
        "excerpt":"I took a short autumn trip to Thessaloniki, Greece.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Views of Thessaloniki and Platamon.          This place is most interesting for it’s culture and history. You can learn all about the Greek civilisation and the evolution of human civilisation in general. Some of my favourite musuems were the Archaeological Museum of Thessaloniki, War Museum of Thessaloniki and the white tower museum. The Museum of Byzantine Culture was unfortunately closed. There are numerous historical churches and archaeological sites scattered through the city.   There are a few places to get some great views of the city. The white tower museum is the most obvious, but it is not that high. The skyline cafe bar offers excellent 360 degree views of the city as you stuff your pie-hole. Another site is the Trigonion Tower which offers stunning views over the whole city.   On the last day, I took a train ride down to Platamon which is about 1.5 hrs away from Thessaloniki. The beaches were practically empty in November and to be fair, it wasn’t very warm to be in the water. But, I had an excellent fried fish dinner at a local restaurant.  ","categories": ["Photography","Travel"],
        "tags": ["Greece"],
        "url": "/greece-thessaloniki/",
        "teaser": null
      }]
